// @aztec/bb.js/dest/browser/cbind/generated/api_types.js
// AUTOGENERATED FILE - DO NOT EDIT
// Conversion functions (exported)
export function toAvmStat(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in AvmStat deserialization");
    }
    if (o.value_ms === undefined) {
        throw new Error("Expected value_ms in AvmStat deserialization");
    }
    ;
    return {
        name: o.name,
        valueMs: o.value_ms,
    };
}
export function toCircuitComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in CircuitComputeVkResponse deserialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in CircuitComputeVkResponse deserialization");
    }
    if (o.hash === undefined) {
        throw new Error("Expected hash in CircuitComputeVkResponse deserialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields,
        hash: o.hash,
    };
}
export function toAvmProve(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in AvmProve deserialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function toAvmVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in AvmVerify deserialization");
    }
    if (o.public_inputs === undefined) {
        throw new Error("Expected public_inputs in AvmVerify deserialization");
    }
    ;
    return {
        proof: o.proof,
        publicInputs: o.public_inputs,
    };
}
export function toAvmCheckCircuit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in AvmCheckCircuit deserialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function toCircuitInput(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInput deserialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInput deserialization");
    }
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitInput deserialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
        verificationKey: o.verification_key,
    };
}
export function toProofSystemSettings(o) {
    if (o.ipa_accumulation === undefined) {
        throw new Error("Expected ipa_accumulation in ProofSystemSettings deserialization");
    }
    if (o.oracle_hash_type === undefined) {
        throw new Error("Expected oracle_hash_type in ProofSystemSettings deserialization");
    }
    if (o.disable_zk === undefined) {
        throw new Error("Expected disable_zk in ProofSystemSettings deserialization");
    }
    if (o.optimized_solidity_verifier === undefined) {
        throw new Error("Expected optimized_solidity_verifier in ProofSystemSettings deserialization");
    }
    ;
    return {
        ipaAccumulation: o.ipa_accumulation,
        oracleHashType: o.oracle_hash_type,
        disableZk: o.disable_zk,
        optimizedSolidityVerifier: o.optimized_solidity_verifier,
    };
}
export function toCircuitProve(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitProve deserialization");
    }
    if (o.witness === undefined) {
        throw new Error("Expected witness in CircuitProve deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitProve deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
        witness: o.witness,
        settings: toProofSystemSettings(o.settings),
    };
}
export function toCircuitInputNoVK(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInputNoVK deserialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInputNoVK deserialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
    };
}
export function toCircuitComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitComputeVk deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitComputeVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
        settings: toProofSystemSettings(o.settings),
    };
}
export function toCircuitStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitStats deserialization");
    }
    if (o.include_gates_per_opcode === undefined) {
        throw new Error("Expected include_gates_per_opcode in CircuitStats deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitStats deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
        includeGatesPerOpcode: o.include_gates_per_opcode,
        settings: toProofSystemSettings(o.settings),
    };
}
export function toCircuitVerify(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitVerify deserialization");
    }
    if (o.public_inputs === undefined) {
        throw new Error("Expected public_inputs in CircuitVerify deserialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitVerify deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitVerify deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
        publicInputs: o.public_inputs,
        proof: o.proof,
        settings: toProofSystemSettings(o.settings),
    };
}
export function toChonkComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkComputeVk deserialization");
    }
    if (o.use_zk_flavor === undefined) {
        throw new Error("Expected use_zk_flavor in ChonkComputeVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
        useZkFlavor: o.use_zk_flavor,
    };
}
export function toChonkStart(o) {
    if (o.num_circuits === undefined) {
        throw new Error("Expected num_circuits in ChonkStart deserialization");
    }
    ;
    return {
        numCircuits: o.num_circuits,
    };
}
export function toChonkLoad(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkLoad deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
    };
}
export function toChonkAccumulate(o) {
    if (o.witness === undefined) {
        throw new Error("Expected witness in ChonkAccumulate deserialization");
    }
    ;
    return {
        witness: o.witness,
    };
}
export function toChonkProve(o) {
    return {};
}
export function toChonkProof(o) {
    if (o.hiding_oink_proof === undefined) {
        throw new Error("Expected hiding_oink_proof in ChonkProof deserialization");
    }
    if (o.merge_proof === undefined) {
        throw new Error("Expected merge_proof in ChonkProof deserialization");
    }
    if (o.eccvm_proof === undefined) {
        throw new Error("Expected eccvm_proof in ChonkProof deserialization");
    }
    if (o.ipa_proof === undefined) {
        throw new Error("Expected ipa_proof in ChonkProof deserialization");
    }
    if (o.joint_proof === undefined) {
        throw new Error("Expected joint_proof in ChonkProof deserialization");
    }
    ;
    return {
        hidingOinkProof: o.hiding_oink_proof,
        mergeProof: o.merge_proof,
        eccvmProof: o.eccvm_proof,
        ipaProof: o.ipa_proof,
        jointProof: o.joint_proof,
    };
}
export function toChonkVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkVerify deserialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ChonkVerify deserialization");
    }
    ;
    return {
        proof: toChonkProof(o.proof),
        vk: o.vk,
    };
}
export function toChonkVerifyFromFields(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkVerifyFromFields deserialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ChonkVerifyFromFields deserialization");
    }
    ;
    return {
        proof: o.proof,
        vk: o.vk,
    };
}
export function toChonkBatchVerify(o) {
    if (o.proofs === undefined) {
        throw new Error("Expected proofs in ChonkBatchVerify deserialization");
    }
    if (o.vks === undefined) {
        throw new Error("Expected vks in ChonkBatchVerify deserialization");
    }
    ;
    return {
        proofs: o.proofs.map((v) => toChonkProof(v)),
        vks: o.vks,
    };
}
export function toVkAsFields(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in VkAsFields deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
    };
}
export function toMegaVkAsFields(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in MegaVkAsFields deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
    };
}
export function toCircuitWriteSolidityVerifier(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitWriteSolidityVerifier deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitWriteSolidityVerifier deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
        settings: toProofSystemSettings(o.settings),
    };
}
export function toChonkCheckPrecomputedVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkCheckPrecomputedVk deserialization");
    }
    if (o.use_zk_flavor === undefined) {
        throw new Error("Expected use_zk_flavor in ChonkCheckPrecomputedVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
        useZkFlavor: o.use_zk_flavor,
    };
}
export function toChonkStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkStats deserialization");
    }
    if (o.include_gates_per_opcode === undefined) {
        throw new Error("Expected include_gates_per_opcode in ChonkStats deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
        includeGatesPerOpcode: o.include_gates_per_opcode,
    };
}
export function toChonkCompressProof(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkCompressProof deserialization");
    }
    ;
    return {
        proof: toChonkProof(o.proof),
    };
}
export function toChonkDecompressProof(o) {
    if (o.compressed_proof === undefined) {
        throw new Error("Expected compressed_proof in ChonkDecompressProof deserialization");
    }
    ;
    return {
        compressedProof: o.compressed_proof,
    };
}
export function toPoseidon2Hash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Hash deserialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function toPoseidon2Permutation(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Permutation deserialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function toPedersenCommit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenCommit deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenCommit deserialization");
    }
    ;
    return {
        inputs: o.inputs,
        hashIndex: o.hash_index,
    };
}
export function toPedersenHash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenHash deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenHash deserialization");
    }
    ;
    return {
        inputs: o.inputs,
        hashIndex: o.hash_index,
    };
}
export function toPedersenHashBuffer(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in PedersenHashBuffer deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenHashBuffer deserialization");
    }
    ;
    return {
        input: o.input,
        hashIndex: o.hash_index,
    };
}
export function toBlake2s(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2s deserialization");
    }
    ;
    return {
        data: o.data,
    };
}
export function toBlake2sToField(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2sToField deserialization");
    }
    ;
    return {
        data: o.data,
    };
}
export function toAesEncrypt(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesEncrypt deserialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesEncrypt deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesEncrypt deserialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesEncrypt deserialization");
    }
    ;
    return {
        plaintext: o.plaintext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
export function toAesDecrypt(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesDecrypt deserialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesDecrypt deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesDecrypt deserialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesDecrypt deserialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
export function toGrumpkinPoint(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in GrumpkinPoint deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in GrumpkinPoint deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function toGrumpkinMul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinMul deserialization");
    }
    ;
    return {
        point: toGrumpkinPoint(o.point),
        scalar: o.scalar,
    };
}
export function toGrumpkinAdd(o) {
    if (o.point_a === undefined) {
        throw new Error("Expected point_a in GrumpkinAdd deserialization");
    }
    if (o.point_b === undefined) {
        throw new Error("Expected point_b in GrumpkinAdd deserialization");
    }
    ;
    return {
        pointA: toGrumpkinPoint(o.point_a),
        pointB: toGrumpkinPoint(o.point_b),
    };
}
export function toGrumpkinBatchMul(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinBatchMul deserialization");
    }
    ;
    return {
        points: o.points.map((v) => toGrumpkinPoint(v)),
        scalar: o.scalar,
    };
}
export function toGrumpkinGetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in GrumpkinGetRandomFr deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function toGrumpkinReduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in GrumpkinReduce512 deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function toSecp256k1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Secp256k1Point deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Secp256k1Point deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function toSecp256k1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1Mul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Secp256k1Mul deserialization");
    }
    ;
    return {
        point: toSecp256k1Point(o.point),
        scalar: o.scalar,
    };
}
export function toSecp256k1GetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in Secp256k1GetRandomFr deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function toSecp256k1Reduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Secp256k1Reduce512 deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function toBn254FrSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FrSqrt deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function toBn254FqSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FqSqrt deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function toBn254G1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Bn254G1Point deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Bn254G1Point deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function toBn254G1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1Mul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Bn254G1Mul deserialization");
    }
    ;
    return {
        point: toBn254G1Point(o.point),
        scalar: o.scalar,
    };
}
export function toBn254G2Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Bn254G2Point deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Bn254G2Point deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function toBn254G2Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G2Mul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Bn254G2Mul deserialization");
    }
    ;
    return {
        point: toBn254G2Point(o.point),
        scalar: o.scalar,
    };
}
export function toBn254G1IsOnCurve(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1IsOnCurve deserialization");
    }
    ;
    return {
        point: toBn254G1Point(o.point),
    };
}
export function toBn254G1FromCompressed(o) {
    if (o.compressed === undefined) {
        throw new Error("Expected compressed in Bn254G1FromCompressed deserialization");
    }
    ;
    return {
        compressed: o.compressed,
    };
}
export function toSchnorrComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in SchnorrComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
export function toSchnorrConstructSignature(o) {
    if (o.message_field === undefined) {
        throw new Error("Expected message_field in SchnorrConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in SchnorrConstructSignature deserialization");
    }
    ;
    return {
        messageField: o.message_field,
        privateKey: o.private_key,
    };
}
export function toSchnorrVerifySignature(o) {
    if (o.message_field === undefined) {
        throw new Error("Expected message_field in SchnorrVerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in SchnorrVerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrVerifySignature deserialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrVerifySignature deserialization");
    }
    ;
    return {
        messageField: o.message_field,
        publicKey: toGrumpkinPoint(o.public_key),
        s: o.s,
        e: o.e,
    };
}
export function toEcdsaSecp256k1ComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256k1ComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
export function toEcdsaSecp256r1ComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256r1ComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
export function toEcdsaSecp256k1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1ConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256k1ConstructSignature deserialization");
    }
    ;
    return {
        message: o.message,
        privateKey: o.private_key,
    };
}
export function toEcdsaSecp256r1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1ConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256r1ConstructSignature deserialization");
    }
    ;
    return {
        message: o.message,
        privateKey: o.private_key,
    };
}
export function toEcdsaSecp256k1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toEcdsaSecp256r1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toEcdsaSecp256k1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1VerifySignature deserialization");
    }
    ;
    return {
        message: o.message,
        publicKey: toSecp256k1Point(o.public_key),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toSecp256r1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Secp256r1Point deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Secp256r1Point deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function toEcdsaSecp256r1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1VerifySignature deserialization");
    }
    ;
    return {
        message: o.message,
        publicKey: toSecp256r1Point(o.public_key),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toSrsInitSrs(o) {
    if (o.points_buf === undefined) {
        throw new Error("Expected points_buf in SrsInitSrs deserialization");
    }
    if (o.num_points === undefined) {
        throw new Error("Expected num_points in SrsInitSrs deserialization");
    }
    if (o.g2_point === undefined) {
        throw new Error("Expected g2_point in SrsInitSrs deserialization");
    }
    ;
    return {
        pointsBuf: o.points_buf,
        numPoints: o.num_points,
        g2Point: o.g2_point,
    };
}
export function toChonkBatchVerifierStart(o) {
    if (o.vks === undefined) {
        throw new Error("Expected vks in ChonkBatchVerifierStart deserialization");
    }
    if (o.num_cores === undefined) {
        throw new Error("Expected num_cores in ChonkBatchVerifierStart deserialization");
    }
    if (o.batch_size === undefined) {
        throw new Error("Expected batch_size in ChonkBatchVerifierStart deserialization");
    }
    if (o.fifo_path === undefined) {
        throw new Error("Expected fifo_path in ChonkBatchVerifierStart deserialization");
    }
    ;
    return {
        vks: o.vks,
        numCores: o.num_cores,
        batchSize: o.batch_size,
        fifoPath: o.fifo_path,
    };
}
export function toChonkBatchVerifierQueue(o) {
    if (o.request_id === undefined) {
        throw new Error("Expected request_id in ChonkBatchVerifierQueue deserialization");
    }
    if (o.vk_index === undefined) {
        throw new Error("Expected vk_index in ChonkBatchVerifierQueue deserialization");
    }
    if (o.proof_fields === undefined) {
        throw new Error("Expected proof_fields in ChonkBatchVerifierQueue deserialization");
    }
    ;
    return {
        requestId: o.request_id,
        vkIndex: o.vk_index,
        proofFields: o.proof_fields,
    };
}
export function toChonkBatchVerifierStop(o) {
    return {};
}
export function toSrsInitGrumpkinSrs(o) {
    if (o.points_buf === undefined) {
        throw new Error("Expected points_buf in SrsInitGrumpkinSrs deserialization");
    }
    if (o.num_points === undefined) {
        throw new Error("Expected num_points in SrsInitGrumpkinSrs deserialization");
    }
    ;
    return {
        pointsBuf: o.points_buf,
        numPoints: o.num_points,
    };
}
export function toShutdown(o) {
    return {};
}
export function toErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in ErrorResponse deserialization");
    }
    ;
    return {
        message: o.message,
    };
}
export function toAvmProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in AvmProveResponse deserialization");
    }
    if (o.stats === undefined) {
        throw new Error("Expected stats in AvmProveResponse deserialization");
    }
    ;
    return {
        proof: o.proof,
        stats: o.stats.map((v) => toAvmStat(v)),
    };
}
export function toAvmVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in AvmVerifyResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function toAvmCheckCircuitResponse(o) {
    if (o.passed === undefined) {
        throw new Error("Expected passed in AvmCheckCircuitResponse deserialization");
    }
    if (o.stats === undefined) {
        throw new Error("Expected stats in AvmCheckCircuitResponse deserialization");
    }
    ;
    return {
        passed: o.passed,
        stats: o.stats.map((v) => toAvmStat(v)),
    };
}
export function toCircuitProveResponse(o) {
    if (o.public_inputs === undefined) {
        throw new Error("Expected public_inputs in CircuitProveResponse deserialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitProveResponse deserialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in CircuitProveResponse deserialization");
    }
    ;
    return {
        publicInputs: o.public_inputs,
        proof: o.proof,
        vk: toCircuitComputeVkResponse(o.vk),
    };
}
export function toCircuitInfoResponse(o) {
    if (o.num_gates === undefined) {
        throw new Error("Expected num_gates in CircuitInfoResponse deserialization");
    }
    if (o.num_gates_dyadic === undefined) {
        throw new Error("Expected num_gates_dyadic in CircuitInfoResponse deserialization");
    }
    if (o.num_acir_opcodes === undefined) {
        throw new Error("Expected num_acir_opcodes in CircuitInfoResponse deserialization");
    }
    if (o.gates_per_opcode === undefined) {
        throw new Error("Expected gates_per_opcode in CircuitInfoResponse deserialization");
    }
    ;
    return {
        numGates: o.num_gates,
        numGatesDyadic: o.num_gates_dyadic,
        numAcirOpcodes: o.num_acir_opcodes,
        gatesPerOpcode: o.gates_per_opcode,
    };
}
export function toCircuitVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in CircuitVerifyResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function toChonkComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ChonkComputeVkResponse deserialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in ChonkComputeVkResponse deserialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields,
    };
}
export function toChonkStartResponse(o) {
    return {};
}
export function toChonkLoadResponse(o) {
    return {};
}
export function toChonkAccumulateResponse(o) {
    return {};
}
export function toChonkProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkProveResponse deserialization");
    }
    ;
    return {
        proof: toChonkProof(o.proof),
    };
}
export function toChonkVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkVerifyResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function toChonkVerifyFromFieldsResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkVerifyFromFieldsResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function toChonkBatchVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkBatchVerifyResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function toVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in VkAsFieldsResponse deserialization");
    }
    ;
    return {
        fields: o.fields,
    };
}
export function toMegaVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in MegaVkAsFieldsResponse deserialization");
    }
    ;
    return {
        fields: o.fields,
    };
}
export function toCircuitWriteSolidityVerifierResponse(o) {
    if (o.solidity_code === undefined) {
        throw new Error("Expected solidity_code in CircuitWriteSolidityVerifierResponse deserialization");
    }
    ;
    return {
        solidityCode: o.solidity_code,
    };
}
export function toChonkCheckPrecomputedVkResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkCheckPrecomputedVkResponse deserialization");
    }
    if (o.actual_vk === undefined) {
        throw new Error("Expected actual_vk in ChonkCheckPrecomputedVkResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
        actualVk: o.actual_vk,
    };
}
export function toChonkStatsResponse(o) {
    if (o.acir_opcodes === undefined) {
        throw new Error("Expected acir_opcodes in ChonkStatsResponse deserialization");
    }
    if (o.circuit_size === undefined) {
        throw new Error("Expected circuit_size in ChonkStatsResponse deserialization");
    }
    if (o.gates_per_opcode === undefined) {
        throw new Error("Expected gates_per_opcode in ChonkStatsResponse deserialization");
    }
    ;
    return {
        acirOpcodes: o.acir_opcodes,
        circuitSize: o.circuit_size,
        gatesPerOpcode: o.gates_per_opcode,
    };
}
export function toChonkCompressProofResponse(o) {
    if (o.compressed_proof === undefined) {
        throw new Error("Expected compressed_proof in ChonkCompressProofResponse deserialization");
    }
    ;
    return {
        compressedProof: o.compressed_proof,
    };
}
export function toChonkDecompressProofResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkDecompressProofResponse deserialization");
    }
    ;
    return {
        proof: toChonkProof(o.proof),
    };
}
export function toPoseidon2HashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function toPoseidon2PermutationResponse(o) {
    if (o.outputs === undefined) {
        throw new Error("Expected outputs in Poseidon2PermutationResponse deserialization");
    }
    ;
    return {
        outputs: o.outputs,
    };
}
export function toPedersenCommitResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in PedersenCommitResponse deserialization");
    }
    ;
    return {
        point: toGrumpkinPoint(o.point),
    };
}
export function toPedersenHashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function toPedersenHashBufferResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashBufferResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function toBlake2sResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Blake2sResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function toBlake2sToFieldResponse(o) {
    if (o.field === undefined) {
        throw new Error("Expected field in Blake2sToFieldResponse deserialization");
    }
    ;
    return {
        field: o.field,
    };
}
export function toAesEncryptResponse(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesEncryptResponse deserialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
    };
}
export function toAesDecryptResponse(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesDecryptResponse deserialization");
    }
    ;
    return {
        plaintext: o.plaintext,
    };
}
export function toGrumpkinMulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMulResponse deserialization");
    }
    ;
    return {
        point: toGrumpkinPoint(o.point),
    };
}
export function toGrumpkinAddResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinAddResponse deserialization");
    }
    ;
    return {
        point: toGrumpkinPoint(o.point),
    };
}
export function toGrumpkinBatchMulResponse(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMulResponse deserialization");
    }
    ;
    return {
        points: o.points.map((v) => toGrumpkinPoint(v)),
    };
}
export function toGrumpkinGetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinGetRandomFrResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function toGrumpkinReduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinReduce512Response deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function toSecp256k1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1MulResponse deserialization");
    }
    ;
    return {
        point: toSecp256k1Point(o.point),
    };
}
export function toSecp256k1GetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1GetRandomFrResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function toSecp256k1Reduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1Reduce512Response deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function toBn254FrSqrtResponse(o) {
    if (o.is_square_root === undefined) {
        throw new Error("Expected is_square_root in Bn254FrSqrtResponse deserialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FrSqrtResponse deserialization");
    }
    ;
    return {
        isSquareRoot: o.is_square_root,
        value: o.value,
    };
}
export function toBn254FqSqrtResponse(o) {
    if (o.is_square_root === undefined) {
        throw new Error("Expected is_square_root in Bn254FqSqrtResponse deserialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FqSqrtResponse deserialization");
    }
    ;
    return {
        isSquareRoot: o.is_square_root,
        value: o.value,
    };
}
export function toBn254G1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1MulResponse deserialization");
    }
    ;
    return {
        point: toBn254G1Point(o.point),
    };
}
export function toBn254G2MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G2MulResponse deserialization");
    }
    ;
    return {
        point: toBn254G2Point(o.point),
    };
}
export function toBn254G1IsOnCurveResponse(o) {
    if (o.is_on_curve === undefined) {
        throw new Error("Expected is_on_curve in Bn254G1IsOnCurveResponse deserialization");
    }
    ;
    return {
        isOnCurve: o.is_on_curve,
    };
}
export function toBn254G1FromCompressedResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1FromCompressedResponse deserialization");
    }
    ;
    return {
        point: toBn254G1Point(o.point),
    };
}
export function toSchnorrComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in SchnorrComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toGrumpkinPoint(o.public_key),
    };
}
export function toSchnorrConstructSignatureResponse(o) {
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrConstructSignatureResponse deserialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrConstructSignatureResponse deserialization");
    }
    ;
    return {
        s: o.s,
        e: o.e,
    };
}
export function toSchnorrVerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in SchnorrVerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function toEcdsaSecp256k1ComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1ComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toSecp256k1Point(o.public_key),
    };
}
export function toEcdsaSecp256r1ComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1ComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toSecp256r1Point(o.public_key),
    };
}
export function toEcdsaSecp256k1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toEcdsaSecp256r1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function toEcdsaSecp256k1RecoverPublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1RecoverPublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toSecp256k1Point(o.public_key),
    };
}
export function toEcdsaSecp256r1RecoverPublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1RecoverPublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toSecp256r1Point(o.public_key),
    };
}
export function toEcdsaSecp256k1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function toEcdsaSecp256r1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function toSrsInitSrsResponse(o) {
    if (o.points_buf === undefined) {
        throw new Error("Expected points_buf in SrsInitSrsResponse deserialization");
    }
    ;
    return {
        pointsBuf: o.points_buf,
    };
}
export function toChonkBatchVerifierStartResponse(o) {
    return {};
}
export function toChonkBatchVerifierQueueResponse(o) {
    return {};
}
export function toChonkBatchVerifierStopResponse(o) {
    return {};
}
export function toSrsInitGrumpkinSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function toShutdownResponse(o) {
    return {};
}
export function fromAvmStat(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in AvmStat serialization");
    }
    if (o.valueMs === undefined) {
        throw new Error("Expected valueMs in AvmStat serialization");
    }
    ;
    return {
        name: o.name,
        value_ms: o.valueMs,
    };
}
export function fromCircuitComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in CircuitComputeVkResponse serialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in CircuitComputeVkResponse serialization");
    }
    if (o.hash === undefined) {
        throw new Error("Expected hash in CircuitComputeVkResponse serialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields,
        hash: o.hash,
    };
}
export function fromAvmProve(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in AvmProve serialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function fromAvmVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in AvmVerify serialization");
    }
    if (o.publicInputs === undefined) {
        throw new Error("Expected publicInputs in AvmVerify serialization");
    }
    ;
    return {
        proof: o.proof,
        public_inputs: o.publicInputs,
    };
}
export function fromAvmCheckCircuit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in AvmCheckCircuit serialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function fromCircuitInput(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInput serialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInput serialization");
    }
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitInput serialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
        verification_key: o.verificationKey,
    };
}
export function fromProofSystemSettings(o) {
    if (o.ipaAccumulation === undefined) {
        throw new Error("Expected ipaAccumulation in ProofSystemSettings serialization");
    }
    if (o.oracleHashType === undefined) {
        throw new Error("Expected oracleHashType in ProofSystemSettings serialization");
    }
    if (o.disableZk === undefined) {
        throw new Error("Expected disableZk in ProofSystemSettings serialization");
    }
    if (o.optimizedSolidityVerifier === undefined) {
        throw new Error("Expected optimizedSolidityVerifier in ProofSystemSettings serialization");
    }
    ;
    return {
        ipa_accumulation: o.ipaAccumulation,
        oracle_hash_type: o.oracleHashType,
        disable_zk: o.disableZk,
        optimized_solidity_verifier: o.optimizedSolidityVerifier,
    };
}
export function fromCircuitProve(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitProve serialization");
    }
    if (o.witness === undefined) {
        throw new Error("Expected witness in CircuitProve serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitProve serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
        witness: o.witness,
        settings: fromProofSystemSettings(o.settings),
    };
}
export function fromCircuitInputNoVK(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInputNoVK serialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInputNoVK serialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
    };
}
export function fromCircuitComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitComputeVk serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitComputeVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
        settings: fromProofSystemSettings(o.settings),
    };
}
export function fromCircuitStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitStats serialization");
    }
    if (o.includeGatesPerOpcode === undefined) {
        throw new Error("Expected includeGatesPerOpcode in CircuitStats serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitStats serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
        include_gates_per_opcode: o.includeGatesPerOpcode,
        settings: fromProofSystemSettings(o.settings),
    };
}
export function fromCircuitVerify(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitVerify serialization");
    }
    if (o.publicInputs === undefined) {
        throw new Error("Expected publicInputs in CircuitVerify serialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitVerify serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitVerify serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
        public_inputs: o.publicInputs,
        proof: o.proof,
        settings: fromProofSystemSettings(o.settings),
    };
}
export function fromChonkComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkComputeVk serialization");
    }
    if (o.useZkFlavor === undefined) {
        throw new Error("Expected useZkFlavor in ChonkComputeVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
        use_zk_flavor: o.useZkFlavor,
    };
}
export function fromChonkStart(o) {
    if (o.numCircuits === undefined) {
        throw new Error("Expected numCircuits in ChonkStart serialization");
    }
    ;
    return {
        num_circuits: o.numCircuits,
    };
}
export function fromChonkLoad(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkLoad serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
    };
}
export function fromChonkAccumulate(o) {
    if (o.witness === undefined) {
        throw new Error("Expected witness in ChonkAccumulate serialization");
    }
    ;
    return {
        witness: o.witness,
    };
}
export function fromChonkProve(o) {
    return {};
}
export function fromChonkProof(o) {
    if (o.hidingOinkProof === undefined) {
        throw new Error("Expected hidingOinkProof in ChonkProof serialization");
    }
    if (o.mergeProof === undefined) {
        throw new Error("Expected mergeProof in ChonkProof serialization");
    }
    if (o.eccvmProof === undefined) {
        throw new Error("Expected eccvmProof in ChonkProof serialization");
    }
    if (o.ipaProof === undefined) {
        throw new Error("Expected ipaProof in ChonkProof serialization");
    }
    if (o.jointProof === undefined) {
        throw new Error("Expected jointProof in ChonkProof serialization");
    }
    ;
    return {
        hiding_oink_proof: o.hidingOinkProof,
        merge_proof: o.mergeProof,
        eccvm_proof: o.eccvmProof,
        ipa_proof: o.ipaProof,
        joint_proof: o.jointProof,
    };
}
export function fromChonkVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkVerify serialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ChonkVerify serialization");
    }
    ;
    return {
        proof: fromChonkProof(o.proof),
        vk: o.vk,
    };
}
export function fromChonkVerifyFromFields(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkVerifyFromFields serialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ChonkVerifyFromFields serialization");
    }
    ;
    return {
        proof: o.proof,
        vk: o.vk,
    };
}
export function fromChonkBatchVerify(o) {
    if (o.proofs === undefined) {
        throw new Error("Expected proofs in ChonkBatchVerify serialization");
    }
    if (o.vks === undefined) {
        throw new Error("Expected vks in ChonkBatchVerify serialization");
    }
    ;
    return {
        proofs: o.proofs.map((v) => fromChonkProof(v)),
        vks: o.vks,
    };
}
export function fromVkAsFields(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in VkAsFields serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
    };
}
export function fromMegaVkAsFields(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in MegaVkAsFields serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
    };
}
export function fromCircuitWriteSolidityVerifier(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitWriteSolidityVerifier serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitWriteSolidityVerifier serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
        settings: fromProofSystemSettings(o.settings),
    };
}
export function fromChonkCheckPrecomputedVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkCheckPrecomputedVk serialization");
    }
    if (o.useZkFlavor === undefined) {
        throw new Error("Expected useZkFlavor in ChonkCheckPrecomputedVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
        use_zk_flavor: o.useZkFlavor,
    };
}
export function fromChonkStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ChonkStats serialization");
    }
    if (o.includeGatesPerOpcode === undefined) {
        throw new Error("Expected includeGatesPerOpcode in ChonkStats serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
        include_gates_per_opcode: o.includeGatesPerOpcode,
    };
}
export function fromChonkCompressProof(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkCompressProof serialization");
    }
    ;
    return {
        proof: fromChonkProof(o.proof),
    };
}
export function fromChonkDecompressProof(o) {
    if (o.compressedProof === undefined) {
        throw new Error("Expected compressedProof in ChonkDecompressProof serialization");
    }
    ;
    return {
        compressed_proof: o.compressedProof,
    };
}
export function fromPoseidon2Hash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Hash serialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function fromPoseidon2Permutation(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Permutation serialization");
    }
    ;
    return {
        inputs: o.inputs,
    };
}
export function fromPedersenCommit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenCommit serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenCommit serialization");
    }
    ;
    return {
        inputs: o.inputs,
        hash_index: o.hashIndex,
    };
}
export function fromPedersenHash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenHash serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenHash serialization");
    }
    ;
    return {
        inputs: o.inputs,
        hash_index: o.hashIndex,
    };
}
export function fromPedersenHashBuffer(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in PedersenHashBuffer serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenHashBuffer serialization");
    }
    ;
    return {
        input: o.input,
        hash_index: o.hashIndex,
    };
}
export function fromBlake2s(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2s serialization");
    }
    ;
    return {
        data: o.data,
    };
}
export function fromBlake2sToField(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2sToField serialization");
    }
    ;
    return {
        data: o.data,
    };
}
export function fromAesEncrypt(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesEncrypt serialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesEncrypt serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesEncrypt serialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesEncrypt serialization");
    }
    ;
    return {
        plaintext: o.plaintext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
export function fromAesDecrypt(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesDecrypt serialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesDecrypt serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesDecrypt serialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesDecrypt serialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
export function fromGrumpkinPoint(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in GrumpkinPoint serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in GrumpkinPoint serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function fromGrumpkinMul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinMul serialization");
    }
    ;
    return {
        point: fromGrumpkinPoint(o.point),
        scalar: o.scalar,
    };
}
export function fromGrumpkinAdd(o) {
    if (o.pointA === undefined) {
        throw new Error("Expected pointA in GrumpkinAdd serialization");
    }
    if (o.pointB === undefined) {
        throw new Error("Expected pointB in GrumpkinAdd serialization");
    }
    ;
    return {
        point_a: fromGrumpkinPoint(o.pointA),
        point_b: fromGrumpkinPoint(o.pointB),
    };
}
export function fromGrumpkinBatchMul(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinBatchMul serialization");
    }
    ;
    return {
        points: o.points.map((v) => fromGrumpkinPoint(v)),
        scalar: o.scalar,
    };
}
export function fromGrumpkinGetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in GrumpkinGetRandomFr serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function fromGrumpkinReduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in GrumpkinReduce512 serialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function fromSecp256k1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Secp256k1Point serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Secp256k1Point serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function fromSecp256k1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1Mul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Secp256k1Mul serialization");
    }
    ;
    return {
        point: fromSecp256k1Point(o.point),
        scalar: o.scalar,
    };
}
export function fromSecp256k1GetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in Secp256k1GetRandomFr serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function fromSecp256k1Reduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Secp256k1Reduce512 serialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function fromBn254FrSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FrSqrt serialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function fromBn254FqSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FqSqrt serialization");
    }
    ;
    return {
        input: o.input,
    };
}
export function fromBn254G1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Bn254G1Point serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Bn254G1Point serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function fromBn254G1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1Mul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Bn254G1Mul serialization");
    }
    ;
    return {
        point: fromBn254G1Point(o.point),
        scalar: o.scalar,
    };
}
export function fromBn254G2Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Bn254G2Point serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Bn254G2Point serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function fromBn254G2Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G2Mul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Bn254G2Mul serialization");
    }
    ;
    return {
        point: fromBn254G2Point(o.point),
        scalar: o.scalar,
    };
}
export function fromBn254G1IsOnCurve(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1IsOnCurve serialization");
    }
    ;
    return {
        point: fromBn254G1Point(o.point),
    };
}
export function fromBn254G1FromCompressed(o) {
    if (o.compressed === undefined) {
        throw new Error("Expected compressed in Bn254G1FromCompressed serialization");
    }
    ;
    return {
        compressed: o.compressed,
    };
}
export function fromSchnorrComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in SchnorrComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
export function fromSchnorrConstructSignature(o) {
    if (o.messageField === undefined) {
        throw new Error("Expected messageField in SchnorrConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in SchnorrConstructSignature serialization");
    }
    ;
    return {
        message_field: o.messageField,
        private_key: o.privateKey,
    };
}
export function fromSchnorrVerifySignature(o) {
    if (o.messageField === undefined) {
        throw new Error("Expected messageField in SchnorrVerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in SchnorrVerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrVerifySignature serialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrVerifySignature serialization");
    }
    ;
    return {
        message_field: o.messageField,
        public_key: fromGrumpkinPoint(o.publicKey),
        s: o.s,
        e: o.e,
    };
}
export function fromEcdsaSecp256k1ComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256k1ComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
export function fromEcdsaSecp256r1ComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256r1ComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
export function fromEcdsaSecp256k1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1ConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256k1ConstructSignature serialization");
    }
    ;
    return {
        message: o.message,
        private_key: o.privateKey,
    };
}
export function fromEcdsaSecp256r1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1ConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256r1ConstructSignature serialization");
    }
    ;
    return {
        message: o.message,
        private_key: o.privateKey,
    };
}
export function fromEcdsaSecp256k1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromEcdsaSecp256r1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromEcdsaSecp256k1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1VerifySignature serialization");
    }
    ;
    return {
        message: o.message,
        public_key: fromSecp256k1Point(o.publicKey),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromSecp256r1Point(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in Secp256r1Point serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in Secp256r1Point serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
export function fromEcdsaSecp256r1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1VerifySignature serialization");
    }
    ;
    return {
        message: o.message,
        public_key: fromSecp256r1Point(o.publicKey),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromSrsInitSrs(o) {
    if (o.pointsBuf === undefined) {
        throw new Error("Expected pointsBuf in SrsInitSrs serialization");
    }
    if (o.numPoints === undefined) {
        throw new Error("Expected numPoints in SrsInitSrs serialization");
    }
    if (o.g2Point === undefined) {
        throw new Error("Expected g2Point in SrsInitSrs serialization");
    }
    ;
    return {
        points_buf: o.pointsBuf,
        num_points: o.numPoints,
        g2_point: o.g2Point,
    };
}
export function fromChonkBatchVerifierStart(o) {
    if (o.vks === undefined) {
        throw new Error("Expected vks in ChonkBatchVerifierStart serialization");
    }
    if (o.numCores === undefined) {
        throw new Error("Expected numCores in ChonkBatchVerifierStart serialization");
    }
    if (o.batchSize === undefined) {
        throw new Error("Expected batchSize in ChonkBatchVerifierStart serialization");
    }
    if (o.fifoPath === undefined) {
        throw new Error("Expected fifoPath in ChonkBatchVerifierStart serialization");
    }
    ;
    return {
        vks: o.vks,
        num_cores: o.numCores,
        batch_size: o.batchSize,
        fifo_path: o.fifoPath,
    };
}
export function fromChonkBatchVerifierQueue(o) {
    if (o.requestId === undefined) {
        throw new Error("Expected requestId in ChonkBatchVerifierQueue serialization");
    }
    if (o.vkIndex === undefined) {
        throw new Error("Expected vkIndex in ChonkBatchVerifierQueue serialization");
    }
    if (o.proofFields === undefined) {
        throw new Error("Expected proofFields in ChonkBatchVerifierQueue serialization");
    }
    ;
    return {
        request_id: o.requestId,
        vk_index: o.vkIndex,
        proof_fields: o.proofFields,
    };
}
export function fromChonkBatchVerifierStop(o) {
    return {};
}
export function fromSrsInitGrumpkinSrs(o) {
    if (o.pointsBuf === undefined) {
        throw new Error("Expected pointsBuf in SrsInitGrumpkinSrs serialization");
    }
    if (o.numPoints === undefined) {
        throw new Error("Expected numPoints in SrsInitGrumpkinSrs serialization");
    }
    ;
    return {
        points_buf: o.pointsBuf,
        num_points: o.numPoints,
    };
}
export function fromShutdown(o) {
    return {};
}
export function fromErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in ErrorResponse serialization");
    }
    ;
    return {
        message: o.message,
    };
}
export function fromAvmProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in AvmProveResponse serialization");
    }
    if (o.stats === undefined) {
        throw new Error("Expected stats in AvmProveResponse serialization");
    }
    ;
    return {
        proof: o.proof,
        stats: o.stats.map((v) => fromAvmStat(v)),
    };
}
export function fromAvmVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in AvmVerifyResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function fromAvmCheckCircuitResponse(o) {
    if (o.passed === undefined) {
        throw new Error("Expected passed in AvmCheckCircuitResponse serialization");
    }
    if (o.stats === undefined) {
        throw new Error("Expected stats in AvmCheckCircuitResponse serialization");
    }
    ;
    return {
        passed: o.passed,
        stats: o.stats.map((v) => fromAvmStat(v)),
    };
}
export function fromCircuitProveResponse(o) {
    if (o.publicInputs === undefined) {
        throw new Error("Expected publicInputs in CircuitProveResponse serialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitProveResponse serialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in CircuitProveResponse serialization");
    }
    ;
    return {
        public_inputs: o.publicInputs,
        proof: o.proof,
        vk: fromCircuitComputeVkResponse(o.vk),
    };
}
export function fromCircuitInfoResponse(o) {
    if (o.numGates === undefined) {
        throw new Error("Expected numGates in CircuitInfoResponse serialization");
    }
    if (o.numGatesDyadic === undefined) {
        throw new Error("Expected numGatesDyadic in CircuitInfoResponse serialization");
    }
    if (o.numAcirOpcodes === undefined) {
        throw new Error("Expected numAcirOpcodes in CircuitInfoResponse serialization");
    }
    if (o.gatesPerOpcode === undefined) {
        throw new Error("Expected gatesPerOpcode in CircuitInfoResponse serialization");
    }
    ;
    return {
        num_gates: o.numGates,
        num_gates_dyadic: o.numGatesDyadic,
        num_acir_opcodes: o.numAcirOpcodes,
        gates_per_opcode: o.gatesPerOpcode,
    };
}
export function fromCircuitVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in CircuitVerifyResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function fromChonkComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ChonkComputeVkResponse serialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in ChonkComputeVkResponse serialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields,
    };
}
export function fromChonkStartResponse(o) {
    return {};
}
export function fromChonkLoadResponse(o) {
    return {};
}
export function fromChonkAccumulateResponse(o) {
    return {};
}
export function fromChonkProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkProveResponse serialization");
    }
    ;
    return {
        proof: fromChonkProof(o.proof),
    };
}
export function fromChonkVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkVerifyResponse serialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function fromChonkVerifyFromFieldsResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkVerifyFromFieldsResponse serialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function fromChonkBatchVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkBatchVerifyResponse serialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
export function fromVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in VkAsFieldsResponse serialization");
    }
    ;
    return {
        fields: o.fields,
    };
}
export function fromMegaVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in MegaVkAsFieldsResponse serialization");
    }
    ;
    return {
        fields: o.fields,
    };
}
export function fromCircuitWriteSolidityVerifierResponse(o) {
    if (o.solidityCode === undefined) {
        throw new Error("Expected solidityCode in CircuitWriteSolidityVerifierResponse serialization");
    }
    ;
    return {
        solidity_code: o.solidityCode,
    };
}
export function fromChonkCheckPrecomputedVkResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ChonkCheckPrecomputedVkResponse serialization");
    }
    if (o.actualVk === undefined) {
        throw new Error("Expected actualVk in ChonkCheckPrecomputedVkResponse serialization");
    }
    ;
    return {
        valid: o.valid,
        actual_vk: o.actualVk,
    };
}
export function fromChonkStatsResponse(o) {
    if (o.acirOpcodes === undefined) {
        throw new Error("Expected acirOpcodes in ChonkStatsResponse serialization");
    }
    if (o.circuitSize === undefined) {
        throw new Error("Expected circuitSize in ChonkStatsResponse serialization");
    }
    if (o.gatesPerOpcode === undefined) {
        throw new Error("Expected gatesPerOpcode in ChonkStatsResponse serialization");
    }
    ;
    return {
        acir_opcodes: o.acirOpcodes,
        circuit_size: o.circuitSize,
        gates_per_opcode: o.gatesPerOpcode,
    };
}
export function fromChonkCompressProofResponse(o) {
    if (o.compressedProof === undefined) {
        throw new Error("Expected compressedProof in ChonkCompressProofResponse serialization");
    }
    ;
    return {
        compressed_proof: o.compressedProof,
    };
}
export function fromChonkDecompressProofResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ChonkDecompressProofResponse serialization");
    }
    ;
    return {
        proof: fromChonkProof(o.proof),
    };
}
export function fromPoseidon2HashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function fromPoseidon2PermutationResponse(o) {
    if (o.outputs === undefined) {
        throw new Error("Expected outputs in Poseidon2PermutationResponse serialization");
    }
    ;
    return {
        outputs: o.outputs,
    };
}
export function fromPedersenCommitResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in PedersenCommitResponse serialization");
    }
    ;
    return {
        point: fromGrumpkinPoint(o.point),
    };
}
export function fromPedersenHashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function fromPedersenHashBufferResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashBufferResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function fromBlake2sResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Blake2sResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
export function fromBlake2sToFieldResponse(o) {
    if (o.field === undefined) {
        throw new Error("Expected field in Blake2sToFieldResponse serialization");
    }
    ;
    return {
        field: o.field,
    };
}
export function fromAesEncryptResponse(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesEncryptResponse serialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
    };
}
export function fromAesDecryptResponse(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesDecryptResponse serialization");
    }
    ;
    return {
        plaintext: o.plaintext,
    };
}
export function fromGrumpkinMulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMulResponse serialization");
    }
    ;
    return {
        point: fromGrumpkinPoint(o.point),
    };
}
export function fromGrumpkinAddResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinAddResponse serialization");
    }
    ;
    return {
        point: fromGrumpkinPoint(o.point),
    };
}
export function fromGrumpkinBatchMulResponse(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMulResponse serialization");
    }
    ;
    return {
        points: o.points.map((v) => fromGrumpkinPoint(v)),
    };
}
export function fromGrumpkinGetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinGetRandomFrResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function fromGrumpkinReduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinReduce512Response serialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function fromSecp256k1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1MulResponse serialization");
    }
    ;
    return {
        point: fromSecp256k1Point(o.point),
    };
}
export function fromSecp256k1GetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1GetRandomFrResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function fromSecp256k1Reduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1Reduce512Response serialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function fromBn254FrSqrtResponse(o) {
    if (o.isSquareRoot === undefined) {
        throw new Error("Expected isSquareRoot in Bn254FrSqrtResponse serialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FrSqrtResponse serialization");
    }
    ;
    return {
        is_square_root: o.isSquareRoot,
        value: o.value,
    };
}
export function fromBn254FqSqrtResponse(o) {
    if (o.isSquareRoot === undefined) {
        throw new Error("Expected isSquareRoot in Bn254FqSqrtResponse serialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FqSqrtResponse serialization");
    }
    ;
    return {
        is_square_root: o.isSquareRoot,
        value: o.value,
    };
}
export function fromBn254G1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1MulResponse serialization");
    }
    ;
    return {
        point: fromBn254G1Point(o.point),
    };
}
export function fromBn254G2MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G2MulResponse serialization");
    }
    ;
    return {
        point: fromBn254G2Point(o.point),
    };
}
export function fromBn254G1IsOnCurveResponse(o) {
    if (o.isOnCurve === undefined) {
        throw new Error("Expected isOnCurve in Bn254G1IsOnCurveResponse serialization");
    }
    ;
    return {
        is_on_curve: o.isOnCurve,
    };
}
export function fromBn254G1FromCompressedResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Bn254G1FromCompressedResponse serialization");
    }
    ;
    return {
        point: fromBn254G1Point(o.point),
    };
}
export function fromSchnorrComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in SchnorrComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromGrumpkinPoint(o.publicKey),
    };
}
export function fromSchnorrConstructSignatureResponse(o) {
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrConstructSignatureResponse serialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrConstructSignatureResponse serialization");
    }
    ;
    return {
        s: o.s,
        e: o.e,
    };
}
export function fromSchnorrVerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in SchnorrVerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function fromEcdsaSecp256k1ComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1ComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromSecp256k1Point(o.publicKey),
    };
}
export function fromEcdsaSecp256r1ComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1ComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromSecp256r1Point(o.publicKey),
    };
}
export function fromEcdsaSecp256k1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromEcdsaSecp256r1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
export function fromEcdsaSecp256k1RecoverPublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1RecoverPublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromSecp256k1Point(o.publicKey),
    };
}
export function fromEcdsaSecp256r1RecoverPublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1RecoverPublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromSecp256r1Point(o.publicKey),
    };
}
export function fromEcdsaSecp256k1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function fromEcdsaSecp256r1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
export function fromSrsInitSrsResponse(o) {
    if (o.pointsBuf === undefined) {
        throw new Error("Expected pointsBuf in SrsInitSrsResponse serialization");
    }
    ;
    return {
        points_buf: o.pointsBuf,
    };
}
export function fromChonkBatchVerifierStartResponse(o) {
    return {};
}
export function fromChonkBatchVerifierQueueResponse(o) {
    return {};
}
export function fromChonkBatchVerifierStopResponse(o) {
    return {};
}
export function fromSrsInitGrumpkinSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
export function fromShutdownResponse(o) {
    return {};
}
