// @aztec/bb.js/dest/node-cjs/bbapi/chonk_pinned_inputs.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const msgpackr_1 = require("msgpackr");
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const pako_1 = require("pako");
const index_js_1 = require("../index.js");
const TEST_TIMEOUT_MS = 30 * 60 * 1000;
const DEFAULT_WASM_FLOW_LIMIT = 1;
function findRepoRoot() {
    return process.env.AZTEC_REPO_ROOT ?? (0, node_path_1.resolve)(process.cwd(), '../..');
}
function ensurePinnedInputsRoot() {
    const explicit = process.env.CHONK_PINNED_IVC_INPUTS_DIR;
    if (explicit) {
        return explicit;
    }
    const repoRoot = findRepoRoot();
    return (0, node_path_1.join)(repoRoot, 'barretenberg/cpp/chonk-pinned-flows');
}
function discoverFlows(root) {
    if (!(0, node_fs_1.existsSync)(root)) {
        return [];
    }
    return (0, node_fs_1.readdirSync)(root)
        .map(name => (0, node_path_1.join)(root, name))
        .filter(path => (0, node_fs_1.statSync)(path).isDirectory() && (0, node_fs_1.existsSync)((0, node_path_1.join)(path, 'ivc-inputs.msgpack')))
        .sort();
}
function selectFlows(flows) {
    const filter = process.env.CHONK_PINNED_IVC_FLOW;
    const filtered = filter ? flows.filter(flow => (0, node_path_1.basename)(flow).includes(filter)) : flows;
    const limit = Number(process.env.CHONK_PINNED_IVC_FLOW_LIMIT);
    return Number.isInteger(limit) && limit > 0 ? filtered.slice(0, limit) : filtered;
}
function loadPinnedFlow(flowDir) {
    const buf = (0, node_fs_1.readFileSync)((0, node_path_1.join)(flowDir, 'ivc-inputs.msgpack'));
    const steps = new msgpackr_1.Decoder({ useRecords: false }).unpack(buf);
    if (steps.length === 0) {
        throw new Error(`No execution steps in ${(0, node_path_1.join)(flowDir, 'ivc-inputs.msgpack')}`);
    }
    return {
        bytecodes: steps.map(step => (0, pako_1.ungzip)(step.bytecode)),
        witnesses: steps.map(step => (0, pako_1.ungzip)(step.witness)),
        vks: steps.map(step => new Uint8Array(step.vk)),
        names: steps.map(step => step.functionName),
    };
}
function getWasmFlowLimit() {
    const parsed = Number(process.env.CHONK_PINNED_IVC_WASM_FLOW_LIMIT ?? DEFAULT_WASM_FLOW_LIMIT);
    return Number.isInteger(parsed) && parsed > 0 ? parsed : DEFAULT_WASM_FLOW_LIMIT;
}
const wasmFlowLimit = getWasmFlowLimit();
const backendCases = [
    {
        label: 'native',
        backendType: index_js_1.BackendType.NativeUnixSocket,
        threads: 16,
        selectFlows: (flows) => flows,
    },
    {
        label: 'wasm',
        backendType: index_js_1.BackendType.Wasm,
        threads: 1,
        selectFlows: (flows) => flows.slice(0, Math.max(1, wasmFlowLimit)),
    },
];
describe('Chonk pinned IVC inputs through bb.js', () => {
    let flows;
    let bbPath;
    beforeAll(() => {
        const repoRoot = findRepoRoot();
        const defaultBbPath = (0, node_path_1.join)(repoRoot, 'barretenberg/cpp/build/bin/bb');
        bbPath = process.env.BB_BINARY_PATH ?? ((0, node_fs_1.existsSync)(defaultBbPath) ? defaultBbPath : undefined);
        const pinnedRoot = ensurePinnedInputsRoot();
        const discoveredFlows = discoverFlows(pinnedRoot);
        if (discoveredFlows.length === 0) {
            throw new Error(`No pinned ivc-inputs.msgpack files found under ${pinnedRoot}. Run barretenberg/cpp/scripts/chonk_inputs.sh download first.`);
        }
        flows = selectFlows(discoveredFlows);
        if (flows.length === 0 && process.env.CHONK_PINNED_IVC_FLOW) {
            throw new Error(`CHONK_PINNED_IVC_FLOW='${process.env.CHONK_PINNED_IVC_FLOW}' matched no pinned flows`);
        }
        if (flows.length === 0) {
            throw new Error(`No pinned ivc-inputs.msgpack files found under ${pinnedRoot}`);
        }
    }, TEST_TIMEOUT_MS);
    it.each(backendCases)('proves and verifies pinned flows with $label backend', async backendCase => {
        const barretenberg = await index_js_1.Barretenberg.initSingleton({
            backend: backendCase.backendType,
            threads: backendCase.threads,
            ...(backendCase.backendType === index_js_1.BackendType.NativeUnixSocket && bbPath ? { bbPath } : {}),
        });
        try {
            for (const flowDir of backendCase.selectFlows(flows)) {
                const { bytecodes, witnesses, vks, names } = loadPinnedFlow(flowDir);
                const backend = new index_js_1.AztecClientBackend(bytecodes, barretenberg, names);
                const { proof, vk } = await backend.prove(witnesses, vks);
                const verified = await backend.verify(proof, vk);
                expect(verified).toBe(true);
            }
        }
        finally {
            await index_js_1.Barretenberg.destroySingleton();
        }
    }, TEST_TIMEOUT_MS);
});
