// @aztec/bb.js/dest/node-cjs/aztec-wsdb/generated/api_types.js
"use strict";
// AUTOGENERATED FILE - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSiblingPathAndIndex = toSiblingPathAndIndex;
exports.toWorldStateStatusSummary = toWorldStateStatusSummary;
exports.toDBStats = toDBStats;
exports.toTreeDBStats = toTreeDBStats;
exports.toWorldStateDBStats = toWorldStateDBStats;
exports.toTreeMeta = toTreeMeta;
exports.toWorldStateMeta = toWorldStateMeta;
exports.toWorldStateStatusFull = toWorldStateStatusFull;
exports.toWorldStateRevision = toWorldStateRevision;
exports.toWsdbGetTreeInfo = toWsdbGetTreeInfo;
exports.toWsdbGetStateReference = toWsdbGetStateReference;
exports.toWsdbGetInitialStateReference = toWsdbGetInitialStateReference;
exports.toWsdbGetLeafValue = toWsdbGetLeafValue;
exports.toWsdbGetLeafPreimage = toWsdbGetLeafPreimage;
exports.toWsdbGetSiblingPath = toWsdbGetSiblingPath;
exports.toWsdbGetBlockNumbersForLeafIndices = toWsdbGetBlockNumbersForLeafIndices;
exports.toWsdbFindLeafIndices = toWsdbFindLeafIndices;
exports.toWsdbFindLowLeaf = toWsdbFindLowLeaf;
exports.toWsdbFindSiblingPaths = toWsdbFindSiblingPaths;
exports.toWsdbAppendLeaves = toWsdbAppendLeaves;
exports.toWsdbBatchInsert = toWsdbBatchInsert;
exports.toWsdbSequentialInsert = toWsdbSequentialInsert;
exports.toWsdbUpdateArchive = toWsdbUpdateArchive;
exports.toWsdbCommit = toWsdbCommit;
exports.toWsdbRollback = toWsdbRollback;
exports.toNullifierLeafValue = toNullifierLeafValue;
exports.toPublicDataLeafValue = toPublicDataLeafValue;
exports.toWsdbSyncBlock = toWsdbSyncBlock;
exports.toWsdbCreateFork = toWsdbCreateFork;
exports.toWsdbDeleteFork = toWsdbDeleteFork;
exports.toWsdbFinalizeBlocks = toWsdbFinalizeBlocks;
exports.toWsdbUnwindBlocks = toWsdbUnwindBlocks;
exports.toWsdbRemoveHistoricalBlocks = toWsdbRemoveHistoricalBlocks;
exports.toWsdbGetStatus = toWsdbGetStatus;
exports.toWsdbCreateCheckpoint = toWsdbCreateCheckpoint;
exports.toWsdbCommitCheckpoint = toWsdbCommitCheckpoint;
exports.toWsdbRevertCheckpoint = toWsdbRevertCheckpoint;
exports.toWsdbCommitAllCheckpoints = toWsdbCommitAllCheckpoints;
exports.toWsdbRevertAllCheckpoints = toWsdbRevertAllCheckpoints;
exports.toWsdbCopyStores = toWsdbCopyStores;
exports.toWsdbShutdown = toWsdbShutdown;
exports.toWsdbErrorResponse = toWsdbErrorResponse;
exports.toWsdbGetTreeInfoResponse = toWsdbGetTreeInfoResponse;
exports.toWsdbGetStateReferenceResponse = toWsdbGetStateReferenceResponse;
exports.toWsdbGetInitialStateReferenceResponse = toWsdbGetInitialStateReferenceResponse;
exports.toWsdbGetLeafValueResponse = toWsdbGetLeafValueResponse;
exports.toWsdbGetLeafPreimageResponse = toWsdbGetLeafPreimageResponse;
exports.toWsdbGetSiblingPathResponse = toWsdbGetSiblingPathResponse;
exports.toWsdbGetBlockNumbersForLeafIndicesResponse = toWsdbGetBlockNumbersForLeafIndicesResponse;
exports.toWsdbFindLeafIndicesResponse = toWsdbFindLeafIndicesResponse;
exports.toWsdbFindLowLeafResponse = toWsdbFindLowLeafResponse;
exports.toWsdbFindSiblingPathsResponse = toWsdbFindSiblingPathsResponse;
exports.toWsdbAppendLeavesResponse = toWsdbAppendLeavesResponse;
exports.toWsdbBatchInsertResponse = toWsdbBatchInsertResponse;
exports.toWsdbSequentialInsertResponse = toWsdbSequentialInsertResponse;
exports.toWsdbUpdateArchiveResponse = toWsdbUpdateArchiveResponse;
exports.toWsdbCommitResponse = toWsdbCommitResponse;
exports.toWsdbRollbackResponse = toWsdbRollbackResponse;
exports.toWsdbSyncBlockResponse = toWsdbSyncBlockResponse;
exports.toWsdbCreateForkResponse = toWsdbCreateForkResponse;
exports.toWsdbDeleteForkResponse = toWsdbDeleteForkResponse;
exports.toWsdbFinalizeBlocksResponse = toWsdbFinalizeBlocksResponse;
exports.toWsdbUnwindBlocksResponse = toWsdbUnwindBlocksResponse;
exports.toWsdbRemoveHistoricalBlocksResponse = toWsdbRemoveHistoricalBlocksResponse;
exports.toWsdbGetStatusResponse = toWsdbGetStatusResponse;
exports.toWsdbCreateCheckpointResponse = toWsdbCreateCheckpointResponse;
exports.toWsdbCommitCheckpointResponse = toWsdbCommitCheckpointResponse;
exports.toWsdbRevertCheckpointResponse = toWsdbRevertCheckpointResponse;
exports.toWsdbCommitAllCheckpointsResponse = toWsdbCommitAllCheckpointsResponse;
exports.toWsdbRevertAllCheckpointsResponse = toWsdbRevertAllCheckpointsResponse;
exports.toWsdbCopyStoresResponse = toWsdbCopyStoresResponse;
exports.toWsdbShutdownResponse = toWsdbShutdownResponse;
exports.fromSiblingPathAndIndex = fromSiblingPathAndIndex;
exports.fromWorldStateStatusSummary = fromWorldStateStatusSummary;
exports.fromDBStats = fromDBStats;
exports.fromTreeDBStats = fromTreeDBStats;
exports.fromWorldStateDBStats = fromWorldStateDBStats;
exports.fromTreeMeta = fromTreeMeta;
exports.fromWorldStateMeta = fromWorldStateMeta;
exports.fromWorldStateStatusFull = fromWorldStateStatusFull;
exports.fromWorldStateRevision = fromWorldStateRevision;
exports.fromWsdbGetTreeInfo = fromWsdbGetTreeInfo;
exports.fromWsdbGetStateReference = fromWsdbGetStateReference;
exports.fromWsdbGetInitialStateReference = fromWsdbGetInitialStateReference;
exports.fromWsdbGetLeafValue = fromWsdbGetLeafValue;
exports.fromWsdbGetLeafPreimage = fromWsdbGetLeafPreimage;
exports.fromWsdbGetSiblingPath = fromWsdbGetSiblingPath;
exports.fromWsdbGetBlockNumbersForLeafIndices = fromWsdbGetBlockNumbersForLeafIndices;
exports.fromWsdbFindLeafIndices = fromWsdbFindLeafIndices;
exports.fromWsdbFindLowLeaf = fromWsdbFindLowLeaf;
exports.fromWsdbFindSiblingPaths = fromWsdbFindSiblingPaths;
exports.fromWsdbAppendLeaves = fromWsdbAppendLeaves;
exports.fromWsdbBatchInsert = fromWsdbBatchInsert;
exports.fromWsdbSequentialInsert = fromWsdbSequentialInsert;
exports.fromWsdbUpdateArchive = fromWsdbUpdateArchive;
exports.fromWsdbCommit = fromWsdbCommit;
exports.fromWsdbRollback = fromWsdbRollback;
exports.fromNullifierLeafValue = fromNullifierLeafValue;
exports.fromPublicDataLeafValue = fromPublicDataLeafValue;
exports.fromWsdbSyncBlock = fromWsdbSyncBlock;
exports.fromWsdbCreateFork = fromWsdbCreateFork;
exports.fromWsdbDeleteFork = fromWsdbDeleteFork;
exports.fromWsdbFinalizeBlocks = fromWsdbFinalizeBlocks;
exports.fromWsdbUnwindBlocks = fromWsdbUnwindBlocks;
exports.fromWsdbRemoveHistoricalBlocks = fromWsdbRemoveHistoricalBlocks;
exports.fromWsdbGetStatus = fromWsdbGetStatus;
exports.fromWsdbCreateCheckpoint = fromWsdbCreateCheckpoint;
exports.fromWsdbCommitCheckpoint = fromWsdbCommitCheckpoint;
exports.fromWsdbRevertCheckpoint = fromWsdbRevertCheckpoint;
exports.fromWsdbCommitAllCheckpoints = fromWsdbCommitAllCheckpoints;
exports.fromWsdbRevertAllCheckpoints = fromWsdbRevertAllCheckpoints;
exports.fromWsdbCopyStores = fromWsdbCopyStores;
exports.fromWsdbShutdown = fromWsdbShutdown;
exports.fromWsdbErrorResponse = fromWsdbErrorResponse;
exports.fromWsdbGetTreeInfoResponse = fromWsdbGetTreeInfoResponse;
exports.fromWsdbGetStateReferenceResponse = fromWsdbGetStateReferenceResponse;
exports.fromWsdbGetInitialStateReferenceResponse = fromWsdbGetInitialStateReferenceResponse;
exports.fromWsdbGetLeafValueResponse = fromWsdbGetLeafValueResponse;
exports.fromWsdbGetLeafPreimageResponse = fromWsdbGetLeafPreimageResponse;
exports.fromWsdbGetSiblingPathResponse = fromWsdbGetSiblingPathResponse;
exports.fromWsdbGetBlockNumbersForLeafIndicesResponse = fromWsdbGetBlockNumbersForLeafIndicesResponse;
exports.fromWsdbFindLeafIndicesResponse = fromWsdbFindLeafIndicesResponse;
exports.fromWsdbFindLowLeafResponse = fromWsdbFindLowLeafResponse;
exports.fromWsdbFindSiblingPathsResponse = fromWsdbFindSiblingPathsResponse;
exports.fromWsdbAppendLeavesResponse = fromWsdbAppendLeavesResponse;
exports.fromWsdbBatchInsertResponse = fromWsdbBatchInsertResponse;
exports.fromWsdbSequentialInsertResponse = fromWsdbSequentialInsertResponse;
exports.fromWsdbUpdateArchiveResponse = fromWsdbUpdateArchiveResponse;
exports.fromWsdbCommitResponse = fromWsdbCommitResponse;
exports.fromWsdbRollbackResponse = fromWsdbRollbackResponse;
exports.fromWsdbSyncBlockResponse = fromWsdbSyncBlockResponse;
exports.fromWsdbCreateForkResponse = fromWsdbCreateForkResponse;
exports.fromWsdbDeleteForkResponse = fromWsdbDeleteForkResponse;
exports.fromWsdbFinalizeBlocksResponse = fromWsdbFinalizeBlocksResponse;
exports.fromWsdbUnwindBlocksResponse = fromWsdbUnwindBlocksResponse;
exports.fromWsdbRemoveHistoricalBlocksResponse = fromWsdbRemoveHistoricalBlocksResponse;
exports.fromWsdbGetStatusResponse = fromWsdbGetStatusResponse;
exports.fromWsdbCreateCheckpointResponse = fromWsdbCreateCheckpointResponse;
exports.fromWsdbCommitCheckpointResponse = fromWsdbCommitCheckpointResponse;
exports.fromWsdbRevertCheckpointResponse = fromWsdbRevertCheckpointResponse;
exports.fromWsdbCommitAllCheckpointsResponse = fromWsdbCommitAllCheckpointsResponse;
exports.fromWsdbRevertAllCheckpointsResponse = fromWsdbRevertAllCheckpointsResponse;
exports.fromWsdbCopyStoresResponse = fromWsdbCopyStoresResponse;
exports.fromWsdbShutdownResponse = fromWsdbShutdownResponse;
// Conversion functions (exported)
function toSiblingPathAndIndex(o) {
    if (o.index === undefined) {
        throw new Error("Expected index in SiblingPathAndIndex deserialization");
    }
    if (o.path === undefined) {
        throw new Error("Expected path in SiblingPathAndIndex deserialization");
    }
    ;
    return {
        index: o.index,
        path: o.path,
    };
}
function toWorldStateStatusSummary(o) {
    if (o.unfinalizedBlockNumber === undefined) {
        throw new Error("Expected unfinalizedBlockNumber in WorldStateStatusSummary deserialization");
    }
    if (o.finalizedBlockNumber === undefined) {
        throw new Error("Expected finalizedBlockNumber in WorldStateStatusSummary deserialization");
    }
    if (o.oldestHistoricalBlock === undefined) {
        throw new Error("Expected oldestHistoricalBlock in WorldStateStatusSummary deserialization");
    }
    if (o.treesAreSynched === undefined) {
        throw new Error("Expected treesAreSynched in WorldStateStatusSummary deserialization");
    }
    ;
    return {
        unfinalizedblocknumber: o.unfinalizedBlockNumber,
        finalizedblocknumber: o.finalizedBlockNumber,
        oldesthistoricalblock: o.oldestHistoricalBlock,
        treesaresynched: o.treesAreSynched,
    };
}
function toDBStats(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in DBStats deserialization");
    }
    if (o.numDataItems === undefined) {
        throw new Error("Expected numDataItems in DBStats deserialization");
    }
    if (o.totalUsedSize === undefined) {
        throw new Error("Expected totalUsedSize in DBStats deserialization");
    }
    ;
    return {
        name: o.name,
        numdataitems: o.numDataItems,
        totalusedsize: o.totalUsedSize,
    };
}
function toTreeDBStats(o) {
    if (o.mapSize === undefined) {
        throw new Error("Expected mapSize in TreeDBStats deserialization");
    }
    if (o.physicalFileSize === undefined) {
        throw new Error("Expected physicalFileSize in TreeDBStats deserialization");
    }
    if (o.blocksDBStats === undefined) {
        throw new Error("Expected blocksDBStats in TreeDBStats deserialization");
    }
    if (o.nodesDBStats === undefined) {
        throw new Error("Expected nodesDBStats in TreeDBStats deserialization");
    }
    if (o.leafPreimagesDBStats === undefined) {
        throw new Error("Expected leafPreimagesDBStats in TreeDBStats deserialization");
    }
    if (o.leafIndicesDBStats === undefined) {
        throw new Error("Expected leafIndicesDBStats in TreeDBStats deserialization");
    }
    if (o.blockIndicesDBStats === undefined) {
        throw new Error("Expected blockIndicesDBStats in TreeDBStats deserialization");
    }
    ;
    return {
        mapsize: o.mapSize,
        physicalfilesize: o.physicalFileSize,
        blocksdbstats: toDBStats(o.blocksDBStats),
        nodesdbstats: toDBStats(o.nodesDBStats),
        leafpreimagesdbstats: toDBStats(o.leafPreimagesDBStats),
        leafindicesdbstats: toDBStats(o.leafIndicesDBStats),
        blockindicesdbstats: toDBStats(o.blockIndicesDBStats),
    };
}
function toWorldStateDBStats(o) {
    if (o.noteHashTreeStats === undefined) {
        throw new Error("Expected noteHashTreeStats in WorldStateDBStats deserialization");
    }
    if (o.messageTreeStats === undefined) {
        throw new Error("Expected messageTreeStats in WorldStateDBStats deserialization");
    }
    if (o.archiveTreeStats === undefined) {
        throw new Error("Expected archiveTreeStats in WorldStateDBStats deserialization");
    }
    if (o.publicDataTreeStats === undefined) {
        throw new Error("Expected publicDataTreeStats in WorldStateDBStats deserialization");
    }
    if (o.nullifierTreeStats === undefined) {
        throw new Error("Expected nullifierTreeStats in WorldStateDBStats deserialization");
    }
    ;
    return {
        notehashtreestats: toTreeDBStats(o.noteHashTreeStats),
        messagetreestats: toTreeDBStats(o.messageTreeStats),
        archivetreestats: toTreeDBStats(o.archiveTreeStats),
        publicdatatreestats: toTreeDBStats(o.publicDataTreeStats),
        nullifiertreestats: toTreeDBStats(o.nullifierTreeStats),
    };
}
function toTreeMeta(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in TreeMeta deserialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in TreeMeta deserialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in TreeMeta deserialization");
    }
    if (o.committedSize === undefined) {
        throw new Error("Expected committedSize in TreeMeta deserialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in TreeMeta deserialization");
    }
    if (o.initialSize === undefined) {
        throw new Error("Expected initialSize in TreeMeta deserialization");
    }
    if (o.initialRoot === undefined) {
        throw new Error("Expected initialRoot in TreeMeta deserialization");
    }
    if (o.oldestHistoricBlock === undefined) {
        throw new Error("Expected oldestHistoricBlock in TreeMeta deserialization");
    }
    if (o.unfinalizedBlockHeight === undefined) {
        throw new Error("Expected unfinalizedBlockHeight in TreeMeta deserialization");
    }
    if (o.finalizedBlockHeight === undefined) {
        throw new Error("Expected finalizedBlockHeight in TreeMeta deserialization");
    }
    ;
    return {
        name: o.name,
        depth: o.depth,
        size: o.size,
        committedsize: o.committedSize,
        root: o.root,
        initialsize: o.initialSize,
        initialroot: o.initialRoot,
        oldesthistoricblock: o.oldestHistoricBlock,
        unfinalizedblockheight: o.unfinalizedBlockHeight,
        finalizedblockheight: o.finalizedBlockHeight,
    };
}
function toWorldStateMeta(o) {
    if (o.noteHashTreeMeta === undefined) {
        throw new Error("Expected noteHashTreeMeta in WorldStateMeta deserialization");
    }
    if (o.messageTreeMeta === undefined) {
        throw new Error("Expected messageTreeMeta in WorldStateMeta deserialization");
    }
    if (o.archiveTreeMeta === undefined) {
        throw new Error("Expected archiveTreeMeta in WorldStateMeta deserialization");
    }
    if (o.publicDataTreeMeta === undefined) {
        throw new Error("Expected publicDataTreeMeta in WorldStateMeta deserialization");
    }
    if (o.nullifierTreeMeta === undefined) {
        throw new Error("Expected nullifierTreeMeta in WorldStateMeta deserialization");
    }
    ;
    return {
        notehashtreemeta: toTreeMeta(o.noteHashTreeMeta),
        messagetreemeta: toTreeMeta(o.messageTreeMeta),
        archivetreemeta: toTreeMeta(o.archiveTreeMeta),
        publicdatatreemeta: toTreeMeta(o.publicDataTreeMeta),
        nullifiertreemeta: toTreeMeta(o.nullifierTreeMeta),
    };
}
function toWorldStateStatusFull(o) {
    if (o.summary === undefined) {
        throw new Error("Expected summary in WorldStateStatusFull deserialization");
    }
    if (o.dbStats === undefined) {
        throw new Error("Expected dbStats in WorldStateStatusFull deserialization");
    }
    if (o.meta === undefined) {
        throw new Error("Expected meta in WorldStateStatusFull deserialization");
    }
    ;
    return {
        summary: toWorldStateStatusSummary(o.summary),
        dbstats: toWorldStateDBStats(o.dbStats),
        meta: toWorldStateMeta(o.meta),
    };
}
function toWorldStateRevision(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WorldStateRevision deserialization");
    }
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WorldStateRevision deserialization");
    }
    if (o.includeUncommitted === undefined) {
        throw new Error("Expected includeUncommitted in WorldStateRevision deserialization");
    }
    ;
    return {
        forkid: o.forkId,
        blocknumber: o.blockNumber,
        includeuncommitted: o.includeUncommitted,
    };
}
function toWsdbGetTreeInfo(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetTreeInfo deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetTreeInfo deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
    };
}
function toWsdbGetStateReference(o) {
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetStateReference deserialization");
    }
    ;
    return {
        revision: toWorldStateRevision(o.revision),
    };
}
function toWsdbGetInitialStateReference(o) {
    return {};
}
function toWsdbGetLeafValue(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetLeafValue deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafValue deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetLeafValue deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
function toWsdbGetLeafPreimage(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetLeafPreimage deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafPreimage deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetLeafPreimage deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
function toWsdbGetSiblingPath(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetSiblingPath deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetSiblingPath deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetSiblingPath deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
function toWsdbGetBlockNumbersForLeafIndices(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    if (o.leafIndices === undefined) {
        throw new Error("Expected leafIndices in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindices: o.leafIndices,
    };
}
function toWsdbFindLeafIndices(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindLeafIndices deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLeafIndices deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindLeafIndices deserialization");
    }
    if (o.startIndex === undefined) {
        throw new Error("Expected startIndex in WsdbFindLeafIndices deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leaves: o.leaves,
        startindex: o.startIndex,
    };
}
function toWsdbFindLowLeaf(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindLowLeaf deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLowLeaf deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in WsdbFindLowLeaf deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        key: o.key,
    };
}
function toWsdbFindSiblingPaths(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindSiblingPaths deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindSiblingPaths deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindSiblingPaths deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leaves: o.leaves,
    };
}
function toWsdbAppendLeaves(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbAppendLeaves deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbAppendLeaves deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbAppendLeaves deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        forkid: o.forkId,
    };
}
function toWsdbBatchInsert(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbBatchInsert deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbBatchInsert deserialization");
    }
    if (o.subtreeDepth === undefined) {
        throw new Error("Expected subtreeDepth in WsdbBatchInsert deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbBatchInsert deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        subtreedepth: o.subtreeDepth,
        forkid: o.forkId,
    };
}
function toWsdbSequentialInsert(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbSequentialInsert deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbSequentialInsert deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbSequentialInsert deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        forkid: o.forkId,
    };
}
function toWsdbUpdateArchive(o) {
    if (o.blockStateRef === undefined) {
        throw new Error("Expected blockStateRef in WsdbUpdateArchive deserialization");
    }
    if (o.blockHeaderHash === undefined) {
        throw new Error("Expected blockHeaderHash in WsdbUpdateArchive deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbUpdateArchive deserialization");
    }
    ;
    return {
        blockstateref: o.blockStateRef,
        blockheaderhash: o.blockHeaderHash,
        forkid: o.forkId,
    };
}
function toWsdbCommit(o) {
    return {};
}
function toWsdbRollback(o) {
    return {};
}
function toNullifierLeafValue(o) {
    if (o.nullifier === undefined) {
        throw new Error("Expected nullifier in NullifierLeafValue deserialization");
    }
    ;
    return {
        nullifier: o.nullifier,
    };
}
function toPublicDataLeafValue(o) {
    if (o.slot === undefined) {
        throw new Error("Expected slot in PublicDataLeafValue deserialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in PublicDataLeafValue deserialization");
    }
    ;
    return {
        slot: o.slot,
        value: o.value,
    };
}
function toWsdbSyncBlock(o) {
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WsdbSyncBlock deserialization");
    }
    if (o.blockStateRef === undefined) {
        throw new Error("Expected blockStateRef in WsdbSyncBlock deserialization");
    }
    if (o.blockHeaderHash === undefined) {
        throw new Error("Expected blockHeaderHash in WsdbSyncBlock deserialization");
    }
    if (o.paddedNoteHashes === undefined) {
        throw new Error("Expected paddedNoteHashes in WsdbSyncBlock deserialization");
    }
    if (o.paddedL1ToL2Messages === undefined) {
        throw new Error("Expected paddedL1ToL2Messages in WsdbSyncBlock deserialization");
    }
    if (o.paddedNullifiers === undefined) {
        throw new Error("Expected paddedNullifiers in WsdbSyncBlock deserialization");
    }
    if (o.publicDataWrites === undefined) {
        throw new Error("Expected publicDataWrites in WsdbSyncBlock deserialization");
    }
    ;
    return {
        blocknumber: o.blockNumber,
        blockstateref: o.blockStateRef,
        blockheaderhash: o.blockHeaderHash,
        paddednotehashes: o.paddedNoteHashes,
        paddedl1tol2messages: o.paddedL1ToL2Messages,
        paddednullifiers: o.paddedNullifiers.map((v) => toNullifierLeafValue(v)),
        publicdatawrites: o.publicDataWrites.map((v) => toPublicDataLeafValue(v)),
    };
}
function toWsdbCreateFork(o) {
    if (o.latest === undefined) {
        throw new Error("Expected latest in WsdbCreateFork deserialization");
    }
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WsdbCreateFork deserialization");
    }
    ;
    return {
        latest: o.latest,
        blocknumber: o.blockNumber,
    };
}
function toWsdbDeleteFork(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbDeleteFork deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbFinalizeBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbFinalizeBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
function toWsdbUnwindBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbUnwindBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
function toWsdbRemoveHistoricalBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbRemoveHistoricalBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
function toWsdbGetStatus(o) {
    return {};
}
function toWsdbCreateCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCreateCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbCommitCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCommitCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbRevertCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbRevertCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbCommitAllCheckpoints(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCommitAllCheckpoints deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbRevertAllCheckpoints(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbRevertAllCheckpoints deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbCopyStores(o) {
    if (o.dstPath === undefined) {
        throw new Error("Expected dstPath in WsdbCopyStores deserialization");
    }
    if (o.compact === undefined) {
        throw new Error("Expected compact in WsdbCopyStores deserialization");
    }
    ;
    return {
        dstpath: o.dstPath,
        compact: o.compact,
    };
}
function toWsdbShutdown(o) {
    return {};
}
function toWsdbErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in WsdbErrorResponse deserialization");
    }
    ;
    return {
        message: o.message,
    };
}
function toWsdbGetTreeInfoResponse(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in WsdbGetTreeInfoResponse deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        root: o.root,
        size: o.size,
        depth: o.depth,
    };
}
function toWsdbGetStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetStateReferenceResponse deserialization");
    }
    ;
    return {
        state: o.state,
    };
}
function toWsdbGetInitialStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetInitialStateReferenceResponse deserialization");
    }
    ;
    return {
        state: o.state,
    };
}
function toWsdbGetLeafValueResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in WsdbGetLeafValueResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
function toWsdbGetLeafPreimageResponse(o) {
    if (o.preimage === undefined) {
        throw new Error("Expected preimage in WsdbGetLeafPreimageResponse deserialization");
    }
    ;
    return {
        preimage: o.preimage,
    };
}
function toWsdbGetSiblingPathResponse(o) {
    if (o.path === undefined) {
        throw new Error("Expected path in WsdbGetSiblingPathResponse deserialization");
    }
    ;
    return {
        path: o.path,
    };
}
function toWsdbGetBlockNumbersForLeafIndicesResponse(o) {
    if (o.blockNumbers === undefined) {
        throw new Error("Expected blockNumbers in WsdbGetBlockNumbersForLeafIndicesResponse deserialization");
    }
    ;
    return {
        blocknumbers: o.blockNumbers,
    };
}
function toWsdbFindLeafIndicesResponse(o) {
    if (o.indices === undefined) {
        throw new Error("Expected indices in WsdbFindLeafIndicesResponse deserialization");
    }
    ;
    return {
        indices: o.indices,
    };
}
function toWsdbFindLowLeafResponse(o) {
    if (o.alreadyPresent === undefined) {
        throw new Error("Expected alreadyPresent in WsdbFindLowLeafResponse deserialization");
    }
    if (o.index === undefined) {
        throw new Error("Expected index in WsdbFindLowLeafResponse deserialization");
    }
    ;
    return {
        alreadypresent: o.alreadyPresent,
        index: o.index,
    };
}
function toWsdbFindSiblingPathsResponse(o) {
    if (o.paths === undefined) {
        throw new Error("Expected paths in WsdbFindSiblingPathsResponse deserialization");
    }
    ;
    return {
        paths: o.paths.map((v) => v != null ? toSiblingPathAndIndex(v) : undefined),
    };
}
function toWsdbAppendLeavesResponse(o) {
    return {};
}
function toWsdbBatchInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbBatchInsertResponse deserialization");
    }
    ;
    return {
        result: o.result,
    };
}
function toWsdbSequentialInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbSequentialInsertResponse deserialization");
    }
    ;
    return {
        result: o.result,
    };
}
function toWsdbUpdateArchiveResponse(o) {
    return {};
}
function toWsdbCommitResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbCommitResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
function toWsdbRollbackResponse(o) {
    return {};
}
function toWsdbSyncBlockResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbSyncBlockResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
function toWsdbCreateForkResponse(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCreateForkResponse deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
function toWsdbDeleteForkResponse(o) {
    return {};
}
function toWsdbFinalizeBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbFinalizeBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusSummary(o.status),
    };
}
function toWsdbUnwindBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbUnwindBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
function toWsdbRemoveHistoricalBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
function toWsdbGetStatusResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbGetStatusResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusSummary(o.status),
    };
}
function toWsdbCreateCheckpointResponse(o) {
    return {};
}
function toWsdbCommitCheckpointResponse(o) {
    return {};
}
function toWsdbRevertCheckpointResponse(o) {
    return {};
}
function toWsdbCommitAllCheckpointsResponse(o) {
    return {};
}
function toWsdbRevertAllCheckpointsResponse(o) {
    return {};
}
function toWsdbCopyStoresResponse(o) {
    return {};
}
function toWsdbShutdownResponse(o) {
    return {};
}
function fromSiblingPathAndIndex(o) {
    if (o.index === undefined) {
        throw new Error("Expected index in SiblingPathAndIndex serialization");
    }
    if (o.path === undefined) {
        throw new Error("Expected path in SiblingPathAndIndex serialization");
    }
    ;
    return {
        index: o.index,
        path: o.path,
    };
}
function fromWorldStateStatusSummary(o) {
    if (o.unfinalizedblocknumber === undefined) {
        throw new Error("Expected unfinalizedblocknumber in WorldStateStatusSummary serialization");
    }
    if (o.finalizedblocknumber === undefined) {
        throw new Error("Expected finalizedblocknumber in WorldStateStatusSummary serialization");
    }
    if (o.oldesthistoricalblock === undefined) {
        throw new Error("Expected oldesthistoricalblock in WorldStateStatusSummary serialization");
    }
    if (o.treesaresynched === undefined) {
        throw new Error("Expected treesaresynched in WorldStateStatusSummary serialization");
    }
    ;
    return {
        unfinalizedBlockNumber: o.unfinalizedblocknumber,
        finalizedBlockNumber: o.finalizedblocknumber,
        oldestHistoricalBlock: o.oldesthistoricalblock,
        treesAreSynched: o.treesaresynched,
    };
}
function fromDBStats(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in DBStats serialization");
    }
    if (o.numdataitems === undefined) {
        throw new Error("Expected numdataitems in DBStats serialization");
    }
    if (o.totalusedsize === undefined) {
        throw new Error("Expected totalusedsize in DBStats serialization");
    }
    ;
    return {
        name: o.name,
        numDataItems: o.numdataitems,
        totalUsedSize: o.totalusedsize,
    };
}
function fromTreeDBStats(o) {
    if (o.mapsize === undefined) {
        throw new Error("Expected mapsize in TreeDBStats serialization");
    }
    if (o.physicalfilesize === undefined) {
        throw new Error("Expected physicalfilesize in TreeDBStats serialization");
    }
    if (o.blocksdbstats === undefined) {
        throw new Error("Expected blocksdbstats in TreeDBStats serialization");
    }
    if (o.nodesdbstats === undefined) {
        throw new Error("Expected nodesdbstats in TreeDBStats serialization");
    }
    if (o.leafpreimagesdbstats === undefined) {
        throw new Error("Expected leafpreimagesdbstats in TreeDBStats serialization");
    }
    if (o.leafindicesdbstats === undefined) {
        throw new Error("Expected leafindicesdbstats in TreeDBStats serialization");
    }
    if (o.blockindicesdbstats === undefined) {
        throw new Error("Expected blockindicesdbstats in TreeDBStats serialization");
    }
    ;
    return {
        mapSize: o.mapsize,
        physicalFileSize: o.physicalfilesize,
        blocksDBStats: fromDBStats(o.blocksdbstats),
        nodesDBStats: fromDBStats(o.nodesdbstats),
        leafPreimagesDBStats: fromDBStats(o.leafpreimagesdbstats),
        leafIndicesDBStats: fromDBStats(o.leafindicesdbstats),
        blockIndicesDBStats: fromDBStats(o.blockindicesdbstats),
    };
}
function fromWorldStateDBStats(o) {
    if (o.notehashtreestats === undefined) {
        throw new Error("Expected notehashtreestats in WorldStateDBStats serialization");
    }
    if (o.messagetreestats === undefined) {
        throw new Error("Expected messagetreestats in WorldStateDBStats serialization");
    }
    if (o.archivetreestats === undefined) {
        throw new Error("Expected archivetreestats in WorldStateDBStats serialization");
    }
    if (o.publicdatatreestats === undefined) {
        throw new Error("Expected publicdatatreestats in WorldStateDBStats serialization");
    }
    if (o.nullifiertreestats === undefined) {
        throw new Error("Expected nullifiertreestats in WorldStateDBStats serialization");
    }
    ;
    return {
        noteHashTreeStats: fromTreeDBStats(o.notehashtreestats),
        messageTreeStats: fromTreeDBStats(o.messagetreestats),
        archiveTreeStats: fromTreeDBStats(o.archivetreestats),
        publicDataTreeStats: fromTreeDBStats(o.publicdatatreestats),
        nullifierTreeStats: fromTreeDBStats(o.nullifiertreestats),
    };
}
function fromTreeMeta(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in TreeMeta serialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in TreeMeta serialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in TreeMeta serialization");
    }
    if (o.committedsize === undefined) {
        throw new Error("Expected committedsize in TreeMeta serialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in TreeMeta serialization");
    }
    if (o.initialsize === undefined) {
        throw new Error("Expected initialsize in TreeMeta serialization");
    }
    if (o.initialroot === undefined) {
        throw new Error("Expected initialroot in TreeMeta serialization");
    }
    if (o.oldesthistoricblock === undefined) {
        throw new Error("Expected oldesthistoricblock in TreeMeta serialization");
    }
    if (o.unfinalizedblockheight === undefined) {
        throw new Error("Expected unfinalizedblockheight in TreeMeta serialization");
    }
    if (o.finalizedblockheight === undefined) {
        throw new Error("Expected finalizedblockheight in TreeMeta serialization");
    }
    ;
    return {
        name: o.name,
        depth: o.depth,
        size: o.size,
        committedSize: o.committedsize,
        root: o.root,
        initialSize: o.initialsize,
        initialRoot: o.initialroot,
        oldestHistoricBlock: o.oldesthistoricblock,
        unfinalizedBlockHeight: o.unfinalizedblockheight,
        finalizedBlockHeight: o.finalizedblockheight,
    };
}
function fromWorldStateMeta(o) {
    if (o.notehashtreemeta === undefined) {
        throw new Error("Expected notehashtreemeta in WorldStateMeta serialization");
    }
    if (o.messagetreemeta === undefined) {
        throw new Error("Expected messagetreemeta in WorldStateMeta serialization");
    }
    if (o.archivetreemeta === undefined) {
        throw new Error("Expected archivetreemeta in WorldStateMeta serialization");
    }
    if (o.publicdatatreemeta === undefined) {
        throw new Error("Expected publicdatatreemeta in WorldStateMeta serialization");
    }
    if (o.nullifiertreemeta === undefined) {
        throw new Error("Expected nullifiertreemeta in WorldStateMeta serialization");
    }
    ;
    return {
        noteHashTreeMeta: fromTreeMeta(o.notehashtreemeta),
        messageTreeMeta: fromTreeMeta(o.messagetreemeta),
        archiveTreeMeta: fromTreeMeta(o.archivetreemeta),
        publicDataTreeMeta: fromTreeMeta(o.publicdatatreemeta),
        nullifierTreeMeta: fromTreeMeta(o.nullifiertreemeta),
    };
}
function fromWorldStateStatusFull(o) {
    if (o.summary === undefined) {
        throw new Error("Expected summary in WorldStateStatusFull serialization");
    }
    if (o.dbstats === undefined) {
        throw new Error("Expected dbstats in WorldStateStatusFull serialization");
    }
    if (o.meta === undefined) {
        throw new Error("Expected meta in WorldStateStatusFull serialization");
    }
    ;
    return {
        summary: fromWorldStateStatusSummary(o.summary),
        dbStats: fromWorldStateDBStats(o.dbstats),
        meta: fromWorldStateMeta(o.meta),
    };
}
function fromWorldStateRevision(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WorldStateRevision serialization");
    }
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WorldStateRevision serialization");
    }
    if (o.includeuncommitted === undefined) {
        throw new Error("Expected includeuncommitted in WorldStateRevision serialization");
    }
    ;
    return {
        forkId: o.forkid,
        blockNumber: o.blocknumber,
        includeUncommitted: o.includeuncommitted,
    };
}
function fromWsdbGetTreeInfo(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetTreeInfo serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetTreeInfo serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
    };
}
function fromWsdbGetStateReference(o) {
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetStateReference serialization");
    }
    ;
    return {
        revision: fromWorldStateRevision(o.revision),
    };
}
function fromWsdbGetInitialStateReference(o) {
    return {};
}
function fromWsdbGetLeafValue(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetLeafValue serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafValue serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetLeafValue serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
function fromWsdbGetLeafPreimage(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetLeafPreimage serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafPreimage serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetLeafPreimage serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
function fromWsdbGetSiblingPath(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetSiblingPath serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetSiblingPath serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetSiblingPath serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
function fromWsdbGetBlockNumbersForLeafIndices(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    if (o.leafindices === undefined) {
        throw new Error("Expected leafindices in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndices: o.leafindices,
    };
}
function fromWsdbFindLeafIndices(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindLeafIndices serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLeafIndices serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindLeafIndices serialization");
    }
    if (o.startindex === undefined) {
        throw new Error("Expected startindex in WsdbFindLeafIndices serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leaves: o.leaves,
        startIndex: o.startindex,
    };
}
function fromWsdbFindLowLeaf(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindLowLeaf serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLowLeaf serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in WsdbFindLowLeaf serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        key: o.key,
    };
}
function fromWsdbFindSiblingPaths(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindSiblingPaths serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindSiblingPaths serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindSiblingPaths serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leaves: o.leaves,
    };
}
function fromWsdbAppendLeaves(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbAppendLeaves serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbAppendLeaves serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbAppendLeaves serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        forkId: o.forkid,
    };
}
function fromWsdbBatchInsert(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbBatchInsert serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbBatchInsert serialization");
    }
    if (o.subtreedepth === undefined) {
        throw new Error("Expected subtreedepth in WsdbBatchInsert serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbBatchInsert serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        subtreeDepth: o.subtreedepth,
        forkId: o.forkid,
    };
}
function fromWsdbSequentialInsert(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbSequentialInsert serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbSequentialInsert serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbSequentialInsert serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        forkId: o.forkid,
    };
}
function fromWsdbUpdateArchive(o) {
    if (o.blockstateref === undefined) {
        throw new Error("Expected blockstateref in WsdbUpdateArchive serialization");
    }
    if (o.blockheaderhash === undefined) {
        throw new Error("Expected blockheaderhash in WsdbUpdateArchive serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbUpdateArchive serialization");
    }
    ;
    return {
        blockStateRef: o.blockstateref,
        blockHeaderHash: o.blockheaderhash,
        forkId: o.forkid,
    };
}
function fromWsdbCommit(o) {
    return {};
}
function fromWsdbRollback(o) {
    return {};
}
function fromNullifierLeafValue(o) {
    if (o.nullifier === undefined) {
        throw new Error("Expected nullifier in NullifierLeafValue serialization");
    }
    ;
    return {
        nullifier: o.nullifier,
    };
}
function fromPublicDataLeafValue(o) {
    if (o.slot === undefined) {
        throw new Error("Expected slot in PublicDataLeafValue serialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in PublicDataLeafValue serialization");
    }
    ;
    return {
        slot: o.slot,
        value: o.value,
    };
}
function fromWsdbSyncBlock(o) {
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WsdbSyncBlock serialization");
    }
    if (o.blockstateref === undefined) {
        throw new Error("Expected blockstateref in WsdbSyncBlock serialization");
    }
    if (o.blockheaderhash === undefined) {
        throw new Error("Expected blockheaderhash in WsdbSyncBlock serialization");
    }
    if (o.paddednotehashes === undefined) {
        throw new Error("Expected paddednotehashes in WsdbSyncBlock serialization");
    }
    if (o.paddedl1tol2messages === undefined) {
        throw new Error("Expected paddedl1tol2messages in WsdbSyncBlock serialization");
    }
    if (o.paddednullifiers === undefined) {
        throw new Error("Expected paddednullifiers in WsdbSyncBlock serialization");
    }
    if (o.publicdatawrites === undefined) {
        throw new Error("Expected publicdatawrites in WsdbSyncBlock serialization");
    }
    ;
    return {
        blockNumber: o.blocknumber,
        blockStateRef: o.blockstateref,
        blockHeaderHash: o.blockheaderhash,
        paddedNoteHashes: o.paddednotehashes,
        paddedL1ToL2Messages: o.paddedl1tol2messages,
        paddedNullifiers: o.paddednullifiers.map((v) => fromNullifierLeafValue(v)),
        publicDataWrites: o.publicdatawrites.map((v) => fromPublicDataLeafValue(v)),
    };
}
function fromWsdbCreateFork(o) {
    if (o.latest === undefined) {
        throw new Error("Expected latest in WsdbCreateFork serialization");
    }
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WsdbCreateFork serialization");
    }
    ;
    return {
        latest: o.latest,
        blockNumber: o.blocknumber,
    };
}
function fromWsdbDeleteFork(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbDeleteFork serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbFinalizeBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbFinalizeBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
function fromWsdbUnwindBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbUnwindBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
function fromWsdbRemoveHistoricalBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbRemoveHistoricalBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
function fromWsdbGetStatus(o) {
    return {};
}
function fromWsdbCreateCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCreateCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbCommitCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCommitCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbRevertCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbRevertCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbCommitAllCheckpoints(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCommitAllCheckpoints serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbRevertAllCheckpoints(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbRevertAllCheckpoints serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbCopyStores(o) {
    if (o.dstpath === undefined) {
        throw new Error("Expected dstpath in WsdbCopyStores serialization");
    }
    if (o.compact === undefined) {
        throw new Error("Expected compact in WsdbCopyStores serialization");
    }
    ;
    return {
        dstPath: o.dstpath,
        compact: o.compact,
    };
}
function fromWsdbShutdown(o) {
    return {};
}
function fromWsdbErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in WsdbErrorResponse serialization");
    }
    ;
    return {
        message: o.message,
    };
}
function fromWsdbGetTreeInfoResponse(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetTreeInfoResponse serialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in WsdbGetTreeInfoResponse serialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in WsdbGetTreeInfoResponse serialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in WsdbGetTreeInfoResponse serialization");
    }
    ;
    return {
        treeId: o.treeid,
        root: o.root,
        size: o.size,
        depth: o.depth,
    };
}
function fromWsdbGetStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetStateReferenceResponse serialization");
    }
    ;
    return {
        state: o.state,
    };
}
function fromWsdbGetInitialStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetInitialStateReferenceResponse serialization");
    }
    ;
    return {
        state: o.state,
    };
}
function fromWsdbGetLeafValueResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in WsdbGetLeafValueResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
function fromWsdbGetLeafPreimageResponse(o) {
    if (o.preimage === undefined) {
        throw new Error("Expected preimage in WsdbGetLeafPreimageResponse serialization");
    }
    ;
    return {
        preimage: o.preimage,
    };
}
function fromWsdbGetSiblingPathResponse(o) {
    if (o.path === undefined) {
        throw new Error("Expected path in WsdbGetSiblingPathResponse serialization");
    }
    ;
    return {
        path: o.path,
    };
}
function fromWsdbGetBlockNumbersForLeafIndicesResponse(o) {
    if (o.blocknumbers === undefined) {
        throw new Error("Expected blocknumbers in WsdbGetBlockNumbersForLeafIndicesResponse serialization");
    }
    ;
    return {
        blockNumbers: o.blocknumbers,
    };
}
function fromWsdbFindLeafIndicesResponse(o) {
    if (o.indices === undefined) {
        throw new Error("Expected indices in WsdbFindLeafIndicesResponse serialization");
    }
    ;
    return {
        indices: o.indices,
    };
}
function fromWsdbFindLowLeafResponse(o) {
    if (o.alreadypresent === undefined) {
        throw new Error("Expected alreadypresent in WsdbFindLowLeafResponse serialization");
    }
    if (o.index === undefined) {
        throw new Error("Expected index in WsdbFindLowLeafResponse serialization");
    }
    ;
    return {
        alreadyPresent: o.alreadypresent,
        index: o.index,
    };
}
function fromWsdbFindSiblingPathsResponse(o) {
    if (o.paths === undefined) {
        throw new Error("Expected paths in WsdbFindSiblingPathsResponse serialization");
    }
    ;
    return {
        paths: o.paths.map((v) => v != null ? fromSiblingPathAndIndex(v) : undefined),
    };
}
function fromWsdbAppendLeavesResponse(o) {
    return {};
}
function fromWsdbBatchInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbBatchInsertResponse serialization");
    }
    ;
    return {
        result: o.result,
    };
}
function fromWsdbSequentialInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbSequentialInsertResponse serialization");
    }
    ;
    return {
        result: o.result,
    };
}
function fromWsdbUpdateArchiveResponse(o) {
    return {};
}
function fromWsdbCommitResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbCommitResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
function fromWsdbRollbackResponse(o) {
    return {};
}
function fromWsdbSyncBlockResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbSyncBlockResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
function fromWsdbCreateForkResponse(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCreateForkResponse serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
function fromWsdbDeleteForkResponse(o) {
    return {};
}
function fromWsdbFinalizeBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbFinalizeBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusSummary(o.status),
    };
}
function fromWsdbUnwindBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbUnwindBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
function fromWsdbRemoveHistoricalBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
function fromWsdbGetStatusResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbGetStatusResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusSummary(o.status),
    };
}
function fromWsdbCreateCheckpointResponse(o) {
    return {};
}
function fromWsdbCommitCheckpointResponse(o) {
    return {};
}
function fromWsdbRevertCheckpointResponse(o) {
    return {};
}
function fromWsdbCommitAllCheckpointsResponse(o) {
    return {};
}
function fromWsdbRevertAllCheckpointsResponse(o) {
    return {};
}
function fromWsdbCopyStoresResponse(o) {
    return {};
}
function fromWsdbShutdownResponse(o) {
    return {};
}
