// @aztec/bb.js/dest/node-cjs/aztec-wsdb/generated/async.js
"use strict";
// AUTOGENERATED FILE - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncApi = void 0;
const msgpackr_1 = require("msgpackr");
const bbapi_exception_js_1 = require("../../bbapi_exception.js");
const api_types_js_1 = require("./api_types.js");
async function msgpackCall(backend, input) {
    const inputBuffer = new msgpackr_1.Encoder({ useRecords: false }).pack(input);
    const encodedResult = await backend.call(inputBuffer);
    return new msgpackr_1.Decoder({ useRecords: false }).unpack(encodedResult);
}
class AsyncApi {
    backend;
    constructor(backend) {
        this.backend = backend;
    }
    wsdbGetTreeInfo(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetTreeInfo)(command);
        return msgpackCall(this.backend, [["WsdbGetTreeInfo", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetTreeInfoResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetTreeInfoResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetTreeInfoResponse)(result);
        });
    }
    wsdbGetStateReference(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetStateReference)(command);
        return msgpackCall(this.backend, [["WsdbGetStateReference", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetStateReferenceResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetStateReferenceResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetStateReferenceResponse)(result);
        });
    }
    wsdbGetInitialStateReference(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetInitialStateReference)(command);
        return msgpackCall(this.backend, [["WsdbGetInitialStateReference", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetInitialStateReferenceResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetInitialStateReferenceResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetInitialStateReferenceResponse)(result);
        });
    }
    wsdbGetLeafValue(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetLeafValue)(command);
        return msgpackCall(this.backend, [["WsdbGetLeafValue", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetLeafValueResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetLeafValueResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetLeafValueResponse)(result);
        });
    }
    wsdbGetLeafPreimage(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetLeafPreimage)(command);
        return msgpackCall(this.backend, [["WsdbGetLeafPreimage", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetLeafPreimageResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetLeafPreimageResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetLeafPreimageResponse)(result);
        });
    }
    wsdbGetSiblingPath(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetSiblingPath)(command);
        return msgpackCall(this.backend, [["WsdbGetSiblingPath", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetSiblingPathResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetSiblingPathResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetSiblingPathResponse)(result);
        });
    }
    wsdbGetBlockNumbersForLeafIndices(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetBlockNumbersForLeafIndices)(command);
        return msgpackCall(this.backend, [["WsdbGetBlockNumbersForLeafIndices", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetBlockNumbersForLeafIndicesResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetBlockNumbersForLeafIndicesResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetBlockNumbersForLeafIndicesResponse)(result);
        });
    }
    wsdbFindLeafIndices(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbFindLeafIndices)(command);
        return msgpackCall(this.backend, [["WsdbFindLeafIndices", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindLeafIndicesResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbFindLeafIndicesResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbFindLeafIndicesResponse)(result);
        });
    }
    wsdbFindLowLeaf(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbFindLowLeaf)(command);
        return msgpackCall(this.backend, [["WsdbFindLowLeaf", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindLowLeafResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbFindLowLeafResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbFindLowLeafResponse)(result);
        });
    }
    wsdbFindSiblingPaths(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbFindSiblingPaths)(command);
        return msgpackCall(this.backend, [["WsdbFindSiblingPaths", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindSiblingPathsResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbFindSiblingPathsResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbFindSiblingPathsResponse)(result);
        });
    }
    wsdbAppendLeaves(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbAppendLeaves)(command);
        return msgpackCall(this.backend, [["WsdbAppendLeaves", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbAppendLeavesResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbAppendLeavesResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbAppendLeavesResponse)(result);
        });
    }
    wsdbBatchInsert(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbBatchInsert)(command);
        return msgpackCall(this.backend, [["WsdbBatchInsert", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbBatchInsertResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbBatchInsertResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbBatchInsertResponse)(result);
        });
    }
    wsdbSequentialInsert(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbSequentialInsert)(command);
        return msgpackCall(this.backend, [["WsdbSequentialInsert", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbSequentialInsertResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbSequentialInsertResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbSequentialInsertResponse)(result);
        });
    }
    wsdbUpdateArchive(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbUpdateArchive)(command);
        return msgpackCall(this.backend, [["WsdbUpdateArchive", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbUpdateArchiveResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbUpdateArchiveResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbUpdateArchiveResponse)(result);
        });
    }
    wsdbCommit(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCommit)(command);
        return msgpackCall(this.backend, [["WsdbCommit", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCommitResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCommitResponse)(result);
        });
    }
    wsdbRollback(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbRollback)(command);
        return msgpackCall(this.backend, [["WsdbRollback", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRollbackResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbRollbackResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbRollbackResponse)(result);
        });
    }
    wsdbSyncBlock(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbSyncBlock)(command);
        return msgpackCall(this.backend, [["WsdbSyncBlock", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbSyncBlockResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbSyncBlockResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbSyncBlockResponse)(result);
        });
    }
    wsdbCreateFork(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCreateFork)(command);
        return msgpackCall(this.backend, [["WsdbCreateFork", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCreateForkResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCreateForkResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCreateForkResponse)(result);
        });
    }
    wsdbDeleteFork(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbDeleteFork)(command);
        return msgpackCall(this.backend, [["WsdbDeleteFork", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbDeleteForkResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbDeleteForkResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbDeleteForkResponse)(result);
        });
    }
    wsdbFinalizeBlocks(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbFinalizeBlocks)(command);
        return msgpackCall(this.backend, [["WsdbFinalizeBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFinalizeBlocksResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbFinalizeBlocksResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbFinalizeBlocksResponse)(result);
        });
    }
    wsdbUnwindBlocks(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbUnwindBlocks)(command);
        return msgpackCall(this.backend, [["WsdbUnwindBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbUnwindBlocksResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbUnwindBlocksResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbUnwindBlocksResponse)(result);
        });
    }
    wsdbRemoveHistoricalBlocks(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbRemoveHistoricalBlocks)(command);
        return msgpackCall(this.backend, [["WsdbRemoveHistoricalBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRemoveHistoricalBlocksResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbRemoveHistoricalBlocksResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbRemoveHistoricalBlocksResponse)(result);
        });
    }
    wsdbGetStatus(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbGetStatus)(command);
        return msgpackCall(this.backend, [["WsdbGetStatus", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetStatusResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbGetStatusResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbGetStatusResponse)(result);
        });
    }
    wsdbCreateCheckpoint(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCreateCheckpoint)(command);
        return msgpackCall(this.backend, [["WsdbCreateCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCreateCheckpointResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCreateCheckpointResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCreateCheckpointResponse)(result);
        });
    }
    wsdbCommitCheckpoint(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCommitCheckpoint)(command);
        return msgpackCall(this.backend, [["WsdbCommitCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitCheckpointResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCommitCheckpointResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCommitCheckpointResponse)(result);
        });
    }
    wsdbRevertCheckpoint(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbRevertCheckpoint)(command);
        return msgpackCall(this.backend, [["WsdbRevertCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRevertCheckpointResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbRevertCheckpointResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbRevertCheckpointResponse)(result);
        });
    }
    wsdbCommitAllCheckpoints(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCommitAllCheckpoints)(command);
        return msgpackCall(this.backend, [["WsdbCommitAllCheckpoints", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitAllCheckpointsResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCommitAllCheckpointsResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCommitAllCheckpointsResponse)(result);
        });
    }
    wsdbRevertAllCheckpoints(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbRevertAllCheckpoints)(command);
        return msgpackCall(this.backend, [["WsdbRevertAllCheckpoints", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRevertAllCheckpointsResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbRevertAllCheckpointsResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbRevertAllCheckpointsResponse)(result);
        });
    }
    wsdbCopyStores(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbCopyStores)(command);
        return msgpackCall(this.backend, [["WsdbCopyStores", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCopyStoresResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbCopyStoresResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbCopyStoresResponse)(result);
        });
    }
    wsdbShutdown(command) {
        const msgpackCommand = (0, api_types_js_1.fromWsdbShutdown)(command);
        return msgpackCall(this.backend, [["WsdbShutdown", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new bbapi_exception_js_1.BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbShutdownResponse') {
                throw new bbapi_exception_js_1.BBApiException(`Expected variant name 'WsdbShutdownResponse' but got '${variantName}'`);
            }
            return (0, api_types_js_1.toWsdbShutdownResponse)(result);
        });
    }
    destroy() {
        return this.backend.destroy ? this.backend.destroy() : Promise.resolve();
    }
}
exports.AsyncApi = AsyncApi;
