// @aztec/bb.js/dest/node-cjs/cbind/cpp_codegen.js
"use strict";
/**
 * C++ IPC Client Code Generator
 *
 * Generates a C++ IPC client from a CompiledSchema. The generated client:
 *   - Connects to a server over Unix Domain Socket via ipc::IpcClient
 *   - Wraps each command in a NamedUnion, serializes with msgpack, sends, receives, deserializes
 *   - Has one method per command, returning the typed response
 *
 * Usage:
 *   const gen = new CppCodegen({ namespace: 'bb::cdb', prefix: 'Cdb' });
 *   const header = gen.generateHeader(schema);
 *   const impl = gen.generateImpl(schema);
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CppCodegen = void 0;
const naming_js_1 = require("./naming.js");
class CppCodegen {
    opts;
    constructor(opts) {
        this.opts = opts;
    }
    /** Convert a command name to a C++ method name (snake_case without prefix) */
    methodName(commandName) {
        // Strip prefix: "CdbGetContractInstance" -> "GetContractInstance" -> "get_contract_instance"
        const withoutPrefix = commandName.startsWith(this.opts.prefix)
            ? commandName.slice(this.opts.prefix.length)
            : commandName;
        return (0, naming_js_1.toSnakeCase)(withoutPrefix);
    }
    /** Check if the response has fields (non-void return) */
    hasResponseFields(command, schema) {
        const resp = schema.responses.get(command.responseType);
        return !!resp && resp.fields.length > 0;
    }
    /** Generate the method signature using command struct types directly */
    generateMethodSignature(command, schema, className) {
        const method = this.methodName(command.name);
        const hasFields = this.hasResponseFields(command, schema);
        const retType = hasFields ? `${command.name}::Response` : 'void';
        // If the command has fields, take the whole command struct by value
        const params = command.fields.length > 0 ? `${command.name} cmd` : '';
        const prefix = className ? `${className}::` : '';
        const constSuffix = !this.isWriteCommand(command) ? ' const' : '';
        return `${retType} ${prefix}${method}(${params})${constSuffix}`;
    }
    /** Check if a command modifies state (non-const) */
    isWriteCommand(command) {
        const name = command.name.toLowerCase();
        return name.includes('add') || name.includes('create') ||
            name.includes('commit') || name.includes('revert') ||
            name.includes('register') || name.includes('shutdown') ||
            name.includes('delete') || name.includes('sync') ||
            name.includes('rollback') || name.includes('unwind');
    }
    /** Generate the header file */
    generateHeader(schema) {
        const { namespace: ns, prefix } = this.opts;
        const className = `${prefix}IpcClient`;
        const guardName = `${ns.replace(/::/g, '_').toUpperCase()}_${prefix.toUpperCase()}_IPC_CLIENT_GENERATED_HPP`;
        const methods = schema.commands.map(cmd => {
            const sig = this.generateMethodSignature(cmd, schema);
            return `    ${sig};`;
        }).join('\n');
        return `// AUTOGENERATED FILE - DO NOT EDIT
#pragma once

#include "barretenberg/common/try_catch_shim.hpp"
#include "${this.opts.executeHeader}"
#include "barretenberg/ipc/ipc_client.hpp"

#include <memory>
#include <string>

namespace ${ns} {

/**
 * @brief Auto-generated IPC client.
 *
 * Each method sends a msgpack-serialized command to the server over UDS
 * and returns the typed response. All methods block until the response arrives.
 */
class ${className} {
  public:
    explicit ${className}(const std::string& socket_path);
    ~${className}();

    ${className}(const ${className}&) = delete;
    ${className}& operator=(const ${className}&) = delete;

${methods}

  private:
    template <typename Cmd>
    typename Cmd::Response send(Cmd&& cmd) const;

    mutable std::unique_ptr<ipc::IpcClient> client_;
};

} // namespace ${ns}
`;
    }
    /** Generate the implementation file */
    generateImpl(schema) {
        const { namespace: ns, prefix } = this.opts;
        const className = `${prefix}IpcClient`;
        const commandType = `${prefix}Command`;
        const responseType = `${prefix}CommandResponse`;
        const errorType = `${prefix}ErrorResponse`;
        const methods = schema.commands.map(cmd => {
            return this.generateMethodImpl(cmd, schema, className);
        }).join('\n');
        return `// AUTOGENERATED FILE - DO NOT EDIT

#include "${this.headerIncludePath()}"
#include "${this.opts.executeHeader}"
#include "barretenberg/serialize/msgpack.hpp"
#include "barretenberg/serialize/msgpack_impl.hpp"

#include <stdexcept>
#include <tuple>

namespace ${ns} {

${className}::${className}(const std::string& socket_path)
    : client_(ipc::IpcClient::create_socket(socket_path))
{
    if (!client_->connect()) {
        throw std::runtime_error("Failed to connect to server at " + socket_path);
    }
}

${className}::~${className}()
{
    if (client_) {
        client_->close();
    }
}

template <typename Cmd>
typename Cmd::Response ${className}::send(Cmd&& cmd) const
{
    // Wrap command in ${commandType} NamedUnion, then in a 1-element tuple (matches server expectations)
    ${commandType} command = std::forward<Cmd>(cmd);
    auto wrapped = std::make_tuple(std::move(command));

    // Serialize to msgpack
    msgpack::sbuffer send_buffer;
    msgpack::pack(send_buffer, wrapped);

    // Send to server
    constexpr uint64_t timeout_ns = 30'000'000'000ULL; // 30 seconds
    if (!client_->send(send_buffer.data(), send_buffer.size(), timeout_ns)) {
        throw std::runtime_error("Failed to send command to server");
    }

    // Receive response
    auto response_span = client_->receive(timeout_ns);
    if (response_span.empty()) {
        throw std::runtime_error("Empty response from server");
    }

    // Deserialize response
    auto unpacked = msgpack::unpack(reinterpret_cast<const char*>(response_span.data()), response_span.size());
    auto response_obj = unpacked.get();

    ${responseType} response;
    response_obj.convert(response);

    // Release the receive buffer
    client_->release(response_span.size());

    // Check for error response
    return std::move(response).visit([](auto&& resp) -> typename Cmd::Response {
        using RespType = std::decay_t<decltype(resp)>;

        if constexpr (std::is_same_v<RespType, ${errorType}>) {
            throw std::runtime_error("Server error: " + resp.message);
        } else if constexpr (std::is_same_v<RespType, typename Cmd::Response>) {
            return std::forward<decltype(resp)>(resp);
        } else {
            throw std::runtime_error("Unexpected response type from server");
        }
    });
}

${methods}
} // namespace ${ns}
`;
    }
    /** Generate a single method implementation */
    generateMethodImpl(command, schema, className) {
        const sig = this.generateMethodSignature(command, schema, className);
        const hasFields = this.hasResponseFields(command, schema);
        const cmdExpr = command.fields.length > 0 ? 'std::move(cmd)' : `${command.name}{}`;
        if (!hasFields) {
            return `${sig}
{
    send(${cmdExpr});
}
`;
        }
        return `${sig}
{
    return send(${cmdExpr});
}
`;
    }
    /** Compute the include path for the generated header */
    headerIncludePath() {
        // Derive from the executeHeader path: replace _execute.hpp with _ipc_client_generated.hpp
        const dir = this.opts.executeHeader.substring(0, this.opts.executeHeader.lastIndexOf('/'));
        return `${dir}/${(0, naming_js_1.toSnakeCase)(this.opts.prefix)}_ipc_client_generated.hpp`;
    }
}
exports.CppCodegen = CppCodegen;
