// @aztec/bb.js/dest/browser/aztec-wsdb/generated/async.js
// AUTOGENERATED FILE - DO NOT EDIT
import { Decoder, Encoder } from 'msgpackr';
import { BBApiException } from '../../bbapi_exception.js';
import { fromWsdbAppendLeaves, fromWsdbBatchInsert, fromWsdbCommit, fromWsdbCommitAllCheckpoints, fromWsdbCommitCheckpoint, fromWsdbCopyStores, fromWsdbCreateCheckpoint, fromWsdbCreateFork, fromWsdbDeleteFork, fromWsdbFinalizeBlocks, fromWsdbFindLeafIndices, fromWsdbFindLowLeaf, fromWsdbFindSiblingPaths, fromWsdbGetBlockNumbersForLeafIndices, fromWsdbGetInitialStateReference, fromWsdbGetLeafPreimage, fromWsdbGetLeafValue, fromWsdbGetSiblingPath, fromWsdbGetStateReference, fromWsdbGetStatus, fromWsdbGetTreeInfo, fromWsdbRemoveHistoricalBlocks, fromWsdbRevertAllCheckpoints, fromWsdbRevertCheckpoint, fromWsdbRollback, fromWsdbSequentialInsert, fromWsdbShutdown, fromWsdbSyncBlock, fromWsdbUnwindBlocks, fromWsdbUpdateArchive, toWsdbAppendLeavesResponse, toWsdbBatchInsertResponse, toWsdbCommitAllCheckpointsResponse, toWsdbCommitCheckpointResponse, toWsdbCommitResponse, toWsdbCopyStoresResponse, toWsdbCreateCheckpointResponse, toWsdbCreateForkResponse, toWsdbDeleteForkResponse, toWsdbFinalizeBlocksResponse, toWsdbFindLeafIndicesResponse, toWsdbFindLowLeafResponse, toWsdbFindSiblingPathsResponse, toWsdbGetBlockNumbersForLeafIndicesResponse, toWsdbGetInitialStateReferenceResponse, toWsdbGetLeafPreimageResponse, toWsdbGetLeafValueResponse, toWsdbGetSiblingPathResponse, toWsdbGetStateReferenceResponse, toWsdbGetStatusResponse, toWsdbGetTreeInfoResponse, toWsdbRemoveHistoricalBlocksResponse, toWsdbRevertAllCheckpointsResponse, toWsdbRevertCheckpointResponse, toWsdbRollbackResponse, toWsdbSequentialInsertResponse, toWsdbShutdownResponse, toWsdbSyncBlockResponse, toWsdbUnwindBlocksResponse, toWsdbUpdateArchiveResponse } from './api_types.js';
async function msgpackCall(backend, input) {
    const inputBuffer = new Encoder({ useRecords: false }).pack(input);
    const encodedResult = await backend.call(inputBuffer);
    return new Decoder({ useRecords: false }).unpack(encodedResult);
}
export class AsyncApi {
    backend;
    constructor(backend) {
        this.backend = backend;
    }
    wsdbGetTreeInfo(command) {
        const msgpackCommand = fromWsdbGetTreeInfo(command);
        return msgpackCall(this.backend, [["WsdbGetTreeInfo", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetTreeInfoResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetTreeInfoResponse' but got '${variantName}'`);
            }
            return toWsdbGetTreeInfoResponse(result);
        });
    }
    wsdbGetStateReference(command) {
        const msgpackCommand = fromWsdbGetStateReference(command);
        return msgpackCall(this.backend, [["WsdbGetStateReference", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetStateReferenceResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetStateReferenceResponse' but got '${variantName}'`);
            }
            return toWsdbGetStateReferenceResponse(result);
        });
    }
    wsdbGetInitialStateReference(command) {
        const msgpackCommand = fromWsdbGetInitialStateReference(command);
        return msgpackCall(this.backend, [["WsdbGetInitialStateReference", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetInitialStateReferenceResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetInitialStateReferenceResponse' but got '${variantName}'`);
            }
            return toWsdbGetInitialStateReferenceResponse(result);
        });
    }
    wsdbGetLeafValue(command) {
        const msgpackCommand = fromWsdbGetLeafValue(command);
        return msgpackCall(this.backend, [["WsdbGetLeafValue", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetLeafValueResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetLeafValueResponse' but got '${variantName}'`);
            }
            return toWsdbGetLeafValueResponse(result);
        });
    }
    wsdbGetLeafPreimage(command) {
        const msgpackCommand = fromWsdbGetLeafPreimage(command);
        return msgpackCall(this.backend, [["WsdbGetLeafPreimage", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetLeafPreimageResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetLeafPreimageResponse' but got '${variantName}'`);
            }
            return toWsdbGetLeafPreimageResponse(result);
        });
    }
    wsdbGetSiblingPath(command) {
        const msgpackCommand = fromWsdbGetSiblingPath(command);
        return msgpackCall(this.backend, [["WsdbGetSiblingPath", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetSiblingPathResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetSiblingPathResponse' but got '${variantName}'`);
            }
            return toWsdbGetSiblingPathResponse(result);
        });
    }
    wsdbGetBlockNumbersForLeafIndices(command) {
        const msgpackCommand = fromWsdbGetBlockNumbersForLeafIndices(command);
        return msgpackCall(this.backend, [["WsdbGetBlockNumbersForLeafIndices", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetBlockNumbersForLeafIndicesResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetBlockNumbersForLeafIndicesResponse' but got '${variantName}'`);
            }
            return toWsdbGetBlockNumbersForLeafIndicesResponse(result);
        });
    }
    wsdbFindLeafIndices(command) {
        const msgpackCommand = fromWsdbFindLeafIndices(command);
        return msgpackCall(this.backend, [["WsdbFindLeafIndices", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindLeafIndicesResponse') {
                throw new BBApiException(`Expected variant name 'WsdbFindLeafIndicesResponse' but got '${variantName}'`);
            }
            return toWsdbFindLeafIndicesResponse(result);
        });
    }
    wsdbFindLowLeaf(command) {
        const msgpackCommand = fromWsdbFindLowLeaf(command);
        return msgpackCall(this.backend, [["WsdbFindLowLeaf", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindLowLeafResponse') {
                throw new BBApiException(`Expected variant name 'WsdbFindLowLeafResponse' but got '${variantName}'`);
            }
            return toWsdbFindLowLeafResponse(result);
        });
    }
    wsdbFindSiblingPaths(command) {
        const msgpackCommand = fromWsdbFindSiblingPaths(command);
        return msgpackCall(this.backend, [["WsdbFindSiblingPaths", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFindSiblingPathsResponse') {
                throw new BBApiException(`Expected variant name 'WsdbFindSiblingPathsResponse' but got '${variantName}'`);
            }
            return toWsdbFindSiblingPathsResponse(result);
        });
    }
    wsdbAppendLeaves(command) {
        const msgpackCommand = fromWsdbAppendLeaves(command);
        return msgpackCall(this.backend, [["WsdbAppendLeaves", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbAppendLeavesResponse') {
                throw new BBApiException(`Expected variant name 'WsdbAppendLeavesResponse' but got '${variantName}'`);
            }
            return toWsdbAppendLeavesResponse(result);
        });
    }
    wsdbBatchInsert(command) {
        const msgpackCommand = fromWsdbBatchInsert(command);
        return msgpackCall(this.backend, [["WsdbBatchInsert", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbBatchInsertResponse') {
                throw new BBApiException(`Expected variant name 'WsdbBatchInsertResponse' but got '${variantName}'`);
            }
            return toWsdbBatchInsertResponse(result);
        });
    }
    wsdbSequentialInsert(command) {
        const msgpackCommand = fromWsdbSequentialInsert(command);
        return msgpackCall(this.backend, [["WsdbSequentialInsert", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbSequentialInsertResponse') {
                throw new BBApiException(`Expected variant name 'WsdbSequentialInsertResponse' but got '${variantName}'`);
            }
            return toWsdbSequentialInsertResponse(result);
        });
    }
    wsdbUpdateArchive(command) {
        const msgpackCommand = fromWsdbUpdateArchive(command);
        return msgpackCall(this.backend, [["WsdbUpdateArchive", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbUpdateArchiveResponse') {
                throw new BBApiException(`Expected variant name 'WsdbUpdateArchiveResponse' but got '${variantName}'`);
            }
            return toWsdbUpdateArchiveResponse(result);
        });
    }
    wsdbCommit(command) {
        const msgpackCommand = fromWsdbCommit(command);
        return msgpackCall(this.backend, [["WsdbCommit", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCommitResponse' but got '${variantName}'`);
            }
            return toWsdbCommitResponse(result);
        });
    }
    wsdbRollback(command) {
        const msgpackCommand = fromWsdbRollback(command);
        return msgpackCall(this.backend, [["WsdbRollback", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRollbackResponse') {
                throw new BBApiException(`Expected variant name 'WsdbRollbackResponse' but got '${variantName}'`);
            }
            return toWsdbRollbackResponse(result);
        });
    }
    wsdbSyncBlock(command) {
        const msgpackCommand = fromWsdbSyncBlock(command);
        return msgpackCall(this.backend, [["WsdbSyncBlock", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbSyncBlockResponse') {
                throw new BBApiException(`Expected variant name 'WsdbSyncBlockResponse' but got '${variantName}'`);
            }
            return toWsdbSyncBlockResponse(result);
        });
    }
    wsdbCreateFork(command) {
        const msgpackCommand = fromWsdbCreateFork(command);
        return msgpackCall(this.backend, [["WsdbCreateFork", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCreateForkResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCreateForkResponse' but got '${variantName}'`);
            }
            return toWsdbCreateForkResponse(result);
        });
    }
    wsdbDeleteFork(command) {
        const msgpackCommand = fromWsdbDeleteFork(command);
        return msgpackCall(this.backend, [["WsdbDeleteFork", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbDeleteForkResponse') {
                throw new BBApiException(`Expected variant name 'WsdbDeleteForkResponse' but got '${variantName}'`);
            }
            return toWsdbDeleteForkResponse(result);
        });
    }
    wsdbFinalizeBlocks(command) {
        const msgpackCommand = fromWsdbFinalizeBlocks(command);
        return msgpackCall(this.backend, [["WsdbFinalizeBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbFinalizeBlocksResponse') {
                throw new BBApiException(`Expected variant name 'WsdbFinalizeBlocksResponse' but got '${variantName}'`);
            }
            return toWsdbFinalizeBlocksResponse(result);
        });
    }
    wsdbUnwindBlocks(command) {
        const msgpackCommand = fromWsdbUnwindBlocks(command);
        return msgpackCall(this.backend, [["WsdbUnwindBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbUnwindBlocksResponse') {
                throw new BBApiException(`Expected variant name 'WsdbUnwindBlocksResponse' but got '${variantName}'`);
            }
            return toWsdbUnwindBlocksResponse(result);
        });
    }
    wsdbRemoveHistoricalBlocks(command) {
        const msgpackCommand = fromWsdbRemoveHistoricalBlocks(command);
        return msgpackCall(this.backend, [["WsdbRemoveHistoricalBlocks", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRemoveHistoricalBlocksResponse') {
                throw new BBApiException(`Expected variant name 'WsdbRemoveHistoricalBlocksResponse' but got '${variantName}'`);
            }
            return toWsdbRemoveHistoricalBlocksResponse(result);
        });
    }
    wsdbGetStatus(command) {
        const msgpackCommand = fromWsdbGetStatus(command);
        return msgpackCall(this.backend, [["WsdbGetStatus", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbGetStatusResponse') {
                throw new BBApiException(`Expected variant name 'WsdbGetStatusResponse' but got '${variantName}'`);
            }
            return toWsdbGetStatusResponse(result);
        });
    }
    wsdbCreateCheckpoint(command) {
        const msgpackCommand = fromWsdbCreateCheckpoint(command);
        return msgpackCall(this.backend, [["WsdbCreateCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCreateCheckpointResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCreateCheckpointResponse' but got '${variantName}'`);
            }
            return toWsdbCreateCheckpointResponse(result);
        });
    }
    wsdbCommitCheckpoint(command) {
        const msgpackCommand = fromWsdbCommitCheckpoint(command);
        return msgpackCall(this.backend, [["WsdbCommitCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitCheckpointResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCommitCheckpointResponse' but got '${variantName}'`);
            }
            return toWsdbCommitCheckpointResponse(result);
        });
    }
    wsdbRevertCheckpoint(command) {
        const msgpackCommand = fromWsdbRevertCheckpoint(command);
        return msgpackCall(this.backend, [["WsdbRevertCheckpoint", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRevertCheckpointResponse') {
                throw new BBApiException(`Expected variant name 'WsdbRevertCheckpointResponse' but got '${variantName}'`);
            }
            return toWsdbRevertCheckpointResponse(result);
        });
    }
    wsdbCommitAllCheckpoints(command) {
        const msgpackCommand = fromWsdbCommitAllCheckpoints(command);
        return msgpackCall(this.backend, [["WsdbCommitAllCheckpoints", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCommitAllCheckpointsResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCommitAllCheckpointsResponse' but got '${variantName}'`);
            }
            return toWsdbCommitAllCheckpointsResponse(result);
        });
    }
    wsdbRevertAllCheckpoints(command) {
        const msgpackCommand = fromWsdbRevertAllCheckpoints(command);
        return msgpackCall(this.backend, [["WsdbRevertAllCheckpoints", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbRevertAllCheckpointsResponse') {
                throw new BBApiException(`Expected variant name 'WsdbRevertAllCheckpointsResponse' but got '${variantName}'`);
            }
            return toWsdbRevertAllCheckpointsResponse(result);
        });
    }
    wsdbCopyStores(command) {
        const msgpackCommand = fromWsdbCopyStores(command);
        return msgpackCall(this.backend, [["WsdbCopyStores", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbCopyStoresResponse') {
                throw new BBApiException(`Expected variant name 'WsdbCopyStoresResponse' but got '${variantName}'`);
            }
            return toWsdbCopyStoresResponse(result);
        });
    }
    wsdbShutdown(command) {
        const msgpackCommand = fromWsdbShutdown(command);
        return msgpackCall(this.backend, [["WsdbShutdown", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName === 'WsdbErrorResponse') {
                throw new BBApiException(result.message || 'Unknown error from barretenberg');
            }
            if (variantName !== 'WsdbShutdownResponse') {
                throw new BBApiException(`Expected variant name 'WsdbShutdownResponse' but got '${variantName}'`);
            }
            return toWsdbShutdownResponse(result);
        });
    }
    destroy() {
        return this.backend.destroy ? this.backend.destroy() : Promise.resolve();
    }
}
