// @aztec/bb.js/dest/node-cjs/crs/browser/cached_net_crs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CachedNetGrumpkinCrs = exports.CachedNetCrs = void 0;
const net_crs_js_1 = require("../net_crs.js");
const idb_keyval_1 = require("idb-keyval");
/**
 * Downloader for CRS from the web or local.
 */
class CachedNetCrs {
    numPoints;
    g1Data;
    g2Data;
    constructor(numPoints) {
        this.numPoints = numPoints;
    }
    static async new(numPoints) {
        const crs = new CachedNetCrs(numPoints);
        await crs.init();
        return crs;
    }
    /**
     * Download the data.
     */
    async init() {
        const g2Data = await (0, idb_keyval_1.get)('g2Data');
        // Prefer cached uncompressed (64 bytes/point, fast path: no decompression needed)
        const g1Uncompressed = await (0, idb_keyval_1.get)('g1Data');
        const uncompressedLength = this.numPoints * 64;
        if (g1Uncompressed && g1Uncompressed.length >= uncompressedLength) {
            this.g1Data = g1Uncompressed;
        }
        else {
            // Download compressed from CDN
            const netCrs = new net_crs_js_1.NetCrs(this.numPoints);
            this.g1Data = await netCrs.downloadG1Data();
        }
        if (!g2Data) {
            const netCrs = new net_crs_js_1.NetCrs(this.numPoints);
            this.g2Data = await netCrs.downloadG2Data();
            await (0, idb_keyval_1.set)('g2Data', this.g2Data);
        }
        else {
            this.g2Data = g2Data;
        }
    }
    /**
     * G1 points data for prover key (compressed or uncompressed).
     */
    getG1Data() {
        return this.g1Data;
    }
    /**
     * Cache uncompressed G1 data in IndexedDB after WASM decompression.
     */
    async cacheUncompressed(data) {
        await (0, idb_keyval_1.set)('g1Data', data);
    }
    /**
     * G2 points data for verification key.
     * @returns The points data.
     */
    getG2Data() {
        return this.g2Data;
    }
}
exports.CachedNetCrs = CachedNetCrs;
/**
 * Downloader for CRS from the web or local.
 */
class CachedNetGrumpkinCrs {
    numPoints;
    g1Data;
    constructor(numPoints) {
        this.numPoints = numPoints;
    }
    static async new(numPoints) {
        const crs = new CachedNetGrumpkinCrs(numPoints);
        await crs.init();
        return crs;
    }
    /**
     * Download the data.
     */
    async init() {
        // Check if data is in IndexedDB
        const g1Data = await (0, idb_keyval_1.get)('grumpkinG1Data');
        const netGrumpkinCrs = new net_crs_js_1.NetGrumpkinCrs(this.numPoints);
        const g1DataLength = this.numPoints * 64;
        if (!g1Data || g1Data.length < g1DataLength) {
            this.g1Data = await netGrumpkinCrs.downloadG1Data();
            await (0, idb_keyval_1.set)('grumpkinG1Data', this.g1Data);
        }
        else {
            this.g1Data = g1Data;
        }
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return this.g1Data;
    }
}
exports.CachedNetGrumpkinCrs = CachedNetGrumpkinCrs;
