// @aztec/bb.js/dest/node/aztec-wsdb/generated/api_types.js
// AUTOGENERATED FILE - DO NOT EDIT
// Conversion functions (exported)
export function toSiblingPathAndIndex(o) {
    if (o.index === undefined) {
        throw new Error("Expected index in SiblingPathAndIndex deserialization");
    }
    if (o.path === undefined) {
        throw new Error("Expected path in SiblingPathAndIndex deserialization");
    }
    ;
    return {
        index: o.index,
        path: o.path,
    };
}
export function toWorldStateStatusSummary(o) {
    if (o.unfinalizedBlockNumber === undefined) {
        throw new Error("Expected unfinalizedBlockNumber in WorldStateStatusSummary deserialization");
    }
    if (o.finalizedBlockNumber === undefined) {
        throw new Error("Expected finalizedBlockNumber in WorldStateStatusSummary deserialization");
    }
    if (o.oldestHistoricalBlock === undefined) {
        throw new Error("Expected oldestHistoricalBlock in WorldStateStatusSummary deserialization");
    }
    if (o.treesAreSynched === undefined) {
        throw new Error("Expected treesAreSynched in WorldStateStatusSummary deserialization");
    }
    ;
    return {
        unfinalizedblocknumber: o.unfinalizedBlockNumber,
        finalizedblocknumber: o.finalizedBlockNumber,
        oldesthistoricalblock: o.oldestHistoricalBlock,
        treesaresynched: o.treesAreSynched,
    };
}
export function toDBStats(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in DBStats deserialization");
    }
    if (o.numDataItems === undefined) {
        throw new Error("Expected numDataItems in DBStats deserialization");
    }
    if (o.totalUsedSize === undefined) {
        throw new Error("Expected totalUsedSize in DBStats deserialization");
    }
    ;
    return {
        name: o.name,
        numdataitems: o.numDataItems,
        totalusedsize: o.totalUsedSize,
    };
}
export function toTreeDBStats(o) {
    if (o.mapSize === undefined) {
        throw new Error("Expected mapSize in TreeDBStats deserialization");
    }
    if (o.physicalFileSize === undefined) {
        throw new Error("Expected physicalFileSize in TreeDBStats deserialization");
    }
    if (o.blocksDBStats === undefined) {
        throw new Error("Expected blocksDBStats in TreeDBStats deserialization");
    }
    if (o.nodesDBStats === undefined) {
        throw new Error("Expected nodesDBStats in TreeDBStats deserialization");
    }
    if (o.leafPreimagesDBStats === undefined) {
        throw new Error("Expected leafPreimagesDBStats in TreeDBStats deserialization");
    }
    if (o.leafIndicesDBStats === undefined) {
        throw new Error("Expected leafIndicesDBStats in TreeDBStats deserialization");
    }
    if (o.blockIndicesDBStats === undefined) {
        throw new Error("Expected blockIndicesDBStats in TreeDBStats deserialization");
    }
    ;
    return {
        mapsize: o.mapSize,
        physicalfilesize: o.physicalFileSize,
        blocksdbstats: toDBStats(o.blocksDBStats),
        nodesdbstats: toDBStats(o.nodesDBStats),
        leafpreimagesdbstats: toDBStats(o.leafPreimagesDBStats),
        leafindicesdbstats: toDBStats(o.leafIndicesDBStats),
        blockindicesdbstats: toDBStats(o.blockIndicesDBStats),
    };
}
export function toWorldStateDBStats(o) {
    if (o.noteHashTreeStats === undefined) {
        throw new Error("Expected noteHashTreeStats in WorldStateDBStats deserialization");
    }
    if (o.messageTreeStats === undefined) {
        throw new Error("Expected messageTreeStats in WorldStateDBStats deserialization");
    }
    if (o.archiveTreeStats === undefined) {
        throw new Error("Expected archiveTreeStats in WorldStateDBStats deserialization");
    }
    if (o.publicDataTreeStats === undefined) {
        throw new Error("Expected publicDataTreeStats in WorldStateDBStats deserialization");
    }
    if (o.nullifierTreeStats === undefined) {
        throw new Error("Expected nullifierTreeStats in WorldStateDBStats deserialization");
    }
    ;
    return {
        notehashtreestats: toTreeDBStats(o.noteHashTreeStats),
        messagetreestats: toTreeDBStats(o.messageTreeStats),
        archivetreestats: toTreeDBStats(o.archiveTreeStats),
        publicdatatreestats: toTreeDBStats(o.publicDataTreeStats),
        nullifiertreestats: toTreeDBStats(o.nullifierTreeStats),
    };
}
export function toTreeMeta(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in TreeMeta deserialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in TreeMeta deserialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in TreeMeta deserialization");
    }
    if (o.committedSize === undefined) {
        throw new Error("Expected committedSize in TreeMeta deserialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in TreeMeta deserialization");
    }
    if (o.initialSize === undefined) {
        throw new Error("Expected initialSize in TreeMeta deserialization");
    }
    if (o.initialRoot === undefined) {
        throw new Error("Expected initialRoot in TreeMeta deserialization");
    }
    if (o.oldestHistoricBlock === undefined) {
        throw new Error("Expected oldestHistoricBlock in TreeMeta deserialization");
    }
    if (o.unfinalizedBlockHeight === undefined) {
        throw new Error("Expected unfinalizedBlockHeight in TreeMeta deserialization");
    }
    if (o.finalizedBlockHeight === undefined) {
        throw new Error("Expected finalizedBlockHeight in TreeMeta deserialization");
    }
    ;
    return {
        name: o.name,
        depth: o.depth,
        size: o.size,
        committedsize: o.committedSize,
        root: o.root,
        initialsize: o.initialSize,
        initialroot: o.initialRoot,
        oldesthistoricblock: o.oldestHistoricBlock,
        unfinalizedblockheight: o.unfinalizedBlockHeight,
        finalizedblockheight: o.finalizedBlockHeight,
    };
}
export function toWorldStateMeta(o) {
    if (o.noteHashTreeMeta === undefined) {
        throw new Error("Expected noteHashTreeMeta in WorldStateMeta deserialization");
    }
    if (o.messageTreeMeta === undefined) {
        throw new Error("Expected messageTreeMeta in WorldStateMeta deserialization");
    }
    if (o.archiveTreeMeta === undefined) {
        throw new Error("Expected archiveTreeMeta in WorldStateMeta deserialization");
    }
    if (o.publicDataTreeMeta === undefined) {
        throw new Error("Expected publicDataTreeMeta in WorldStateMeta deserialization");
    }
    if (o.nullifierTreeMeta === undefined) {
        throw new Error("Expected nullifierTreeMeta in WorldStateMeta deserialization");
    }
    ;
    return {
        notehashtreemeta: toTreeMeta(o.noteHashTreeMeta),
        messagetreemeta: toTreeMeta(o.messageTreeMeta),
        archivetreemeta: toTreeMeta(o.archiveTreeMeta),
        publicdatatreemeta: toTreeMeta(o.publicDataTreeMeta),
        nullifiertreemeta: toTreeMeta(o.nullifierTreeMeta),
    };
}
export function toWorldStateStatusFull(o) {
    if (o.summary === undefined) {
        throw new Error("Expected summary in WorldStateStatusFull deserialization");
    }
    if (o.dbStats === undefined) {
        throw new Error("Expected dbStats in WorldStateStatusFull deserialization");
    }
    if (o.meta === undefined) {
        throw new Error("Expected meta in WorldStateStatusFull deserialization");
    }
    ;
    return {
        summary: toWorldStateStatusSummary(o.summary),
        dbstats: toWorldStateDBStats(o.dbStats),
        meta: toWorldStateMeta(o.meta),
    };
}
export function toWorldStateRevision(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WorldStateRevision deserialization");
    }
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WorldStateRevision deserialization");
    }
    if (o.includeUncommitted === undefined) {
        throw new Error("Expected includeUncommitted in WorldStateRevision deserialization");
    }
    ;
    return {
        forkid: o.forkId,
        blocknumber: o.blockNumber,
        includeuncommitted: o.includeUncommitted,
    };
}
export function toWsdbGetTreeInfo(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetTreeInfo deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetTreeInfo deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
    };
}
export function toWsdbGetStateReference(o) {
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetStateReference deserialization");
    }
    ;
    return {
        revision: toWorldStateRevision(o.revision),
    };
}
export function toWsdbGetInitialStateReference(o) {
    return {};
}
export function toWsdbGetLeafValue(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetLeafValue deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafValue deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetLeafValue deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
export function toWsdbGetLeafPreimage(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetLeafPreimage deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafPreimage deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetLeafPreimage deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
export function toWsdbGetSiblingPath(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetSiblingPath deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetSiblingPath deserialization");
    }
    if (o.leafIndex === undefined) {
        throw new Error("Expected leafIndex in WsdbGetSiblingPath deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindex: o.leafIndex,
    };
}
export function toWsdbGetBlockNumbersForLeafIndices(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    if (o.leafIndices === undefined) {
        throw new Error("Expected leafIndices in WsdbGetBlockNumbersForLeafIndices deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leafindices: o.leafIndices,
    };
}
export function toWsdbFindLeafIndices(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindLeafIndices deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLeafIndices deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindLeafIndices deserialization");
    }
    if (o.startIndex === undefined) {
        throw new Error("Expected startIndex in WsdbFindLeafIndices deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leaves: o.leaves,
        startindex: o.startIndex,
    };
}
export function toWsdbFindLowLeaf(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindLowLeaf deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLowLeaf deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in WsdbFindLowLeaf deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        key: o.key,
    };
}
export function toWsdbFindSiblingPaths(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbFindSiblingPaths deserialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindSiblingPaths deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindSiblingPaths deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        revision: toWorldStateRevision(o.revision),
        leaves: o.leaves,
    };
}
export function toWsdbAppendLeaves(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbAppendLeaves deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbAppendLeaves deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbAppendLeaves deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        forkid: o.forkId,
    };
}
export function toWsdbBatchInsert(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbBatchInsert deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbBatchInsert deserialization");
    }
    if (o.subtreeDepth === undefined) {
        throw new Error("Expected subtreeDepth in WsdbBatchInsert deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbBatchInsert deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        subtreedepth: o.subtreeDepth,
        forkid: o.forkId,
    };
}
export function toWsdbSequentialInsert(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbSequentialInsert deserialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbSequentialInsert deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbSequentialInsert deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        leaves: o.leaves,
        forkid: o.forkId,
    };
}
export function toWsdbUpdateArchive(o) {
    if (o.blockStateRef === undefined) {
        throw new Error("Expected blockStateRef in WsdbUpdateArchive deserialization");
    }
    if (o.blockHeaderHash === undefined) {
        throw new Error("Expected blockHeaderHash in WsdbUpdateArchive deserialization");
    }
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbUpdateArchive deserialization");
    }
    ;
    return {
        blockstateref: o.blockStateRef,
        blockheaderhash: o.blockHeaderHash,
        forkid: o.forkId,
    };
}
export function toWsdbCommit(o) {
    return {};
}
export function toWsdbRollback(o) {
    return {};
}
export function toNullifierLeafValue(o) {
    if (o.nullifier === undefined) {
        throw new Error("Expected nullifier in NullifierLeafValue deserialization");
    }
    ;
    return {
        nullifier: o.nullifier,
    };
}
export function toPublicDataLeafValue(o) {
    if (o.slot === undefined) {
        throw new Error("Expected slot in PublicDataLeafValue deserialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in PublicDataLeafValue deserialization");
    }
    ;
    return {
        slot: o.slot,
        value: o.value,
    };
}
export function toWsdbSyncBlock(o) {
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WsdbSyncBlock deserialization");
    }
    if (o.blockStateRef === undefined) {
        throw new Error("Expected blockStateRef in WsdbSyncBlock deserialization");
    }
    if (o.blockHeaderHash === undefined) {
        throw new Error("Expected blockHeaderHash in WsdbSyncBlock deserialization");
    }
    if (o.paddedNoteHashes === undefined) {
        throw new Error("Expected paddedNoteHashes in WsdbSyncBlock deserialization");
    }
    if (o.paddedL1ToL2Messages === undefined) {
        throw new Error("Expected paddedL1ToL2Messages in WsdbSyncBlock deserialization");
    }
    if (o.paddedNullifiers === undefined) {
        throw new Error("Expected paddedNullifiers in WsdbSyncBlock deserialization");
    }
    if (o.publicDataWrites === undefined) {
        throw new Error("Expected publicDataWrites in WsdbSyncBlock deserialization");
    }
    ;
    return {
        blocknumber: o.blockNumber,
        blockstateref: o.blockStateRef,
        blockheaderhash: o.blockHeaderHash,
        paddednotehashes: o.paddedNoteHashes,
        paddedl1tol2messages: o.paddedL1ToL2Messages,
        paddednullifiers: o.paddedNullifiers.map((v) => toNullifierLeafValue(v)),
        publicdatawrites: o.publicDataWrites.map((v) => toPublicDataLeafValue(v)),
    };
}
export function toWsdbCreateFork(o) {
    if (o.latest === undefined) {
        throw new Error("Expected latest in WsdbCreateFork deserialization");
    }
    if (o.blockNumber === undefined) {
        throw new Error("Expected blockNumber in WsdbCreateFork deserialization");
    }
    ;
    return {
        latest: o.latest,
        blocknumber: o.blockNumber,
    };
}
export function toWsdbDeleteFork(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbDeleteFork deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbFinalizeBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbFinalizeBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
export function toWsdbUnwindBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbUnwindBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
export function toWsdbRemoveHistoricalBlocks(o) {
    if (o.toBlockNumber === undefined) {
        throw new Error("Expected toBlockNumber in WsdbRemoveHistoricalBlocks deserialization");
    }
    ;
    return {
        toblocknumber: o.toBlockNumber,
    };
}
export function toWsdbGetStatus(o) {
    return {};
}
export function toWsdbCreateCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCreateCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbCommitCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCommitCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbRevertCheckpoint(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbRevertCheckpoint deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbCommitAllCheckpoints(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCommitAllCheckpoints deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbRevertAllCheckpoints(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbRevertAllCheckpoints deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbCopyStores(o) {
    if (o.dstPath === undefined) {
        throw new Error("Expected dstPath in WsdbCopyStores deserialization");
    }
    if (o.compact === undefined) {
        throw new Error("Expected compact in WsdbCopyStores deserialization");
    }
    ;
    return {
        dstpath: o.dstPath,
        compact: o.compact,
    };
}
export function toWsdbShutdown(o) {
    return {};
}
export function toWsdbErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in WsdbErrorResponse deserialization");
    }
    ;
    return {
        message: o.message,
    };
}
export function toWsdbGetTreeInfoResponse(o) {
    if (o.treeId === undefined) {
        throw new Error("Expected treeId in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in WsdbGetTreeInfoResponse deserialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in WsdbGetTreeInfoResponse deserialization");
    }
    ;
    return {
        treeid: o.treeId,
        root: o.root,
        size: o.size,
        depth: o.depth,
    };
}
export function toWsdbGetStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetStateReferenceResponse deserialization");
    }
    ;
    return {
        state: o.state,
    };
}
export function toWsdbGetInitialStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetInitialStateReferenceResponse deserialization");
    }
    ;
    return {
        state: o.state,
    };
}
export function toWsdbGetLeafValueResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in WsdbGetLeafValueResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function toWsdbGetLeafPreimageResponse(o) {
    if (o.preimage === undefined) {
        throw new Error("Expected preimage in WsdbGetLeafPreimageResponse deserialization");
    }
    ;
    return {
        preimage: o.preimage,
    };
}
export function toWsdbGetSiblingPathResponse(o) {
    if (o.path === undefined) {
        throw new Error("Expected path in WsdbGetSiblingPathResponse deserialization");
    }
    ;
    return {
        path: o.path,
    };
}
export function toWsdbGetBlockNumbersForLeafIndicesResponse(o) {
    if (o.blockNumbers === undefined) {
        throw new Error("Expected blockNumbers in WsdbGetBlockNumbersForLeafIndicesResponse deserialization");
    }
    ;
    return {
        blocknumbers: o.blockNumbers,
    };
}
export function toWsdbFindLeafIndicesResponse(o) {
    if (o.indices === undefined) {
        throw new Error("Expected indices in WsdbFindLeafIndicesResponse deserialization");
    }
    ;
    return {
        indices: o.indices,
    };
}
export function toWsdbFindLowLeafResponse(o) {
    if (o.alreadyPresent === undefined) {
        throw new Error("Expected alreadyPresent in WsdbFindLowLeafResponse deserialization");
    }
    if (o.index === undefined) {
        throw new Error("Expected index in WsdbFindLowLeafResponse deserialization");
    }
    ;
    return {
        alreadypresent: o.alreadyPresent,
        index: o.index,
    };
}
export function toWsdbFindSiblingPathsResponse(o) {
    if (o.paths === undefined) {
        throw new Error("Expected paths in WsdbFindSiblingPathsResponse deserialization");
    }
    ;
    return {
        paths: o.paths.map((v) => v != null ? toSiblingPathAndIndex(v) : undefined),
    };
}
export function toWsdbAppendLeavesResponse(o) {
    return {};
}
export function toWsdbBatchInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbBatchInsertResponse deserialization");
    }
    ;
    return {
        result: o.result,
    };
}
export function toWsdbSequentialInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbSequentialInsertResponse deserialization");
    }
    ;
    return {
        result: o.result,
    };
}
export function toWsdbUpdateArchiveResponse(o) {
    return {};
}
export function toWsdbCommitResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbCommitResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
export function toWsdbRollbackResponse(o) {
    return {};
}
export function toWsdbSyncBlockResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbSyncBlockResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
export function toWsdbCreateForkResponse(o) {
    if (o.forkId === undefined) {
        throw new Error("Expected forkId in WsdbCreateForkResponse deserialization");
    }
    ;
    return {
        forkid: o.forkId,
    };
}
export function toWsdbDeleteForkResponse(o) {
    return {};
}
export function toWsdbFinalizeBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbFinalizeBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusSummary(o.status),
    };
}
export function toWsdbUnwindBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbUnwindBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
export function toWsdbRemoveHistoricalBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusFull(o.status),
    };
}
export function toWsdbGetStatusResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbGetStatusResponse deserialization");
    }
    ;
    return {
        status: toWorldStateStatusSummary(o.status),
    };
}
export function toWsdbCreateCheckpointResponse(o) {
    return {};
}
export function toWsdbCommitCheckpointResponse(o) {
    return {};
}
export function toWsdbRevertCheckpointResponse(o) {
    return {};
}
export function toWsdbCommitAllCheckpointsResponse(o) {
    return {};
}
export function toWsdbRevertAllCheckpointsResponse(o) {
    return {};
}
export function toWsdbCopyStoresResponse(o) {
    return {};
}
export function toWsdbShutdownResponse(o) {
    return {};
}
export function fromSiblingPathAndIndex(o) {
    if (o.index === undefined) {
        throw new Error("Expected index in SiblingPathAndIndex serialization");
    }
    if (o.path === undefined) {
        throw new Error("Expected path in SiblingPathAndIndex serialization");
    }
    ;
    return {
        index: o.index,
        path: o.path,
    };
}
export function fromWorldStateStatusSummary(o) {
    if (o.unfinalizedblocknumber === undefined) {
        throw new Error("Expected unfinalizedblocknumber in WorldStateStatusSummary serialization");
    }
    if (o.finalizedblocknumber === undefined) {
        throw new Error("Expected finalizedblocknumber in WorldStateStatusSummary serialization");
    }
    if (o.oldesthistoricalblock === undefined) {
        throw new Error("Expected oldesthistoricalblock in WorldStateStatusSummary serialization");
    }
    if (o.treesaresynched === undefined) {
        throw new Error("Expected treesaresynched in WorldStateStatusSummary serialization");
    }
    ;
    return {
        unfinalizedBlockNumber: o.unfinalizedblocknumber,
        finalizedBlockNumber: o.finalizedblocknumber,
        oldestHistoricalBlock: o.oldesthistoricalblock,
        treesAreSynched: o.treesaresynched,
    };
}
export function fromDBStats(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in DBStats serialization");
    }
    if (o.numdataitems === undefined) {
        throw new Error("Expected numdataitems in DBStats serialization");
    }
    if (o.totalusedsize === undefined) {
        throw new Error("Expected totalusedsize in DBStats serialization");
    }
    ;
    return {
        name: o.name,
        numDataItems: o.numdataitems,
        totalUsedSize: o.totalusedsize,
    };
}
export function fromTreeDBStats(o) {
    if (o.mapsize === undefined) {
        throw new Error("Expected mapsize in TreeDBStats serialization");
    }
    if (o.physicalfilesize === undefined) {
        throw new Error("Expected physicalfilesize in TreeDBStats serialization");
    }
    if (o.blocksdbstats === undefined) {
        throw new Error("Expected blocksdbstats in TreeDBStats serialization");
    }
    if (o.nodesdbstats === undefined) {
        throw new Error("Expected nodesdbstats in TreeDBStats serialization");
    }
    if (o.leafpreimagesdbstats === undefined) {
        throw new Error("Expected leafpreimagesdbstats in TreeDBStats serialization");
    }
    if (o.leafindicesdbstats === undefined) {
        throw new Error("Expected leafindicesdbstats in TreeDBStats serialization");
    }
    if (o.blockindicesdbstats === undefined) {
        throw new Error("Expected blockindicesdbstats in TreeDBStats serialization");
    }
    ;
    return {
        mapSize: o.mapsize,
        physicalFileSize: o.physicalfilesize,
        blocksDBStats: fromDBStats(o.blocksdbstats),
        nodesDBStats: fromDBStats(o.nodesdbstats),
        leafPreimagesDBStats: fromDBStats(o.leafpreimagesdbstats),
        leafIndicesDBStats: fromDBStats(o.leafindicesdbstats),
        blockIndicesDBStats: fromDBStats(o.blockindicesdbstats),
    };
}
export function fromWorldStateDBStats(o) {
    if (o.notehashtreestats === undefined) {
        throw new Error("Expected notehashtreestats in WorldStateDBStats serialization");
    }
    if (o.messagetreestats === undefined) {
        throw new Error("Expected messagetreestats in WorldStateDBStats serialization");
    }
    if (o.archivetreestats === undefined) {
        throw new Error("Expected archivetreestats in WorldStateDBStats serialization");
    }
    if (o.publicdatatreestats === undefined) {
        throw new Error("Expected publicdatatreestats in WorldStateDBStats serialization");
    }
    if (o.nullifiertreestats === undefined) {
        throw new Error("Expected nullifiertreestats in WorldStateDBStats serialization");
    }
    ;
    return {
        noteHashTreeStats: fromTreeDBStats(o.notehashtreestats),
        messageTreeStats: fromTreeDBStats(o.messagetreestats),
        archiveTreeStats: fromTreeDBStats(o.archivetreestats),
        publicDataTreeStats: fromTreeDBStats(o.publicdatatreestats),
        nullifierTreeStats: fromTreeDBStats(o.nullifiertreestats),
    };
}
export function fromTreeMeta(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in TreeMeta serialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in TreeMeta serialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in TreeMeta serialization");
    }
    if (o.committedsize === undefined) {
        throw new Error("Expected committedsize in TreeMeta serialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in TreeMeta serialization");
    }
    if (o.initialsize === undefined) {
        throw new Error("Expected initialsize in TreeMeta serialization");
    }
    if (o.initialroot === undefined) {
        throw new Error("Expected initialroot in TreeMeta serialization");
    }
    if (o.oldesthistoricblock === undefined) {
        throw new Error("Expected oldesthistoricblock in TreeMeta serialization");
    }
    if (o.unfinalizedblockheight === undefined) {
        throw new Error("Expected unfinalizedblockheight in TreeMeta serialization");
    }
    if (o.finalizedblockheight === undefined) {
        throw new Error("Expected finalizedblockheight in TreeMeta serialization");
    }
    ;
    return {
        name: o.name,
        depth: o.depth,
        size: o.size,
        committedSize: o.committedsize,
        root: o.root,
        initialSize: o.initialsize,
        initialRoot: o.initialroot,
        oldestHistoricBlock: o.oldesthistoricblock,
        unfinalizedBlockHeight: o.unfinalizedblockheight,
        finalizedBlockHeight: o.finalizedblockheight,
    };
}
export function fromWorldStateMeta(o) {
    if (o.notehashtreemeta === undefined) {
        throw new Error("Expected notehashtreemeta in WorldStateMeta serialization");
    }
    if (o.messagetreemeta === undefined) {
        throw new Error("Expected messagetreemeta in WorldStateMeta serialization");
    }
    if (o.archivetreemeta === undefined) {
        throw new Error("Expected archivetreemeta in WorldStateMeta serialization");
    }
    if (o.publicdatatreemeta === undefined) {
        throw new Error("Expected publicdatatreemeta in WorldStateMeta serialization");
    }
    if (o.nullifiertreemeta === undefined) {
        throw new Error("Expected nullifiertreemeta in WorldStateMeta serialization");
    }
    ;
    return {
        noteHashTreeMeta: fromTreeMeta(o.notehashtreemeta),
        messageTreeMeta: fromTreeMeta(o.messagetreemeta),
        archiveTreeMeta: fromTreeMeta(o.archivetreemeta),
        publicDataTreeMeta: fromTreeMeta(o.publicdatatreemeta),
        nullifierTreeMeta: fromTreeMeta(o.nullifiertreemeta),
    };
}
export function fromWorldStateStatusFull(o) {
    if (o.summary === undefined) {
        throw new Error("Expected summary in WorldStateStatusFull serialization");
    }
    if (o.dbstats === undefined) {
        throw new Error("Expected dbstats in WorldStateStatusFull serialization");
    }
    if (o.meta === undefined) {
        throw new Error("Expected meta in WorldStateStatusFull serialization");
    }
    ;
    return {
        summary: fromWorldStateStatusSummary(o.summary),
        dbStats: fromWorldStateDBStats(o.dbstats),
        meta: fromWorldStateMeta(o.meta),
    };
}
export function fromWorldStateRevision(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WorldStateRevision serialization");
    }
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WorldStateRevision serialization");
    }
    if (o.includeuncommitted === undefined) {
        throw new Error("Expected includeuncommitted in WorldStateRevision serialization");
    }
    ;
    return {
        forkId: o.forkid,
        blockNumber: o.blocknumber,
        includeUncommitted: o.includeuncommitted,
    };
}
export function fromWsdbGetTreeInfo(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetTreeInfo serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetTreeInfo serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
    };
}
export function fromWsdbGetStateReference(o) {
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetStateReference serialization");
    }
    ;
    return {
        revision: fromWorldStateRevision(o.revision),
    };
}
export function fromWsdbGetInitialStateReference(o) {
    return {};
}
export function fromWsdbGetLeafValue(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetLeafValue serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafValue serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetLeafValue serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
export function fromWsdbGetLeafPreimage(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetLeafPreimage serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetLeafPreimage serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetLeafPreimage serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
export function fromWsdbGetSiblingPath(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetSiblingPath serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetSiblingPath serialization");
    }
    if (o.leafindex === undefined) {
        throw new Error("Expected leafindex in WsdbGetSiblingPath serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndex: o.leafindex,
    };
}
export function fromWsdbGetBlockNumbersForLeafIndices(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    if (o.leafindices === undefined) {
        throw new Error("Expected leafindices in WsdbGetBlockNumbersForLeafIndices serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leafIndices: o.leafindices,
    };
}
export function fromWsdbFindLeafIndices(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindLeafIndices serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLeafIndices serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindLeafIndices serialization");
    }
    if (o.startindex === undefined) {
        throw new Error("Expected startindex in WsdbFindLeafIndices serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leaves: o.leaves,
        startIndex: o.startindex,
    };
}
export function fromWsdbFindLowLeaf(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindLowLeaf serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindLowLeaf serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in WsdbFindLowLeaf serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        key: o.key,
    };
}
export function fromWsdbFindSiblingPaths(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbFindSiblingPaths serialization");
    }
    if (o.revision === undefined) {
        throw new Error("Expected revision in WsdbFindSiblingPaths serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbFindSiblingPaths serialization");
    }
    ;
    return {
        treeId: o.treeid,
        revision: fromWorldStateRevision(o.revision),
        leaves: o.leaves,
    };
}
export function fromWsdbAppendLeaves(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbAppendLeaves serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbAppendLeaves serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbAppendLeaves serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        forkId: o.forkid,
    };
}
export function fromWsdbBatchInsert(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbBatchInsert serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbBatchInsert serialization");
    }
    if (o.subtreedepth === undefined) {
        throw new Error("Expected subtreedepth in WsdbBatchInsert serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbBatchInsert serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        subtreeDepth: o.subtreedepth,
        forkId: o.forkid,
    };
}
export function fromWsdbSequentialInsert(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbSequentialInsert serialization");
    }
    if (o.leaves === undefined) {
        throw new Error("Expected leaves in WsdbSequentialInsert serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbSequentialInsert serialization");
    }
    ;
    return {
        treeId: o.treeid,
        leaves: o.leaves,
        forkId: o.forkid,
    };
}
export function fromWsdbUpdateArchive(o) {
    if (o.blockstateref === undefined) {
        throw new Error("Expected blockstateref in WsdbUpdateArchive serialization");
    }
    if (o.blockheaderhash === undefined) {
        throw new Error("Expected blockheaderhash in WsdbUpdateArchive serialization");
    }
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbUpdateArchive serialization");
    }
    ;
    return {
        blockStateRef: o.blockstateref,
        blockHeaderHash: o.blockheaderhash,
        forkId: o.forkid,
    };
}
export function fromWsdbCommit(o) {
    return {};
}
export function fromWsdbRollback(o) {
    return {};
}
export function fromNullifierLeafValue(o) {
    if (o.nullifier === undefined) {
        throw new Error("Expected nullifier in NullifierLeafValue serialization");
    }
    ;
    return {
        nullifier: o.nullifier,
    };
}
export function fromPublicDataLeafValue(o) {
    if (o.slot === undefined) {
        throw new Error("Expected slot in PublicDataLeafValue serialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in PublicDataLeafValue serialization");
    }
    ;
    return {
        slot: o.slot,
        value: o.value,
    };
}
export function fromWsdbSyncBlock(o) {
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WsdbSyncBlock serialization");
    }
    if (o.blockstateref === undefined) {
        throw new Error("Expected blockstateref in WsdbSyncBlock serialization");
    }
    if (o.blockheaderhash === undefined) {
        throw new Error("Expected blockheaderhash in WsdbSyncBlock serialization");
    }
    if (o.paddednotehashes === undefined) {
        throw new Error("Expected paddednotehashes in WsdbSyncBlock serialization");
    }
    if (o.paddedl1tol2messages === undefined) {
        throw new Error("Expected paddedl1tol2messages in WsdbSyncBlock serialization");
    }
    if (o.paddednullifiers === undefined) {
        throw new Error("Expected paddednullifiers in WsdbSyncBlock serialization");
    }
    if (o.publicdatawrites === undefined) {
        throw new Error("Expected publicdatawrites in WsdbSyncBlock serialization");
    }
    ;
    return {
        blockNumber: o.blocknumber,
        blockStateRef: o.blockstateref,
        blockHeaderHash: o.blockheaderhash,
        paddedNoteHashes: o.paddednotehashes,
        paddedL1ToL2Messages: o.paddedl1tol2messages,
        paddedNullifiers: o.paddednullifiers.map((v) => fromNullifierLeafValue(v)),
        publicDataWrites: o.publicdatawrites.map((v) => fromPublicDataLeafValue(v)),
    };
}
export function fromWsdbCreateFork(o) {
    if (o.latest === undefined) {
        throw new Error("Expected latest in WsdbCreateFork serialization");
    }
    if (o.blocknumber === undefined) {
        throw new Error("Expected blocknumber in WsdbCreateFork serialization");
    }
    ;
    return {
        latest: o.latest,
        blockNumber: o.blocknumber,
    };
}
export function fromWsdbDeleteFork(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbDeleteFork serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbFinalizeBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbFinalizeBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
export function fromWsdbUnwindBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbUnwindBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
export function fromWsdbRemoveHistoricalBlocks(o) {
    if (o.toblocknumber === undefined) {
        throw new Error("Expected toblocknumber in WsdbRemoveHistoricalBlocks serialization");
    }
    ;
    return {
        toBlockNumber: o.toblocknumber,
    };
}
export function fromWsdbGetStatus(o) {
    return {};
}
export function fromWsdbCreateCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCreateCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbCommitCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCommitCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbRevertCheckpoint(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbRevertCheckpoint serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbCommitAllCheckpoints(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCommitAllCheckpoints serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbRevertAllCheckpoints(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbRevertAllCheckpoints serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbCopyStores(o) {
    if (o.dstpath === undefined) {
        throw new Error("Expected dstpath in WsdbCopyStores serialization");
    }
    if (o.compact === undefined) {
        throw new Error("Expected compact in WsdbCopyStores serialization");
    }
    ;
    return {
        dstPath: o.dstpath,
        compact: o.compact,
    };
}
export function fromWsdbShutdown(o) {
    return {};
}
export function fromWsdbErrorResponse(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in WsdbErrorResponse serialization");
    }
    ;
    return {
        message: o.message,
    };
}
export function fromWsdbGetTreeInfoResponse(o) {
    if (o.treeid === undefined) {
        throw new Error("Expected treeid in WsdbGetTreeInfoResponse serialization");
    }
    if (o.root === undefined) {
        throw new Error("Expected root in WsdbGetTreeInfoResponse serialization");
    }
    if (o.size === undefined) {
        throw new Error("Expected size in WsdbGetTreeInfoResponse serialization");
    }
    if (o.depth === undefined) {
        throw new Error("Expected depth in WsdbGetTreeInfoResponse serialization");
    }
    ;
    return {
        treeId: o.treeid,
        root: o.root,
        size: o.size,
        depth: o.depth,
    };
}
export function fromWsdbGetStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetStateReferenceResponse serialization");
    }
    ;
    return {
        state: o.state,
    };
}
export function fromWsdbGetInitialStateReferenceResponse(o) {
    if (o.state === undefined) {
        throw new Error("Expected state in WsdbGetInitialStateReferenceResponse serialization");
    }
    ;
    return {
        state: o.state,
    };
}
export function fromWsdbGetLeafValueResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in WsdbGetLeafValueResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
export function fromWsdbGetLeafPreimageResponse(o) {
    if (o.preimage === undefined) {
        throw new Error("Expected preimage in WsdbGetLeafPreimageResponse serialization");
    }
    ;
    return {
        preimage: o.preimage,
    };
}
export function fromWsdbGetSiblingPathResponse(o) {
    if (o.path === undefined) {
        throw new Error("Expected path in WsdbGetSiblingPathResponse serialization");
    }
    ;
    return {
        path: o.path,
    };
}
export function fromWsdbGetBlockNumbersForLeafIndicesResponse(o) {
    if (o.blocknumbers === undefined) {
        throw new Error("Expected blocknumbers in WsdbGetBlockNumbersForLeafIndicesResponse serialization");
    }
    ;
    return {
        blockNumbers: o.blocknumbers,
    };
}
export function fromWsdbFindLeafIndicesResponse(o) {
    if (o.indices === undefined) {
        throw new Error("Expected indices in WsdbFindLeafIndicesResponse serialization");
    }
    ;
    return {
        indices: o.indices,
    };
}
export function fromWsdbFindLowLeafResponse(o) {
    if (o.alreadypresent === undefined) {
        throw new Error("Expected alreadypresent in WsdbFindLowLeafResponse serialization");
    }
    if (o.index === undefined) {
        throw new Error("Expected index in WsdbFindLowLeafResponse serialization");
    }
    ;
    return {
        alreadyPresent: o.alreadypresent,
        index: o.index,
    };
}
export function fromWsdbFindSiblingPathsResponse(o) {
    if (o.paths === undefined) {
        throw new Error("Expected paths in WsdbFindSiblingPathsResponse serialization");
    }
    ;
    return {
        paths: o.paths.map((v) => v != null ? fromSiblingPathAndIndex(v) : undefined),
    };
}
export function fromWsdbAppendLeavesResponse(o) {
    return {};
}
export function fromWsdbBatchInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbBatchInsertResponse serialization");
    }
    ;
    return {
        result: o.result,
    };
}
export function fromWsdbSequentialInsertResponse(o) {
    if (o.result === undefined) {
        throw new Error("Expected result in WsdbSequentialInsertResponse serialization");
    }
    ;
    return {
        result: o.result,
    };
}
export function fromWsdbUpdateArchiveResponse(o) {
    return {};
}
export function fromWsdbCommitResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbCommitResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
export function fromWsdbRollbackResponse(o) {
    return {};
}
export function fromWsdbSyncBlockResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbSyncBlockResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
export function fromWsdbCreateForkResponse(o) {
    if (o.forkid === undefined) {
        throw new Error("Expected forkid in WsdbCreateForkResponse serialization");
    }
    ;
    return {
        forkId: o.forkid,
    };
}
export function fromWsdbDeleteForkResponse(o) {
    return {};
}
export function fromWsdbFinalizeBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbFinalizeBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusSummary(o.status),
    };
}
export function fromWsdbUnwindBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbUnwindBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
export function fromWsdbRemoveHistoricalBlocksResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusFull(o.status),
    };
}
export function fromWsdbGetStatusResponse(o) {
    if (o.status === undefined) {
        throw new Error("Expected status in WsdbGetStatusResponse serialization");
    }
    ;
    return {
        status: fromWorldStateStatusSummary(o.status),
    };
}
export function fromWsdbCreateCheckpointResponse(o) {
    return {};
}
export function fromWsdbCommitCheckpointResponse(o) {
    return {};
}
export function fromWsdbRevertCheckpointResponse(o) {
    return {};
}
export function fromWsdbCommitAllCheckpointsResponse(o) {
    return {};
}
export function fromWsdbRevertAllCheckpointsResponse(o) {
    return {};
}
export function fromWsdbCopyStoresResponse(o) {
    return {};
}
export function fromWsdbShutdownResponse(o) {
    return {};
}
