// @aztec/bb.js/src/aztec-wsdb/generated/api_types.ts
// AUTOGENERATED FILE - DO NOT EDIT

// Type aliases for primitive types
export type Field2 = [Uint8Array, Uint8Array];

// Public interfaces (exported)
export interface SiblingPathAndIndex {
  index: number;
  path: Uint8Array[];
}

export interface WorldStateStatusSummary {
  unfinalizedblocknumber: number;
  finalizedblocknumber: number;
  oldesthistoricalblock: number;
  treesaresynched: boolean;
}

export interface DBStats {
  name: string;
  numdataitems: number;
  totalusedsize: number;
}

export interface TreeDBStats {
  mapsize: number;
  physicalfilesize: number;
  blocksdbstats: DBStats;
  nodesdbstats: DBStats;
  leafpreimagesdbstats: DBStats;
  leafindicesdbstats: DBStats;
  blockindicesdbstats: DBStats;
}

export interface WorldStateDBStats {
  notehashtreestats: TreeDBStats;
  messagetreestats: TreeDBStats;
  archivetreestats: TreeDBStats;
  publicdatatreestats: TreeDBStats;
  nullifiertreestats: TreeDBStats;
}

export interface TreeMeta {
  name: string;
  depth: number;
  size: number;
  committedsize: number;
  root: Uint8Array;
  initialsize: number;
  initialroot: Uint8Array;
  oldesthistoricblock: number;
  unfinalizedblockheight: number;
  finalizedblockheight: number;
}

export interface WorldStateMeta {
  notehashtreemeta: TreeMeta;
  messagetreemeta: TreeMeta;
  archivetreemeta: TreeMeta;
  publicdatatreemeta: TreeMeta;
  nullifiertreemeta: TreeMeta;
}

export interface WorldStateStatusFull {
  summary: WorldStateStatusSummary;
  dbstats: WorldStateDBStats;
  meta: WorldStateMeta;
}

export interface WorldStateRevision {
  forkid: number;
  blocknumber: number;
  includeuncommitted: boolean;
}

export interface WsdbGetTreeInfo {
  treeid: number;
  revision: WorldStateRevision;
}

export interface WsdbGetStateReference {
  revision: WorldStateRevision;
}

export interface WsdbGetInitialStateReference {

}

export interface WsdbGetLeafValue {
  treeid: number;
  revision: WorldStateRevision;
  leafindex: number;
}

export interface WsdbGetLeafPreimage {
  treeid: number;
  revision: WorldStateRevision;
  leafindex: number;
}

export interface WsdbGetSiblingPath {
  treeid: number;
  revision: WorldStateRevision;
  leafindex: number;
}

export interface WsdbGetBlockNumbersForLeafIndices {
  treeid: number;
  revision: WorldStateRevision;
  leafindices: number[];
}

export interface WsdbFindLeafIndices {
  treeid: number;
  revision: WorldStateRevision;
  leaves: Uint8Array[];
  startindex: number;
}

export interface WsdbFindLowLeaf {
  treeid: number;
  revision: WorldStateRevision;
  key: Uint8Array;
}

export interface WsdbFindSiblingPaths {
  treeid: number;
  revision: WorldStateRevision;
  leaves: Uint8Array[];
}

export interface WsdbAppendLeaves {
  treeid: number;
  leaves: Uint8Array[];
  forkid: number;
}

export interface WsdbBatchInsert {
  treeid: number;
  leaves: Uint8Array[];
  subtreedepth: number;
  forkid: number;
}

export interface WsdbSequentialInsert {
  treeid: number;
  leaves: Uint8Array[];
  forkid: number;
}

export interface WsdbUpdateArchive {
  blockstateref: Record<number, [Uint8Array, number]>;
  blockheaderhash: Uint8Array;
  forkid: number;
}

export interface WsdbCommit {

}

export interface WsdbRollback {

}

export interface NullifierLeafValue {
  nullifier: Uint8Array;
}

export interface PublicDataLeafValue {
  slot: Uint8Array;
  value: Uint8Array;
}

export interface WsdbSyncBlock {
  blocknumber: number;
  blockstateref: Record<number, [Uint8Array, number]>;
  blockheaderhash: Uint8Array;
  paddednotehashes: Uint8Array[];
  paddedl1tol2messages: Uint8Array[];
  paddednullifiers: NullifierLeafValue[];
  publicdatawrites: PublicDataLeafValue[];
}

export interface WsdbCreateFork {
  latest: boolean;
  blocknumber: number;
}

export interface WsdbDeleteFork {
  forkid: number;
}

export interface WsdbFinalizeBlocks {
  toblocknumber: number;
}

export interface WsdbUnwindBlocks {
  toblocknumber: number;
}

export interface WsdbRemoveHistoricalBlocks {
  toblocknumber: number;
}

export interface WsdbGetStatus {

}

export interface WsdbCreateCheckpoint {
  forkid: number;
}

export interface WsdbCommitCheckpoint {
  forkid: number;
}

export interface WsdbRevertCheckpoint {
  forkid: number;
}

export interface WsdbCommitAllCheckpoints {
  forkid: number;
}

export interface WsdbRevertAllCheckpoints {
  forkid: number;
}

export interface WsdbCopyStores {
  dstpath: string;
  compact: boolean | undefined;
}

export interface WsdbShutdown {

}

export interface WsdbErrorResponse {
  message: string;
}

export interface WsdbGetTreeInfoResponse {
  treeid: number;
  root: Uint8Array;
  size: number;
  depth: number;
}

export interface WsdbGetStateReferenceResponse {
  state: Record<number, [Uint8Array, number]>;
}

export interface WsdbGetInitialStateReferenceResponse {
  state: Record<number, [Uint8Array, number]>;
}

export interface WsdbGetLeafValueResponse {
  value: Uint8Array | undefined;
}

export interface WsdbGetLeafPreimageResponse {
  preimage: Uint8Array | undefined;
}

export interface WsdbGetSiblingPathResponse {
  path: Uint8Array[];
}

export interface WsdbGetBlockNumbersForLeafIndicesResponse {
  blocknumbers: (number | undefined)[];
}

export interface WsdbFindLeafIndicesResponse {
  indices: (number | undefined)[];
}

export interface WsdbFindLowLeafResponse {
  alreadypresent: boolean;
  index: number;
}

export interface WsdbFindSiblingPathsResponse {
  paths: (SiblingPathAndIndex | undefined)[];
}

export interface WsdbAppendLeavesResponse {

}

export interface WsdbBatchInsertResponse {
  result: Uint8Array;
}

export interface WsdbSequentialInsertResponse {
  result: Uint8Array;
}

export interface WsdbUpdateArchiveResponse {

}

export interface WsdbCommitResponse {
  status: WorldStateStatusFull;
}

export interface WsdbRollbackResponse {

}

export interface WsdbSyncBlockResponse {
  status: WorldStateStatusFull;
}

export interface WsdbCreateForkResponse {
  forkid: number;
}

export interface WsdbDeleteForkResponse {

}

export interface WsdbFinalizeBlocksResponse {
  status: WorldStateStatusSummary;
}

export interface WsdbUnwindBlocksResponse {
  status: WorldStateStatusFull;
}

export interface WsdbRemoveHistoricalBlocksResponse {
  status: WorldStateStatusFull;
}

export interface WsdbGetStatusResponse {
  status: WorldStateStatusSummary;
}

export interface WsdbCreateCheckpointResponse {

}

export interface WsdbCommitCheckpointResponse {

}

export interface WsdbRevertCheckpointResponse {

}

export interface WsdbCommitAllCheckpointsResponse {

}

export interface WsdbRevertAllCheckpointsResponse {

}

export interface WsdbCopyStoresResponse {

}

export interface WsdbShutdownResponse {

}

// Private Msgpack interfaces (not exported)
interface MsgpackSiblingPathAndIndex {
  index: number;
  path: Uint8Array[];
}

interface MsgpackWorldStateStatusSummary {
  unfinalizedBlockNumber: number;
  finalizedBlockNumber: number;
  oldestHistoricalBlock: number;
  treesAreSynched: boolean;
}

interface MsgpackDBStats {
  name: string;
  numDataItems: number;
  totalUsedSize: number;
}

interface MsgpackTreeDBStats {
  mapSize: number;
  physicalFileSize: number;
  blocksDBStats: MsgpackDBStats;
  nodesDBStats: MsgpackDBStats;
  leafPreimagesDBStats: MsgpackDBStats;
  leafIndicesDBStats: MsgpackDBStats;
  blockIndicesDBStats: MsgpackDBStats;
}

interface MsgpackWorldStateDBStats {
  noteHashTreeStats: MsgpackTreeDBStats;
  messageTreeStats: MsgpackTreeDBStats;
  archiveTreeStats: MsgpackTreeDBStats;
  publicDataTreeStats: MsgpackTreeDBStats;
  nullifierTreeStats: MsgpackTreeDBStats;
}

interface MsgpackTreeMeta {
  name: string;
  depth: number;
  size: number;
  committedSize: number;
  root: Uint8Array;
  initialSize: number;
  initialRoot: Uint8Array;
  oldestHistoricBlock: number;
  unfinalizedBlockHeight: number;
  finalizedBlockHeight: number;
}

interface MsgpackWorldStateMeta {
  noteHashTreeMeta: MsgpackTreeMeta;
  messageTreeMeta: MsgpackTreeMeta;
  archiveTreeMeta: MsgpackTreeMeta;
  publicDataTreeMeta: MsgpackTreeMeta;
  nullifierTreeMeta: MsgpackTreeMeta;
}

interface MsgpackWorldStateStatusFull {
  summary: MsgpackWorldStateStatusSummary;
  dbStats: MsgpackWorldStateDBStats;
  meta: MsgpackWorldStateMeta;
}

interface MsgpackWorldStateRevision {
  forkId: number;
  blockNumber: number;
  includeUncommitted: boolean;
}

interface MsgpackWsdbGetTreeInfo {
  treeId: number;
  revision: MsgpackWorldStateRevision;
}

interface MsgpackWsdbGetStateReference {
  revision: MsgpackWorldStateRevision;
}

interface MsgpackWsdbGetInitialStateReference {

}

interface MsgpackWsdbGetLeafValue {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leafIndex: number;
}

interface MsgpackWsdbGetLeafPreimage {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leafIndex: number;
}

interface MsgpackWsdbGetSiblingPath {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leafIndex: number;
}

interface MsgpackWsdbGetBlockNumbersForLeafIndices {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leafIndices: number[];
}

interface MsgpackWsdbFindLeafIndices {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leaves: Uint8Array[];
  startIndex: number;
}

interface MsgpackWsdbFindLowLeaf {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  key: Uint8Array;
}

interface MsgpackWsdbFindSiblingPaths {
  treeId: number;
  revision: MsgpackWorldStateRevision;
  leaves: Uint8Array[];
}

interface MsgpackWsdbAppendLeaves {
  treeId: number;
  leaves: Uint8Array[];
  forkId: number;
}

interface MsgpackWsdbBatchInsert {
  treeId: number;
  leaves: Uint8Array[];
  subtreeDepth: number;
  forkId: number;
}

interface MsgpackWsdbSequentialInsert {
  treeId: number;
  leaves: Uint8Array[];
  forkId: number;
}

interface MsgpackWsdbUpdateArchive {
  blockStateRef: Record<number, [Uint8Array, number]>;
  blockHeaderHash: Uint8Array;
  forkId: number;
}

interface MsgpackWsdbCommit {

}

interface MsgpackWsdbRollback {

}

interface MsgpackNullifierLeafValue {
  nullifier: Uint8Array;
}

interface MsgpackPublicDataLeafValue {
  slot: Uint8Array;
  value: Uint8Array;
}

interface MsgpackWsdbSyncBlock {
  blockNumber: number;
  blockStateRef: Record<number, [Uint8Array, number]>;
  blockHeaderHash: Uint8Array;
  paddedNoteHashes: Uint8Array[];
  paddedL1ToL2Messages: Uint8Array[];
  paddedNullifiers: MsgpackNullifierLeafValue[];
  publicDataWrites: MsgpackPublicDataLeafValue[];
}

interface MsgpackWsdbCreateFork {
  latest: boolean;
  blockNumber: number;
}

interface MsgpackWsdbDeleteFork {
  forkId: number;
}

interface MsgpackWsdbFinalizeBlocks {
  toBlockNumber: number;
}

interface MsgpackWsdbUnwindBlocks {
  toBlockNumber: number;
}

interface MsgpackWsdbRemoveHistoricalBlocks {
  toBlockNumber: number;
}

interface MsgpackWsdbGetStatus {

}

interface MsgpackWsdbCreateCheckpoint {
  forkId: number;
}

interface MsgpackWsdbCommitCheckpoint {
  forkId: number;
}

interface MsgpackWsdbRevertCheckpoint {
  forkId: number;
}

interface MsgpackWsdbCommitAllCheckpoints {
  forkId: number;
}

interface MsgpackWsdbRevertAllCheckpoints {
  forkId: number;
}

interface MsgpackWsdbCopyStores {
  dstPath: string;
  compact: boolean | undefined;
}

interface MsgpackWsdbShutdown {

}

interface MsgpackWsdbErrorResponse {
  message: string;
}

interface MsgpackWsdbGetTreeInfoResponse {
  treeId: number;
  root: Uint8Array;
  size: number;
  depth: number;
}

interface MsgpackWsdbGetStateReferenceResponse {
  state: Record<number, [Uint8Array, number]>;
}

interface MsgpackWsdbGetInitialStateReferenceResponse {
  state: Record<number, [Uint8Array, number]>;
}

interface MsgpackWsdbGetLeafValueResponse {
  value: Uint8Array | undefined;
}

interface MsgpackWsdbGetLeafPreimageResponse {
  preimage: Uint8Array | undefined;
}

interface MsgpackWsdbGetSiblingPathResponse {
  path: Uint8Array[];
}

interface MsgpackWsdbGetBlockNumbersForLeafIndicesResponse {
  blockNumbers: (number | undefined)[];
}

interface MsgpackWsdbFindLeafIndicesResponse {
  indices: (number | undefined)[];
}

interface MsgpackWsdbFindLowLeafResponse {
  alreadyPresent: boolean;
  index: number;
}

interface MsgpackWsdbFindSiblingPathsResponse {
  paths: (MsgpackSiblingPathAndIndex | undefined)[];
}

interface MsgpackWsdbAppendLeavesResponse {

}

interface MsgpackWsdbBatchInsertResponse {
  result: Uint8Array;
}

interface MsgpackWsdbSequentialInsertResponse {
  result: Uint8Array;
}

interface MsgpackWsdbUpdateArchiveResponse {

}

interface MsgpackWsdbCommitResponse {
  status: MsgpackWorldStateStatusFull;
}

interface MsgpackWsdbRollbackResponse {

}

interface MsgpackWsdbSyncBlockResponse {
  status: MsgpackWorldStateStatusFull;
}

interface MsgpackWsdbCreateForkResponse {
  forkId: number;
}

interface MsgpackWsdbDeleteForkResponse {

}

interface MsgpackWsdbFinalizeBlocksResponse {
  status: MsgpackWorldStateStatusSummary;
}

interface MsgpackWsdbUnwindBlocksResponse {
  status: MsgpackWorldStateStatusFull;
}

interface MsgpackWsdbRemoveHistoricalBlocksResponse {
  status: MsgpackWorldStateStatusFull;
}

interface MsgpackWsdbGetStatusResponse {
  status: MsgpackWorldStateStatusSummary;
}

interface MsgpackWsdbCreateCheckpointResponse {

}

interface MsgpackWsdbCommitCheckpointResponse {

}

interface MsgpackWsdbRevertCheckpointResponse {

}

interface MsgpackWsdbCommitAllCheckpointsResponse {

}

interface MsgpackWsdbRevertAllCheckpointsResponse {

}

interface MsgpackWsdbCopyStoresResponse {

}

interface MsgpackWsdbShutdownResponse {

}

// Conversion functions (exported)
export function toSiblingPathAndIndex(o: MsgpackSiblingPathAndIndex): SiblingPathAndIndex {
  if (o.index === undefined) { throw new Error("Expected index in SiblingPathAndIndex deserialization"); }
  if (o.path === undefined) { throw new Error("Expected path in SiblingPathAndIndex deserialization"); };
  return {
    index: o.index,
    path: o.path,
  };
}

export function toWorldStateStatusSummary(o: MsgpackWorldStateStatusSummary): WorldStateStatusSummary {
  if (o.unfinalizedBlockNumber === undefined) { throw new Error("Expected unfinalizedBlockNumber in WorldStateStatusSummary deserialization"); }
  if (o.finalizedBlockNumber === undefined) { throw new Error("Expected finalizedBlockNumber in WorldStateStatusSummary deserialization"); }
  if (o.oldestHistoricalBlock === undefined) { throw new Error("Expected oldestHistoricalBlock in WorldStateStatusSummary deserialization"); }
  if (o.treesAreSynched === undefined) { throw new Error("Expected treesAreSynched in WorldStateStatusSummary deserialization"); };
  return {
    unfinalizedblocknumber: o.unfinalizedBlockNumber,
    finalizedblocknumber: o.finalizedBlockNumber,
    oldesthistoricalblock: o.oldestHistoricalBlock,
    treesaresynched: o.treesAreSynched,
  };
}

export function toDBStats(o: MsgpackDBStats): DBStats {
  if (o.name === undefined) { throw new Error("Expected name in DBStats deserialization"); }
  if (o.numDataItems === undefined) { throw new Error("Expected numDataItems in DBStats deserialization"); }
  if (o.totalUsedSize === undefined) { throw new Error("Expected totalUsedSize in DBStats deserialization"); };
  return {
    name: o.name,
    numdataitems: o.numDataItems,
    totalusedsize: o.totalUsedSize,
  };
}

export function toTreeDBStats(o: MsgpackTreeDBStats): TreeDBStats {
  if (o.mapSize === undefined) { throw new Error("Expected mapSize in TreeDBStats deserialization"); }
  if (o.physicalFileSize === undefined) { throw new Error("Expected physicalFileSize in TreeDBStats deserialization"); }
  if (o.blocksDBStats === undefined) { throw new Error("Expected blocksDBStats in TreeDBStats deserialization"); }
  if (o.nodesDBStats === undefined) { throw new Error("Expected nodesDBStats in TreeDBStats deserialization"); }
  if (o.leafPreimagesDBStats === undefined) { throw new Error("Expected leafPreimagesDBStats in TreeDBStats deserialization"); }
  if (o.leafIndicesDBStats === undefined) { throw new Error("Expected leafIndicesDBStats in TreeDBStats deserialization"); }
  if (o.blockIndicesDBStats === undefined) { throw new Error("Expected blockIndicesDBStats in TreeDBStats deserialization"); };
  return {
    mapsize: o.mapSize,
    physicalfilesize: o.physicalFileSize,
    blocksdbstats: toDBStats(o.blocksDBStats),
    nodesdbstats: toDBStats(o.nodesDBStats),
    leafpreimagesdbstats: toDBStats(o.leafPreimagesDBStats),
    leafindicesdbstats: toDBStats(o.leafIndicesDBStats),
    blockindicesdbstats: toDBStats(o.blockIndicesDBStats),
  };
}

export function toWorldStateDBStats(o: MsgpackWorldStateDBStats): WorldStateDBStats {
  if (o.noteHashTreeStats === undefined) { throw new Error("Expected noteHashTreeStats in WorldStateDBStats deserialization"); }
  if (o.messageTreeStats === undefined) { throw new Error("Expected messageTreeStats in WorldStateDBStats deserialization"); }
  if (o.archiveTreeStats === undefined) { throw new Error("Expected archiveTreeStats in WorldStateDBStats deserialization"); }
  if (o.publicDataTreeStats === undefined) { throw new Error("Expected publicDataTreeStats in WorldStateDBStats deserialization"); }
  if (o.nullifierTreeStats === undefined) { throw new Error("Expected nullifierTreeStats in WorldStateDBStats deserialization"); };
  return {
    notehashtreestats: toTreeDBStats(o.noteHashTreeStats),
    messagetreestats: toTreeDBStats(o.messageTreeStats),
    archivetreestats: toTreeDBStats(o.archiveTreeStats),
    publicdatatreestats: toTreeDBStats(o.publicDataTreeStats),
    nullifiertreestats: toTreeDBStats(o.nullifierTreeStats),
  };
}

export function toTreeMeta(o: MsgpackTreeMeta): TreeMeta {
  if (o.name === undefined) { throw new Error("Expected name in TreeMeta deserialization"); }
  if (o.depth === undefined) { throw new Error("Expected depth in TreeMeta deserialization"); }
  if (o.size === undefined) { throw new Error("Expected size in TreeMeta deserialization"); }
  if (o.committedSize === undefined) { throw new Error("Expected committedSize in TreeMeta deserialization"); }
  if (o.root === undefined) { throw new Error("Expected root in TreeMeta deserialization"); }
  if (o.initialSize === undefined) { throw new Error("Expected initialSize in TreeMeta deserialization"); }
  if (o.initialRoot === undefined) { throw new Error("Expected initialRoot in TreeMeta deserialization"); }
  if (o.oldestHistoricBlock === undefined) { throw new Error("Expected oldestHistoricBlock in TreeMeta deserialization"); }
  if (o.unfinalizedBlockHeight === undefined) { throw new Error("Expected unfinalizedBlockHeight in TreeMeta deserialization"); }
  if (o.finalizedBlockHeight === undefined) { throw new Error("Expected finalizedBlockHeight in TreeMeta deserialization"); };
  return {
    name: o.name,
    depth: o.depth,
    size: o.size,
    committedsize: o.committedSize,
    root: o.root,
    initialsize: o.initialSize,
    initialroot: o.initialRoot,
    oldesthistoricblock: o.oldestHistoricBlock,
    unfinalizedblockheight: o.unfinalizedBlockHeight,
    finalizedblockheight: o.finalizedBlockHeight,
  };
}

export function toWorldStateMeta(o: MsgpackWorldStateMeta): WorldStateMeta {
  if (o.noteHashTreeMeta === undefined) { throw new Error("Expected noteHashTreeMeta in WorldStateMeta deserialization"); }
  if (o.messageTreeMeta === undefined) { throw new Error("Expected messageTreeMeta in WorldStateMeta deserialization"); }
  if (o.archiveTreeMeta === undefined) { throw new Error("Expected archiveTreeMeta in WorldStateMeta deserialization"); }
  if (o.publicDataTreeMeta === undefined) { throw new Error("Expected publicDataTreeMeta in WorldStateMeta deserialization"); }
  if (o.nullifierTreeMeta === undefined) { throw new Error("Expected nullifierTreeMeta in WorldStateMeta deserialization"); };
  return {
    notehashtreemeta: toTreeMeta(o.noteHashTreeMeta),
    messagetreemeta: toTreeMeta(o.messageTreeMeta),
    archivetreemeta: toTreeMeta(o.archiveTreeMeta),
    publicdatatreemeta: toTreeMeta(o.publicDataTreeMeta),
    nullifiertreemeta: toTreeMeta(o.nullifierTreeMeta),
  };
}

export function toWorldStateStatusFull(o: MsgpackWorldStateStatusFull): WorldStateStatusFull {
  if (o.summary === undefined) { throw new Error("Expected summary in WorldStateStatusFull deserialization"); }
  if (o.dbStats === undefined) { throw new Error("Expected dbStats in WorldStateStatusFull deserialization"); }
  if (o.meta === undefined) { throw new Error("Expected meta in WorldStateStatusFull deserialization"); };
  return {
    summary: toWorldStateStatusSummary(o.summary),
    dbstats: toWorldStateDBStats(o.dbStats),
    meta: toWorldStateMeta(o.meta),
  };
}

export function toWorldStateRevision(o: MsgpackWorldStateRevision): WorldStateRevision {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WorldStateRevision deserialization"); }
  if (o.blockNumber === undefined) { throw new Error("Expected blockNumber in WorldStateRevision deserialization"); }
  if (o.includeUncommitted === undefined) { throw new Error("Expected includeUncommitted in WorldStateRevision deserialization"); };
  return {
    forkid: o.forkId,
    blocknumber: o.blockNumber,
    includeuncommitted: o.includeUncommitted,
  };
}

export function toWsdbGetTreeInfo(o: MsgpackWsdbGetTreeInfo): WsdbGetTreeInfo {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetTreeInfo deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetTreeInfo deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
  };
}

export function toWsdbGetStateReference(o: MsgpackWsdbGetStateReference): WsdbGetStateReference {
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetStateReference deserialization"); };
  return {
    revision: toWorldStateRevision(o.revision),
  };
}

export function toWsdbGetInitialStateReference(o: MsgpackWsdbGetInitialStateReference): WsdbGetInitialStateReference {
  return {};
}

export function toWsdbGetLeafValue(o: MsgpackWsdbGetLeafValue): WsdbGetLeafValue {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetLeafValue deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetLeafValue deserialization"); }
  if (o.leafIndex === undefined) { throw new Error("Expected leafIndex in WsdbGetLeafValue deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leafindex: o.leafIndex,
  };
}

export function toWsdbGetLeafPreimage(o: MsgpackWsdbGetLeafPreimage): WsdbGetLeafPreimage {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetLeafPreimage deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetLeafPreimage deserialization"); }
  if (o.leafIndex === undefined) { throw new Error("Expected leafIndex in WsdbGetLeafPreimage deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leafindex: o.leafIndex,
  };
}

export function toWsdbGetSiblingPath(o: MsgpackWsdbGetSiblingPath): WsdbGetSiblingPath {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetSiblingPath deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetSiblingPath deserialization"); }
  if (o.leafIndex === undefined) { throw new Error("Expected leafIndex in WsdbGetSiblingPath deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leafindex: o.leafIndex,
  };
}

export function toWsdbGetBlockNumbersForLeafIndices(o: MsgpackWsdbGetBlockNumbersForLeafIndices): WsdbGetBlockNumbersForLeafIndices {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetBlockNumbersForLeafIndices deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices deserialization"); }
  if (o.leafIndices === undefined) { throw new Error("Expected leafIndices in WsdbGetBlockNumbersForLeafIndices deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leafindices: o.leafIndices,
  };
}

export function toWsdbFindLeafIndices(o: MsgpackWsdbFindLeafIndices): WsdbFindLeafIndices {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbFindLeafIndices deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindLeafIndices deserialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbFindLeafIndices deserialization"); }
  if (o.startIndex === undefined) { throw new Error("Expected startIndex in WsdbFindLeafIndices deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leaves: o.leaves,
    startindex: o.startIndex,
  };
}

export function toWsdbFindLowLeaf(o: MsgpackWsdbFindLowLeaf): WsdbFindLowLeaf {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbFindLowLeaf deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindLowLeaf deserialization"); }
  if (o.key === undefined) { throw new Error("Expected key in WsdbFindLowLeaf deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    key: o.key,
  };
}

export function toWsdbFindSiblingPaths(o: MsgpackWsdbFindSiblingPaths): WsdbFindSiblingPaths {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbFindSiblingPaths deserialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindSiblingPaths deserialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbFindSiblingPaths deserialization"); };
  return {
    treeid: o.treeId,
    revision: toWorldStateRevision(o.revision),
    leaves: o.leaves,
  };
}

export function toWsdbAppendLeaves(o: MsgpackWsdbAppendLeaves): WsdbAppendLeaves {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbAppendLeaves deserialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbAppendLeaves deserialization"); }
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbAppendLeaves deserialization"); };
  return {
    treeid: o.treeId,
    leaves: o.leaves,
    forkid: o.forkId,
  };
}

export function toWsdbBatchInsert(o: MsgpackWsdbBatchInsert): WsdbBatchInsert {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbBatchInsert deserialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbBatchInsert deserialization"); }
  if (o.subtreeDepth === undefined) { throw new Error("Expected subtreeDepth in WsdbBatchInsert deserialization"); }
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbBatchInsert deserialization"); };
  return {
    treeid: o.treeId,
    leaves: o.leaves,
    subtreedepth: o.subtreeDepth,
    forkid: o.forkId,
  };
}

export function toWsdbSequentialInsert(o: MsgpackWsdbSequentialInsert): WsdbSequentialInsert {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbSequentialInsert deserialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbSequentialInsert deserialization"); }
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbSequentialInsert deserialization"); };
  return {
    treeid: o.treeId,
    leaves: o.leaves,
    forkid: o.forkId,
  };
}

export function toWsdbUpdateArchive(o: MsgpackWsdbUpdateArchive): WsdbUpdateArchive {
  if (o.blockStateRef === undefined) { throw new Error("Expected blockStateRef in WsdbUpdateArchive deserialization"); }
  if (o.blockHeaderHash === undefined) { throw new Error("Expected blockHeaderHash in WsdbUpdateArchive deserialization"); }
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbUpdateArchive deserialization"); };
  return {
    blockstateref: o.blockStateRef,
    blockheaderhash: o.blockHeaderHash,
    forkid: o.forkId,
  };
}

export function toWsdbCommit(o: MsgpackWsdbCommit): WsdbCommit {
  return {};
}

export function toWsdbRollback(o: MsgpackWsdbRollback): WsdbRollback {
  return {};
}

export function toNullifierLeafValue(o: MsgpackNullifierLeafValue): NullifierLeafValue {
  if (o.nullifier === undefined) { throw new Error("Expected nullifier in NullifierLeafValue deserialization"); };
  return {
    nullifier: o.nullifier,
  };
}

export function toPublicDataLeafValue(o: MsgpackPublicDataLeafValue): PublicDataLeafValue {
  if (o.slot === undefined) { throw new Error("Expected slot in PublicDataLeafValue deserialization"); }
  if (o.value === undefined) { throw new Error("Expected value in PublicDataLeafValue deserialization"); };
  return {
    slot: o.slot,
    value: o.value,
  };
}

export function toWsdbSyncBlock(o: MsgpackWsdbSyncBlock): WsdbSyncBlock {
  if (o.blockNumber === undefined) { throw new Error("Expected blockNumber in WsdbSyncBlock deserialization"); }
  if (o.blockStateRef === undefined) { throw new Error("Expected blockStateRef in WsdbSyncBlock deserialization"); }
  if (o.blockHeaderHash === undefined) { throw new Error("Expected blockHeaderHash in WsdbSyncBlock deserialization"); }
  if (o.paddedNoteHashes === undefined) { throw new Error("Expected paddedNoteHashes in WsdbSyncBlock deserialization"); }
  if (o.paddedL1ToL2Messages === undefined) { throw new Error("Expected paddedL1ToL2Messages in WsdbSyncBlock deserialization"); }
  if (o.paddedNullifiers === undefined) { throw new Error("Expected paddedNullifiers in WsdbSyncBlock deserialization"); }
  if (o.publicDataWrites === undefined) { throw new Error("Expected publicDataWrites in WsdbSyncBlock deserialization"); };
  return {
    blocknumber: o.blockNumber,
    blockstateref: o.blockStateRef,
    blockheaderhash: o.blockHeaderHash,
    paddednotehashes: o.paddedNoteHashes,
    paddedl1tol2messages: o.paddedL1ToL2Messages,
    paddednullifiers: o.paddedNullifiers.map((v: any) => toNullifierLeafValue(v)),
    publicdatawrites: o.publicDataWrites.map((v: any) => toPublicDataLeafValue(v)),
  };
}

export function toWsdbCreateFork(o: MsgpackWsdbCreateFork): WsdbCreateFork {
  if (o.latest === undefined) { throw new Error("Expected latest in WsdbCreateFork deserialization"); }
  if (o.blockNumber === undefined) { throw new Error("Expected blockNumber in WsdbCreateFork deserialization"); };
  return {
    latest: o.latest,
    blocknumber: o.blockNumber,
  };
}

export function toWsdbDeleteFork(o: MsgpackWsdbDeleteFork): WsdbDeleteFork {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbDeleteFork deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbFinalizeBlocks(o: MsgpackWsdbFinalizeBlocks): WsdbFinalizeBlocks {
  if (o.toBlockNumber === undefined) { throw new Error("Expected toBlockNumber in WsdbFinalizeBlocks deserialization"); };
  return {
    toblocknumber: o.toBlockNumber,
  };
}

export function toWsdbUnwindBlocks(o: MsgpackWsdbUnwindBlocks): WsdbUnwindBlocks {
  if (o.toBlockNumber === undefined) { throw new Error("Expected toBlockNumber in WsdbUnwindBlocks deserialization"); };
  return {
    toblocknumber: o.toBlockNumber,
  };
}

export function toWsdbRemoveHistoricalBlocks(o: MsgpackWsdbRemoveHistoricalBlocks): WsdbRemoveHistoricalBlocks {
  if (o.toBlockNumber === undefined) { throw new Error("Expected toBlockNumber in WsdbRemoveHistoricalBlocks deserialization"); };
  return {
    toblocknumber: o.toBlockNumber,
  };
}

export function toWsdbGetStatus(o: MsgpackWsdbGetStatus): WsdbGetStatus {
  return {};
}

export function toWsdbCreateCheckpoint(o: MsgpackWsdbCreateCheckpoint): WsdbCreateCheckpoint {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbCreateCheckpoint deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbCommitCheckpoint(o: MsgpackWsdbCommitCheckpoint): WsdbCommitCheckpoint {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbCommitCheckpoint deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbRevertCheckpoint(o: MsgpackWsdbRevertCheckpoint): WsdbRevertCheckpoint {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbRevertCheckpoint deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbCommitAllCheckpoints(o: MsgpackWsdbCommitAllCheckpoints): WsdbCommitAllCheckpoints {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbCommitAllCheckpoints deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbRevertAllCheckpoints(o: MsgpackWsdbRevertAllCheckpoints): WsdbRevertAllCheckpoints {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbRevertAllCheckpoints deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbCopyStores(o: MsgpackWsdbCopyStores): WsdbCopyStores {
  if (o.dstPath === undefined) { throw new Error("Expected dstPath in WsdbCopyStores deserialization"); }
  if (o.compact === undefined) { throw new Error("Expected compact in WsdbCopyStores deserialization"); };
  return {
    dstpath: o.dstPath,
    compact: o.compact,
  };
}

export function toWsdbShutdown(o: MsgpackWsdbShutdown): WsdbShutdown {
  return {};
}

export function toWsdbErrorResponse(o: MsgpackWsdbErrorResponse): WsdbErrorResponse {
  if (o.message === undefined) { throw new Error("Expected message in WsdbErrorResponse deserialization"); };
  return {
    message: o.message,
  };
}

export function toWsdbGetTreeInfoResponse(o: MsgpackWsdbGetTreeInfoResponse): WsdbGetTreeInfoResponse {
  if (o.treeId === undefined) { throw new Error("Expected treeId in WsdbGetTreeInfoResponse deserialization"); }
  if (o.root === undefined) { throw new Error("Expected root in WsdbGetTreeInfoResponse deserialization"); }
  if (o.size === undefined) { throw new Error("Expected size in WsdbGetTreeInfoResponse deserialization"); }
  if (o.depth === undefined) { throw new Error("Expected depth in WsdbGetTreeInfoResponse deserialization"); };
  return {
    treeid: o.treeId,
    root: o.root,
    size: o.size,
    depth: o.depth,
  };
}

export function toWsdbGetStateReferenceResponse(o: MsgpackWsdbGetStateReferenceResponse): WsdbGetStateReferenceResponse {
  if (o.state === undefined) { throw new Error("Expected state in WsdbGetStateReferenceResponse deserialization"); };
  return {
    state: o.state,
  };
}

export function toWsdbGetInitialStateReferenceResponse(o: MsgpackWsdbGetInitialStateReferenceResponse): WsdbGetInitialStateReferenceResponse {
  if (o.state === undefined) { throw new Error("Expected state in WsdbGetInitialStateReferenceResponse deserialization"); };
  return {
    state: o.state,
  };
}

export function toWsdbGetLeafValueResponse(o: MsgpackWsdbGetLeafValueResponse): WsdbGetLeafValueResponse {
  if (o.value === undefined) { throw new Error("Expected value in WsdbGetLeafValueResponse deserialization"); };
  return {
    value: o.value,
  };
}

export function toWsdbGetLeafPreimageResponse(o: MsgpackWsdbGetLeafPreimageResponse): WsdbGetLeafPreimageResponse {
  if (o.preimage === undefined) { throw new Error("Expected preimage in WsdbGetLeafPreimageResponse deserialization"); };
  return {
    preimage: o.preimage,
  };
}

export function toWsdbGetSiblingPathResponse(o: MsgpackWsdbGetSiblingPathResponse): WsdbGetSiblingPathResponse {
  if (o.path === undefined) { throw new Error("Expected path in WsdbGetSiblingPathResponse deserialization"); };
  return {
    path: o.path,
  };
}

export function toWsdbGetBlockNumbersForLeafIndicesResponse(o: MsgpackWsdbGetBlockNumbersForLeafIndicesResponse): WsdbGetBlockNumbersForLeafIndicesResponse {
  if (o.blockNumbers === undefined) { throw new Error("Expected blockNumbers in WsdbGetBlockNumbersForLeafIndicesResponse deserialization"); };
  return {
    blocknumbers: o.blockNumbers,
  };
}

export function toWsdbFindLeafIndicesResponse(o: MsgpackWsdbFindLeafIndicesResponse): WsdbFindLeafIndicesResponse {
  if (o.indices === undefined) { throw new Error("Expected indices in WsdbFindLeafIndicesResponse deserialization"); };
  return {
    indices: o.indices,
  };
}

export function toWsdbFindLowLeafResponse(o: MsgpackWsdbFindLowLeafResponse): WsdbFindLowLeafResponse {
  if (o.alreadyPresent === undefined) { throw new Error("Expected alreadyPresent in WsdbFindLowLeafResponse deserialization"); }
  if (o.index === undefined) { throw new Error("Expected index in WsdbFindLowLeafResponse deserialization"); };
  return {
    alreadypresent: o.alreadyPresent,
    index: o.index,
  };
}

export function toWsdbFindSiblingPathsResponse(o: MsgpackWsdbFindSiblingPathsResponse): WsdbFindSiblingPathsResponse {
  if (o.paths === undefined) { throw new Error("Expected paths in WsdbFindSiblingPathsResponse deserialization"); };
  return {
    paths: o.paths.map((v: any) => v != null ? toSiblingPathAndIndex(v) : undefined),
  };
}

export function toWsdbAppendLeavesResponse(o: MsgpackWsdbAppendLeavesResponse): WsdbAppendLeavesResponse {
  return {};
}

export function toWsdbBatchInsertResponse(o: MsgpackWsdbBatchInsertResponse): WsdbBatchInsertResponse {
  if (o.result === undefined) { throw new Error("Expected result in WsdbBatchInsertResponse deserialization"); };
  return {
    result: o.result,
  };
}

export function toWsdbSequentialInsertResponse(o: MsgpackWsdbSequentialInsertResponse): WsdbSequentialInsertResponse {
  if (o.result === undefined) { throw new Error("Expected result in WsdbSequentialInsertResponse deserialization"); };
  return {
    result: o.result,
  };
}

export function toWsdbUpdateArchiveResponse(o: MsgpackWsdbUpdateArchiveResponse): WsdbUpdateArchiveResponse {
  return {};
}

export function toWsdbCommitResponse(o: MsgpackWsdbCommitResponse): WsdbCommitResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbCommitResponse deserialization"); };
  return {
    status: toWorldStateStatusFull(o.status),
  };
}

export function toWsdbRollbackResponse(o: MsgpackWsdbRollbackResponse): WsdbRollbackResponse {
  return {};
}

export function toWsdbSyncBlockResponse(o: MsgpackWsdbSyncBlockResponse): WsdbSyncBlockResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbSyncBlockResponse deserialization"); };
  return {
    status: toWorldStateStatusFull(o.status),
  };
}

export function toWsdbCreateForkResponse(o: MsgpackWsdbCreateForkResponse): WsdbCreateForkResponse {
  if (o.forkId === undefined) { throw new Error("Expected forkId in WsdbCreateForkResponse deserialization"); };
  return {
    forkid: o.forkId,
  };
}

export function toWsdbDeleteForkResponse(o: MsgpackWsdbDeleteForkResponse): WsdbDeleteForkResponse {
  return {};
}

export function toWsdbFinalizeBlocksResponse(o: MsgpackWsdbFinalizeBlocksResponse): WsdbFinalizeBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbFinalizeBlocksResponse deserialization"); };
  return {
    status: toWorldStateStatusSummary(o.status),
  };
}

export function toWsdbUnwindBlocksResponse(o: MsgpackWsdbUnwindBlocksResponse): WsdbUnwindBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbUnwindBlocksResponse deserialization"); };
  return {
    status: toWorldStateStatusFull(o.status),
  };
}

export function toWsdbRemoveHistoricalBlocksResponse(o: MsgpackWsdbRemoveHistoricalBlocksResponse): WsdbRemoveHistoricalBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse deserialization"); };
  return {
    status: toWorldStateStatusFull(o.status),
  };
}

export function toWsdbGetStatusResponse(o: MsgpackWsdbGetStatusResponse): WsdbGetStatusResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbGetStatusResponse deserialization"); };
  return {
    status: toWorldStateStatusSummary(o.status),
  };
}

export function toWsdbCreateCheckpointResponse(o: MsgpackWsdbCreateCheckpointResponse): WsdbCreateCheckpointResponse {
  return {};
}

export function toWsdbCommitCheckpointResponse(o: MsgpackWsdbCommitCheckpointResponse): WsdbCommitCheckpointResponse {
  return {};
}

export function toWsdbRevertCheckpointResponse(o: MsgpackWsdbRevertCheckpointResponse): WsdbRevertCheckpointResponse {
  return {};
}

export function toWsdbCommitAllCheckpointsResponse(o: MsgpackWsdbCommitAllCheckpointsResponse): WsdbCommitAllCheckpointsResponse {
  return {};
}

export function toWsdbRevertAllCheckpointsResponse(o: MsgpackWsdbRevertAllCheckpointsResponse): WsdbRevertAllCheckpointsResponse {
  return {};
}

export function toWsdbCopyStoresResponse(o: MsgpackWsdbCopyStoresResponse): WsdbCopyStoresResponse {
  return {};
}

export function toWsdbShutdownResponse(o: MsgpackWsdbShutdownResponse): WsdbShutdownResponse {
  return {};
}

export function fromSiblingPathAndIndex(o: SiblingPathAndIndex): MsgpackSiblingPathAndIndex {
  if (o.index === undefined) { throw new Error("Expected index in SiblingPathAndIndex serialization"); }
  if (o.path === undefined) { throw new Error("Expected path in SiblingPathAndIndex serialization"); };
  return {
  index: o.index,
  path: o.path,
  };
}

export function fromWorldStateStatusSummary(o: WorldStateStatusSummary): MsgpackWorldStateStatusSummary {
  if (o.unfinalizedblocknumber === undefined) { throw new Error("Expected unfinalizedblocknumber in WorldStateStatusSummary serialization"); }
  if (o.finalizedblocknumber === undefined) { throw new Error("Expected finalizedblocknumber in WorldStateStatusSummary serialization"); }
  if (o.oldesthistoricalblock === undefined) { throw new Error("Expected oldesthistoricalblock in WorldStateStatusSummary serialization"); }
  if (o.treesaresynched === undefined) { throw new Error("Expected treesaresynched in WorldStateStatusSummary serialization"); };
  return {
  unfinalizedBlockNumber: o.unfinalizedblocknumber,
  finalizedBlockNumber: o.finalizedblocknumber,
  oldestHistoricalBlock: o.oldesthistoricalblock,
  treesAreSynched: o.treesaresynched,
  };
}

export function fromDBStats(o: DBStats): MsgpackDBStats {
  if (o.name === undefined) { throw new Error("Expected name in DBStats serialization"); }
  if (o.numdataitems === undefined) { throw new Error("Expected numdataitems in DBStats serialization"); }
  if (o.totalusedsize === undefined) { throw new Error("Expected totalusedsize in DBStats serialization"); };
  return {
  name: o.name,
  numDataItems: o.numdataitems,
  totalUsedSize: o.totalusedsize,
  };
}

export function fromTreeDBStats(o: TreeDBStats): MsgpackTreeDBStats {
  if (o.mapsize === undefined) { throw new Error("Expected mapsize in TreeDBStats serialization"); }
  if (o.physicalfilesize === undefined) { throw new Error("Expected physicalfilesize in TreeDBStats serialization"); }
  if (o.blocksdbstats === undefined) { throw new Error("Expected blocksdbstats in TreeDBStats serialization"); }
  if (o.nodesdbstats === undefined) { throw new Error("Expected nodesdbstats in TreeDBStats serialization"); }
  if (o.leafpreimagesdbstats === undefined) { throw new Error("Expected leafpreimagesdbstats in TreeDBStats serialization"); }
  if (o.leafindicesdbstats === undefined) { throw new Error("Expected leafindicesdbstats in TreeDBStats serialization"); }
  if (o.blockindicesdbstats === undefined) { throw new Error("Expected blockindicesdbstats in TreeDBStats serialization"); };
  return {
  mapSize: o.mapsize,
  physicalFileSize: o.physicalfilesize,
  blocksDBStats: fromDBStats(o.blocksdbstats),
  nodesDBStats: fromDBStats(o.nodesdbstats),
  leafPreimagesDBStats: fromDBStats(o.leafpreimagesdbstats),
  leafIndicesDBStats: fromDBStats(o.leafindicesdbstats),
  blockIndicesDBStats: fromDBStats(o.blockindicesdbstats),
  };
}

export function fromWorldStateDBStats(o: WorldStateDBStats): MsgpackWorldStateDBStats {
  if (o.notehashtreestats === undefined) { throw new Error("Expected notehashtreestats in WorldStateDBStats serialization"); }
  if (o.messagetreestats === undefined) { throw new Error("Expected messagetreestats in WorldStateDBStats serialization"); }
  if (o.archivetreestats === undefined) { throw new Error("Expected archivetreestats in WorldStateDBStats serialization"); }
  if (o.publicdatatreestats === undefined) { throw new Error("Expected publicdatatreestats in WorldStateDBStats serialization"); }
  if (o.nullifiertreestats === undefined) { throw new Error("Expected nullifiertreestats in WorldStateDBStats serialization"); };
  return {
  noteHashTreeStats: fromTreeDBStats(o.notehashtreestats),
  messageTreeStats: fromTreeDBStats(o.messagetreestats),
  archiveTreeStats: fromTreeDBStats(o.archivetreestats),
  publicDataTreeStats: fromTreeDBStats(o.publicdatatreestats),
  nullifierTreeStats: fromTreeDBStats(o.nullifiertreestats),
  };
}

export function fromTreeMeta(o: TreeMeta): MsgpackTreeMeta {
  if (o.name === undefined) { throw new Error("Expected name in TreeMeta serialization"); }
  if (o.depth === undefined) { throw new Error("Expected depth in TreeMeta serialization"); }
  if (o.size === undefined) { throw new Error("Expected size in TreeMeta serialization"); }
  if (o.committedsize === undefined) { throw new Error("Expected committedsize in TreeMeta serialization"); }
  if (o.root === undefined) { throw new Error("Expected root in TreeMeta serialization"); }
  if (o.initialsize === undefined) { throw new Error("Expected initialsize in TreeMeta serialization"); }
  if (o.initialroot === undefined) { throw new Error("Expected initialroot in TreeMeta serialization"); }
  if (o.oldesthistoricblock === undefined) { throw new Error("Expected oldesthistoricblock in TreeMeta serialization"); }
  if (o.unfinalizedblockheight === undefined) { throw new Error("Expected unfinalizedblockheight in TreeMeta serialization"); }
  if (o.finalizedblockheight === undefined) { throw new Error("Expected finalizedblockheight in TreeMeta serialization"); };
  return {
  name: o.name,
  depth: o.depth,
  size: o.size,
  committedSize: o.committedsize,
  root: o.root,
  initialSize: o.initialsize,
  initialRoot: o.initialroot,
  oldestHistoricBlock: o.oldesthistoricblock,
  unfinalizedBlockHeight: o.unfinalizedblockheight,
  finalizedBlockHeight: o.finalizedblockheight,
  };
}

export function fromWorldStateMeta(o: WorldStateMeta): MsgpackWorldStateMeta {
  if (o.notehashtreemeta === undefined) { throw new Error("Expected notehashtreemeta in WorldStateMeta serialization"); }
  if (o.messagetreemeta === undefined) { throw new Error("Expected messagetreemeta in WorldStateMeta serialization"); }
  if (o.archivetreemeta === undefined) { throw new Error("Expected archivetreemeta in WorldStateMeta serialization"); }
  if (o.publicdatatreemeta === undefined) { throw new Error("Expected publicdatatreemeta in WorldStateMeta serialization"); }
  if (o.nullifiertreemeta === undefined) { throw new Error("Expected nullifiertreemeta in WorldStateMeta serialization"); };
  return {
  noteHashTreeMeta: fromTreeMeta(o.notehashtreemeta),
  messageTreeMeta: fromTreeMeta(o.messagetreemeta),
  archiveTreeMeta: fromTreeMeta(o.archivetreemeta),
  publicDataTreeMeta: fromTreeMeta(o.publicdatatreemeta),
  nullifierTreeMeta: fromTreeMeta(o.nullifiertreemeta),
  };
}

export function fromWorldStateStatusFull(o: WorldStateStatusFull): MsgpackWorldStateStatusFull {
  if (o.summary === undefined) { throw new Error("Expected summary in WorldStateStatusFull serialization"); }
  if (o.dbstats === undefined) { throw new Error("Expected dbstats in WorldStateStatusFull serialization"); }
  if (o.meta === undefined) { throw new Error("Expected meta in WorldStateStatusFull serialization"); };
  return {
  summary: fromWorldStateStatusSummary(o.summary),
  dbStats: fromWorldStateDBStats(o.dbstats),
  meta: fromWorldStateMeta(o.meta),
  };
}

export function fromWorldStateRevision(o: WorldStateRevision): MsgpackWorldStateRevision {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WorldStateRevision serialization"); }
  if (o.blocknumber === undefined) { throw new Error("Expected blocknumber in WorldStateRevision serialization"); }
  if (o.includeuncommitted === undefined) { throw new Error("Expected includeuncommitted in WorldStateRevision serialization"); };
  return {
  forkId: o.forkid,
  blockNumber: o.blocknumber,
  includeUncommitted: o.includeuncommitted,
  };
}

export function fromWsdbGetTreeInfo(o: WsdbGetTreeInfo): MsgpackWsdbGetTreeInfo {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetTreeInfo serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetTreeInfo serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  };
}

export function fromWsdbGetStateReference(o: WsdbGetStateReference): MsgpackWsdbGetStateReference {
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetStateReference serialization"); };
  return {
  revision: fromWorldStateRevision(o.revision),
  };
}

export function fromWsdbGetInitialStateReference(o: WsdbGetInitialStateReference): MsgpackWsdbGetInitialStateReference {
  return {};
}

export function fromWsdbGetLeafValue(o: WsdbGetLeafValue): MsgpackWsdbGetLeafValue {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetLeafValue serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetLeafValue serialization"); }
  if (o.leafindex === undefined) { throw new Error("Expected leafindex in WsdbGetLeafValue serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leafIndex: o.leafindex,
  };
}

export function fromWsdbGetLeafPreimage(o: WsdbGetLeafPreimage): MsgpackWsdbGetLeafPreimage {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetLeafPreimage serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetLeafPreimage serialization"); }
  if (o.leafindex === undefined) { throw new Error("Expected leafindex in WsdbGetLeafPreimage serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leafIndex: o.leafindex,
  };
}

export function fromWsdbGetSiblingPath(o: WsdbGetSiblingPath): MsgpackWsdbGetSiblingPath {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetSiblingPath serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetSiblingPath serialization"); }
  if (o.leafindex === undefined) { throw new Error("Expected leafindex in WsdbGetSiblingPath serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leafIndex: o.leafindex,
  };
}

export function fromWsdbGetBlockNumbersForLeafIndices(o: WsdbGetBlockNumbersForLeafIndices): MsgpackWsdbGetBlockNumbersForLeafIndices {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetBlockNumbersForLeafIndices serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbGetBlockNumbersForLeafIndices serialization"); }
  if (o.leafindices === undefined) { throw new Error("Expected leafindices in WsdbGetBlockNumbersForLeafIndices serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leafIndices: o.leafindices,
  };
}

export function fromWsdbFindLeafIndices(o: WsdbFindLeafIndices): MsgpackWsdbFindLeafIndices {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbFindLeafIndices serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindLeafIndices serialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbFindLeafIndices serialization"); }
  if (o.startindex === undefined) { throw new Error("Expected startindex in WsdbFindLeafIndices serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leaves: o.leaves,
  startIndex: o.startindex,
  };
}

export function fromWsdbFindLowLeaf(o: WsdbFindLowLeaf): MsgpackWsdbFindLowLeaf {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbFindLowLeaf serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindLowLeaf serialization"); }
  if (o.key === undefined) { throw new Error("Expected key in WsdbFindLowLeaf serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  key: o.key,
  };
}

export function fromWsdbFindSiblingPaths(o: WsdbFindSiblingPaths): MsgpackWsdbFindSiblingPaths {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbFindSiblingPaths serialization"); }
  if (o.revision === undefined) { throw new Error("Expected revision in WsdbFindSiblingPaths serialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbFindSiblingPaths serialization"); };
  return {
  treeId: o.treeid,
  revision: fromWorldStateRevision(o.revision),
  leaves: o.leaves,
  };
}

export function fromWsdbAppendLeaves(o: WsdbAppendLeaves): MsgpackWsdbAppendLeaves {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbAppendLeaves serialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbAppendLeaves serialization"); }
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbAppendLeaves serialization"); };
  return {
  treeId: o.treeid,
  leaves: o.leaves,
  forkId: o.forkid,
  };
}

export function fromWsdbBatchInsert(o: WsdbBatchInsert): MsgpackWsdbBatchInsert {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbBatchInsert serialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbBatchInsert serialization"); }
  if (o.subtreedepth === undefined) { throw new Error("Expected subtreedepth in WsdbBatchInsert serialization"); }
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbBatchInsert serialization"); };
  return {
  treeId: o.treeid,
  leaves: o.leaves,
  subtreeDepth: o.subtreedepth,
  forkId: o.forkid,
  };
}

export function fromWsdbSequentialInsert(o: WsdbSequentialInsert): MsgpackWsdbSequentialInsert {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbSequentialInsert serialization"); }
  if (o.leaves === undefined) { throw new Error("Expected leaves in WsdbSequentialInsert serialization"); }
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbSequentialInsert serialization"); };
  return {
  treeId: o.treeid,
  leaves: o.leaves,
  forkId: o.forkid,
  };
}

export function fromWsdbUpdateArchive(o: WsdbUpdateArchive): MsgpackWsdbUpdateArchive {
  if (o.blockstateref === undefined) { throw new Error("Expected blockstateref in WsdbUpdateArchive serialization"); }
  if (o.blockheaderhash === undefined) { throw new Error("Expected blockheaderhash in WsdbUpdateArchive serialization"); }
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbUpdateArchive serialization"); };
  return {
  blockStateRef: o.blockstateref,
  blockHeaderHash: o.blockheaderhash,
  forkId: o.forkid,
  };
}

export function fromWsdbCommit(o: WsdbCommit): MsgpackWsdbCommit {
  return {};
}

export function fromWsdbRollback(o: WsdbRollback): MsgpackWsdbRollback {
  return {};
}

export function fromNullifierLeafValue(o: NullifierLeafValue): MsgpackNullifierLeafValue {
  if (o.nullifier === undefined) { throw new Error("Expected nullifier in NullifierLeafValue serialization"); };
  return {
  nullifier: o.nullifier,
  };
}

export function fromPublicDataLeafValue(o: PublicDataLeafValue): MsgpackPublicDataLeafValue {
  if (o.slot === undefined) { throw new Error("Expected slot in PublicDataLeafValue serialization"); }
  if (o.value === undefined) { throw new Error("Expected value in PublicDataLeafValue serialization"); };
  return {
  slot: o.slot,
  value: o.value,
  };
}

export function fromWsdbSyncBlock(o: WsdbSyncBlock): MsgpackWsdbSyncBlock {
  if (o.blocknumber === undefined) { throw new Error("Expected blocknumber in WsdbSyncBlock serialization"); }
  if (o.blockstateref === undefined) { throw new Error("Expected blockstateref in WsdbSyncBlock serialization"); }
  if (o.blockheaderhash === undefined) { throw new Error("Expected blockheaderhash in WsdbSyncBlock serialization"); }
  if (o.paddednotehashes === undefined) { throw new Error("Expected paddednotehashes in WsdbSyncBlock serialization"); }
  if (o.paddedl1tol2messages === undefined) { throw new Error("Expected paddedl1tol2messages in WsdbSyncBlock serialization"); }
  if (o.paddednullifiers === undefined) { throw new Error("Expected paddednullifiers in WsdbSyncBlock serialization"); }
  if (o.publicdatawrites === undefined) { throw new Error("Expected publicdatawrites in WsdbSyncBlock serialization"); };
  return {
  blockNumber: o.blocknumber,
  blockStateRef: o.blockstateref,
  blockHeaderHash: o.blockheaderhash,
  paddedNoteHashes: o.paddednotehashes,
  paddedL1ToL2Messages: o.paddedl1tol2messages,
  paddedNullifiers: o.paddednullifiers.map((v: any) => fromNullifierLeafValue(v)),
  publicDataWrites: o.publicdatawrites.map((v: any) => fromPublicDataLeafValue(v)),
  };
}

export function fromWsdbCreateFork(o: WsdbCreateFork): MsgpackWsdbCreateFork {
  if (o.latest === undefined) { throw new Error("Expected latest in WsdbCreateFork serialization"); }
  if (o.blocknumber === undefined) { throw new Error("Expected blocknumber in WsdbCreateFork serialization"); };
  return {
  latest: o.latest,
  blockNumber: o.blocknumber,
  };
}

export function fromWsdbDeleteFork(o: WsdbDeleteFork): MsgpackWsdbDeleteFork {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbDeleteFork serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbFinalizeBlocks(o: WsdbFinalizeBlocks): MsgpackWsdbFinalizeBlocks {
  if (o.toblocknumber === undefined) { throw new Error("Expected toblocknumber in WsdbFinalizeBlocks serialization"); };
  return {
  toBlockNumber: o.toblocknumber,
  };
}

export function fromWsdbUnwindBlocks(o: WsdbUnwindBlocks): MsgpackWsdbUnwindBlocks {
  if (o.toblocknumber === undefined) { throw new Error("Expected toblocknumber in WsdbUnwindBlocks serialization"); };
  return {
  toBlockNumber: o.toblocknumber,
  };
}

export function fromWsdbRemoveHistoricalBlocks(o: WsdbRemoveHistoricalBlocks): MsgpackWsdbRemoveHistoricalBlocks {
  if (o.toblocknumber === undefined) { throw new Error("Expected toblocknumber in WsdbRemoveHistoricalBlocks serialization"); };
  return {
  toBlockNumber: o.toblocknumber,
  };
}

export function fromWsdbGetStatus(o: WsdbGetStatus): MsgpackWsdbGetStatus {
  return {};
}

export function fromWsdbCreateCheckpoint(o: WsdbCreateCheckpoint): MsgpackWsdbCreateCheckpoint {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbCreateCheckpoint serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbCommitCheckpoint(o: WsdbCommitCheckpoint): MsgpackWsdbCommitCheckpoint {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbCommitCheckpoint serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbRevertCheckpoint(o: WsdbRevertCheckpoint): MsgpackWsdbRevertCheckpoint {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbRevertCheckpoint serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbCommitAllCheckpoints(o: WsdbCommitAllCheckpoints): MsgpackWsdbCommitAllCheckpoints {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbCommitAllCheckpoints serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbRevertAllCheckpoints(o: WsdbRevertAllCheckpoints): MsgpackWsdbRevertAllCheckpoints {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbRevertAllCheckpoints serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbCopyStores(o: WsdbCopyStores): MsgpackWsdbCopyStores {
  if (o.dstpath === undefined) { throw new Error("Expected dstpath in WsdbCopyStores serialization"); }
  if (o.compact === undefined) { throw new Error("Expected compact in WsdbCopyStores serialization"); };
  return {
  dstPath: o.dstpath,
  compact: o.compact,
  };
}

export function fromWsdbShutdown(o: WsdbShutdown): MsgpackWsdbShutdown {
  return {};
}

export function fromWsdbErrorResponse(o: WsdbErrorResponse): MsgpackWsdbErrorResponse {
  if (o.message === undefined) { throw new Error("Expected message in WsdbErrorResponse serialization"); };
  return {
  message: o.message,
  };
}

export function fromWsdbGetTreeInfoResponse(o: WsdbGetTreeInfoResponse): MsgpackWsdbGetTreeInfoResponse {
  if (o.treeid === undefined) { throw new Error("Expected treeid in WsdbGetTreeInfoResponse serialization"); }
  if (o.root === undefined) { throw new Error("Expected root in WsdbGetTreeInfoResponse serialization"); }
  if (o.size === undefined) { throw new Error("Expected size in WsdbGetTreeInfoResponse serialization"); }
  if (o.depth === undefined) { throw new Error("Expected depth in WsdbGetTreeInfoResponse serialization"); };
  return {
  treeId: o.treeid,
  root: o.root,
  size: o.size,
  depth: o.depth,
  };
}

export function fromWsdbGetStateReferenceResponse(o: WsdbGetStateReferenceResponse): MsgpackWsdbGetStateReferenceResponse {
  if (o.state === undefined) { throw new Error("Expected state in WsdbGetStateReferenceResponse serialization"); };
  return {
  state: o.state,
  };
}

export function fromWsdbGetInitialStateReferenceResponse(o: WsdbGetInitialStateReferenceResponse): MsgpackWsdbGetInitialStateReferenceResponse {
  if (o.state === undefined) { throw new Error("Expected state in WsdbGetInitialStateReferenceResponse serialization"); };
  return {
  state: o.state,
  };
}

export function fromWsdbGetLeafValueResponse(o: WsdbGetLeafValueResponse): MsgpackWsdbGetLeafValueResponse {
  if (o.value === undefined) { throw new Error("Expected value in WsdbGetLeafValueResponse serialization"); };
  return {
  value: o.value,
  };
}

export function fromWsdbGetLeafPreimageResponse(o: WsdbGetLeafPreimageResponse): MsgpackWsdbGetLeafPreimageResponse {
  if (o.preimage === undefined) { throw new Error("Expected preimage in WsdbGetLeafPreimageResponse serialization"); };
  return {
  preimage: o.preimage,
  };
}

export function fromWsdbGetSiblingPathResponse(o: WsdbGetSiblingPathResponse): MsgpackWsdbGetSiblingPathResponse {
  if (o.path === undefined) { throw new Error("Expected path in WsdbGetSiblingPathResponse serialization"); };
  return {
  path: o.path,
  };
}

export function fromWsdbGetBlockNumbersForLeafIndicesResponse(o: WsdbGetBlockNumbersForLeafIndicesResponse): MsgpackWsdbGetBlockNumbersForLeafIndicesResponse {
  if (o.blocknumbers === undefined) { throw new Error("Expected blocknumbers in WsdbGetBlockNumbersForLeafIndicesResponse serialization"); };
  return {
  blockNumbers: o.blocknumbers,
  };
}

export function fromWsdbFindLeafIndicesResponse(o: WsdbFindLeafIndicesResponse): MsgpackWsdbFindLeafIndicesResponse {
  if (o.indices === undefined) { throw new Error("Expected indices in WsdbFindLeafIndicesResponse serialization"); };
  return {
  indices: o.indices,
  };
}

export function fromWsdbFindLowLeafResponse(o: WsdbFindLowLeafResponse): MsgpackWsdbFindLowLeafResponse {
  if (o.alreadypresent === undefined) { throw new Error("Expected alreadypresent in WsdbFindLowLeafResponse serialization"); }
  if (o.index === undefined) { throw new Error("Expected index in WsdbFindLowLeafResponse serialization"); };
  return {
  alreadyPresent: o.alreadypresent,
  index: o.index,
  };
}

export function fromWsdbFindSiblingPathsResponse(o: WsdbFindSiblingPathsResponse): MsgpackWsdbFindSiblingPathsResponse {
  if (o.paths === undefined) { throw new Error("Expected paths in WsdbFindSiblingPathsResponse serialization"); };
  return {
  paths: o.paths.map((v: any) => v != null ? fromSiblingPathAndIndex(v) : undefined),
  };
}

export function fromWsdbAppendLeavesResponse(o: WsdbAppendLeavesResponse): MsgpackWsdbAppendLeavesResponse {
  return {};
}

export function fromWsdbBatchInsertResponse(o: WsdbBatchInsertResponse): MsgpackWsdbBatchInsertResponse {
  if (o.result === undefined) { throw new Error("Expected result in WsdbBatchInsertResponse serialization"); };
  return {
  result: o.result,
  };
}

export function fromWsdbSequentialInsertResponse(o: WsdbSequentialInsertResponse): MsgpackWsdbSequentialInsertResponse {
  if (o.result === undefined) { throw new Error("Expected result in WsdbSequentialInsertResponse serialization"); };
  return {
  result: o.result,
  };
}

export function fromWsdbUpdateArchiveResponse(o: WsdbUpdateArchiveResponse): MsgpackWsdbUpdateArchiveResponse {
  return {};
}

export function fromWsdbCommitResponse(o: WsdbCommitResponse): MsgpackWsdbCommitResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbCommitResponse serialization"); };
  return {
  status: fromWorldStateStatusFull(o.status),
  };
}

export function fromWsdbRollbackResponse(o: WsdbRollbackResponse): MsgpackWsdbRollbackResponse {
  return {};
}

export function fromWsdbSyncBlockResponse(o: WsdbSyncBlockResponse): MsgpackWsdbSyncBlockResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbSyncBlockResponse serialization"); };
  return {
  status: fromWorldStateStatusFull(o.status),
  };
}

export function fromWsdbCreateForkResponse(o: WsdbCreateForkResponse): MsgpackWsdbCreateForkResponse {
  if (o.forkid === undefined) { throw new Error("Expected forkid in WsdbCreateForkResponse serialization"); };
  return {
  forkId: o.forkid,
  };
}

export function fromWsdbDeleteForkResponse(o: WsdbDeleteForkResponse): MsgpackWsdbDeleteForkResponse {
  return {};
}

export function fromWsdbFinalizeBlocksResponse(o: WsdbFinalizeBlocksResponse): MsgpackWsdbFinalizeBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbFinalizeBlocksResponse serialization"); };
  return {
  status: fromWorldStateStatusSummary(o.status),
  };
}

export function fromWsdbUnwindBlocksResponse(o: WsdbUnwindBlocksResponse): MsgpackWsdbUnwindBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbUnwindBlocksResponse serialization"); };
  return {
  status: fromWorldStateStatusFull(o.status),
  };
}

export function fromWsdbRemoveHistoricalBlocksResponse(o: WsdbRemoveHistoricalBlocksResponse): MsgpackWsdbRemoveHistoricalBlocksResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbRemoveHistoricalBlocksResponse serialization"); };
  return {
  status: fromWorldStateStatusFull(o.status),
  };
}

export function fromWsdbGetStatusResponse(o: WsdbGetStatusResponse): MsgpackWsdbGetStatusResponse {
  if (o.status === undefined) { throw new Error("Expected status in WsdbGetStatusResponse serialization"); };
  return {
  status: fromWorldStateStatusSummary(o.status),
  };
}

export function fromWsdbCreateCheckpointResponse(o: WsdbCreateCheckpointResponse): MsgpackWsdbCreateCheckpointResponse {
  return {};
}

export function fromWsdbCommitCheckpointResponse(o: WsdbCommitCheckpointResponse): MsgpackWsdbCommitCheckpointResponse {
  return {};
}

export function fromWsdbRevertCheckpointResponse(o: WsdbRevertCheckpointResponse): MsgpackWsdbRevertCheckpointResponse {
  return {};
}

export function fromWsdbCommitAllCheckpointsResponse(o: WsdbCommitAllCheckpointsResponse): MsgpackWsdbCommitAllCheckpointsResponse {
  return {};
}

export function fromWsdbRevertAllCheckpointsResponse(o: WsdbRevertAllCheckpointsResponse): MsgpackWsdbRevertAllCheckpointsResponse {
  return {};
}

export function fromWsdbCopyStoresResponse(o: WsdbCopyStoresResponse): MsgpackWsdbCopyStoresResponse {
  return {};
}

export function fromWsdbShutdownResponse(o: WsdbShutdownResponse): MsgpackWsdbShutdownResponse {
  return {};
}

// Base API interface
export interface BbApiBase {
  wsdbGetTreeInfo(command: WsdbGetTreeInfo): Promise<WsdbGetTreeInfoResponse>;
  wsdbGetStateReference(command: WsdbGetStateReference): Promise<WsdbGetStateReferenceResponse>;
  wsdbGetInitialStateReference(command: WsdbGetInitialStateReference): Promise<WsdbGetInitialStateReferenceResponse>;
  wsdbGetLeafValue(command: WsdbGetLeafValue): Promise<WsdbGetLeafValueResponse>;
  wsdbGetLeafPreimage(command: WsdbGetLeafPreimage): Promise<WsdbGetLeafPreimageResponse>;
  wsdbGetSiblingPath(command: WsdbGetSiblingPath): Promise<WsdbGetSiblingPathResponse>;
  wsdbGetBlockNumbersForLeafIndices(command: WsdbGetBlockNumbersForLeafIndices): Promise<WsdbGetBlockNumbersForLeafIndicesResponse>;
  wsdbFindLeafIndices(command: WsdbFindLeafIndices): Promise<WsdbFindLeafIndicesResponse>;
  wsdbFindLowLeaf(command: WsdbFindLowLeaf): Promise<WsdbFindLowLeafResponse>;
  wsdbFindSiblingPaths(command: WsdbFindSiblingPaths): Promise<WsdbFindSiblingPathsResponse>;
  wsdbAppendLeaves(command: WsdbAppendLeaves): Promise<WsdbAppendLeavesResponse>;
  wsdbBatchInsert(command: WsdbBatchInsert): Promise<WsdbBatchInsertResponse>;
  wsdbSequentialInsert(command: WsdbSequentialInsert): Promise<WsdbSequentialInsertResponse>;
  wsdbUpdateArchive(command: WsdbUpdateArchive): Promise<WsdbUpdateArchiveResponse>;
  wsdbCommit(command: WsdbCommit): Promise<WsdbCommitResponse>;
  wsdbRollback(command: WsdbRollback): Promise<WsdbRollbackResponse>;
  wsdbSyncBlock(command: WsdbSyncBlock): Promise<WsdbSyncBlockResponse>;
  wsdbCreateFork(command: WsdbCreateFork): Promise<WsdbCreateForkResponse>;
  wsdbDeleteFork(command: WsdbDeleteFork): Promise<WsdbDeleteForkResponse>;
  wsdbFinalizeBlocks(command: WsdbFinalizeBlocks): Promise<WsdbFinalizeBlocksResponse>;
  wsdbUnwindBlocks(command: WsdbUnwindBlocks): Promise<WsdbUnwindBlocksResponse>;
  wsdbRemoveHistoricalBlocks(command: WsdbRemoveHistoricalBlocks): Promise<WsdbRemoveHistoricalBlocksResponse>;
  wsdbGetStatus(command: WsdbGetStatus): Promise<WsdbGetStatusResponse>;
  wsdbCreateCheckpoint(command: WsdbCreateCheckpoint): Promise<WsdbCreateCheckpointResponse>;
  wsdbCommitCheckpoint(command: WsdbCommitCheckpoint): Promise<WsdbCommitCheckpointResponse>;
  wsdbRevertCheckpoint(command: WsdbRevertCheckpoint): Promise<WsdbRevertCheckpointResponse>;
  wsdbCommitAllCheckpoints(command: WsdbCommitAllCheckpoints): Promise<WsdbCommitAllCheckpointsResponse>;
  wsdbRevertAllCheckpoints(command: WsdbRevertAllCheckpoints): Promise<WsdbRevertAllCheckpointsResponse>;
  wsdbCopyStores(command: WsdbCopyStores): Promise<WsdbCopyStoresResponse>;
  wsdbShutdown(command: WsdbShutdown): Promise<WsdbShutdownResponse>;
  destroy(): Promise<void>;
}
